{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Next Steps (1/5/21)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1/5/21"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- get signed variable importances for regression models\n",
    "\n",
    "- case study for strongest TSS model: do genes with strong secondary TSS do worse?\n",
    "\n",
    "- what is minimal set of features we can augment 1-hot with to get better generalization?\n",
    "    - simplify 1-hot via featurization\n",
    "    - augment 1-hot with cherry-picked additional features\n",
    "    \n",
    "- take first step towards an inventory of feature engineering decisions/parameters\n",
    "\n",
    "- divide genes by expression and fit individual regressors on each subset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12/18/20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- get feedback\n",
    "- error analysis -> which cases are the models getting right vs wrong?\n",
    "- look at variable importances for SVR and RF\n",
    "- can we classify low, medium, and highly expressed genes? \n",
    "    - and if so, can we make quantitative models for just these buckets?\n",
    "- how do we handle slightly offset TSS for same promoter?\n",
    "    - investigate these cases and compare: offset between TSS, difference in TSS signal height\n",
    "    - try to align promoters by -10 box, not TSS\n",
    "        - profile: is this an issue?\n",
    "- within TU vs random expression for E. coli\n",
    "- [lost to aether of Dan's mind]\n",
    "    \n",
    "- Do our features incorporate base pair relationships?\n",
    "    - kmers -> count di- or tri-nucleotides\n",
    "        - drawback: blows up the number of features\n",
    "    - one-hot -> no\n",
    "    - more sophisticated model architecture -> recurrent neural networks\n",
    "        - these require much more data typically (tens of thousands of samples)\n",
    "\n",
    "- can we get more data?\n",
    "    - synthetic promoter libraries\n",
    "    - are there promoter libraries that haven't already been used for this exact purpose but are out there?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Packages and Construct Bitome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import chain, product\n",
    "import os\n",
    "from pathlib import Path\n",
    "import shutil\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "from Bio.SeqUtils import GC\n",
    "from IPython.display import Image\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pingouin import multicomp\n",
    "from scipy.stats import f_oneway, gaussian_kde, mannwhitneyu, ttest_ind\n",
    "import seaborn as sns\n",
    "import shap\n",
    "from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor\n",
    "from sklearn.linear_model import ElasticNet, LogisticRegression\n",
    "from sklearn.metrics import confusion_matrix, r2_score\n",
    "from sklearn.model_selection import cross_validate, train_test_split, RandomizedSearchCV\n",
    "from sklearn.neural_network import MLPClassifier, MLPRegressor\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.svm import SVC, SVR\n",
    "from weblogo import *\n",
    "\n",
    "sys.path.append('../../bitome2')\n",
    "sys.path.append('../')\n",
    "\n",
    "from bitome.core import Bitome\n",
    "from bitome.util import create_motif, genome_point_to_point, score_motif_match\n",
    "\n",
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('display.max_columns', 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_FLORUM_DATA_PATH = Path('../data/bitome2/m_florum')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# terminus is assumed to be at the opposite end from the origin\n",
    "m_florum = Bitome(\n",
    "    Path(M_FLORUM_DATA_PATH, 'Mesoplasma_florum_L1.fasta'),\n",
    "    origin=(1, 1), terminus=(396612, 396612),\n",
    "    gene_table=Path(M_FLORUM_DATA_PATH, 'gene.csv'),\n",
    "    terminator_table=Path(M_FLORUM_DATA_PATH, 'terminator.csv'),\n",
    "    tss_table=Path(M_FLORUM_DATA_PATH, 'tss.csv'),\n",
    "    tu_table=Path(M_FLORUM_DATA_PATH, 'tu.csv')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>strand</th>\n",
       "      <th>KO</th>\n",
       "      <th>category</th>\n",
       "      <th>subcategory</th>\n",
       "      <th>subsubcategory</th>\n",
       "      <th>product_name</th>\n",
       "      <th>name</th>\n",
       "      <th>fpkm</th>\n",
       "      <th>primary_tu</th>\n",
       "      <th>log_fpkm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>peg.1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>1</td>\n",
       "      <td>K02313</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>DNA Maintenance</td>\n",
       "      <td>DNA replication and partition</td>\n",
       "      <td>dnaA, chromosomal replication initiator protein</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>192.793077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.266791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peg.2</th>\n",
       "      <td>1560.0</td>\n",
       "      <td>2681.0</td>\n",
       "      <td>1</td>\n",
       "      <td>K02338</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>DNA Maintenance</td>\n",
       "      <td>DNA replication and partition</td>\n",
       "      <td>dnaN, DNA polymerase III subunit beta [EC:2.7....</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>642.952373</td>\n",
       "      <td>TU_002</td>\n",
       "      <td>6.467625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peg.3</th>\n",
       "      <td>2708.0</td>\n",
       "      <td>3682.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>K03569</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cytoskeleton</td>\n",
       "      <td>Cytoskeleton proteins</td>\n",
       "      <td>mreB, rod shape-determining protein MreB and r...</td>\n",
       "      <td>mreB</td>\n",
       "      <td>2046.205311</td>\n",
       "      <td>TU_004</td>\n",
       "      <td>7.624231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peg.4</th>\n",
       "      <td>3751.0</td>\n",
       "      <td>4278.0</td>\n",
       "      <td>1</td>\n",
       "      <td>K05985</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>Folding, Sorting and Degradation</td>\n",
       "      <td>Nucleases</td>\n",
       "      <td>rnmV, ribonuclease M5 [EC:3.1.26.8]</td>\n",
       "      <td>rnmV</td>\n",
       "      <td>8.156475</td>\n",
       "      <td>TU_005</td>\n",
       "      <td>2.214461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peg.5</th>\n",
       "      <td>4280.0</td>\n",
       "      <td>5083.0</td>\n",
       "      <td>1</td>\n",
       "      <td>K02528</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Ribosome biogenesis</td>\n",
       "      <td>ksgA, 16S rRNA (adenine1518-N6/adenine1519-N6)...</td>\n",
       "      <td>ksgA</td>\n",
       "      <td>215.862583</td>\n",
       "      <td>TU_005</td>\n",
       "      <td>5.379264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             left   right  strand      KO                        category  \\\n",
       "locus_tag                                                                   \n",
       "peg.1         1.0  1332.0       1  K02313  Genetic Information Processing   \n",
       "peg.2      1560.0  2681.0       1  K02338  Genetic Information Processing   \n",
       "peg.3      2708.0  3682.0      -1  K03569              Cellular Processes   \n",
       "peg.4      3751.0  4278.0       1  K05985  Genetic Information Processing   \n",
       "peg.5      4280.0  5083.0       1  K02528  Genetic Information Processing   \n",
       "\n",
       "                                subcategory                 subsubcategory  \\\n",
       "locus_tag                                                                    \n",
       "peg.1                       DNA Maintenance  DNA replication and partition   \n",
       "peg.2                       DNA Maintenance  DNA replication and partition   \n",
       "peg.3                          Cytoskeleton          Cytoskeleton proteins   \n",
       "peg.4      Folding, Sorting and Degradation                      Nucleases   \n",
       "peg.5                           Translation            Ribosome biogenesis   \n",
       "\n",
       "                                                product_name  name  \\\n",
       "locus_tag                                                            \n",
       "peg.1        dnaA, chromosomal replication initiator protein  dnaA   \n",
       "peg.2      dnaN, DNA polymerase III subunit beta [EC:2.7....  dnaN   \n",
       "peg.3      mreB, rod shape-determining protein MreB and r...  mreB   \n",
       "peg.4                    rnmV, ribonuclease M5 [EC:3.1.26.8]  rnmV   \n",
       "peg.5      ksgA, 16S rRNA (adenine1518-N6/adenine1519-N6)...  ksgA   \n",
       "\n",
       "                  fpkm primary_tu  log_fpkm  \n",
       "locus_tag                                    \n",
       "peg.1       192.793077        NaN  5.266791  \n",
       "peg.2       642.952373     TU_002  6.467625  \n",
       "peg.3      2046.205311     TU_004  7.624231  \n",
       "peg.4         8.156475     TU_005  2.214461  \n",
       "peg.5       215.862583     TU_005  5.379264  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_florum.gene_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Investigate Expression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ultimately the TU is what is transcribed; but how consistent are the FPKM values for the genes within a single TU? Of course this only applies to TUs with multiple genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variance within TUs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Most genes in single TU: 21\n"
     ]
    }
   ],
   "source": [
    "tu_expression_dict = {}\n",
    "\n",
    "# to avoid -inf with log\n",
    "ETA = 1\n",
    "\n",
    "gene_counts = []\n",
    "for tu_row in m_florum.tu_table.itertuples():\n",
    "    genes = list(set(tu_row.gene_names.split(';')).intersection(set(list(m_florum.gene_table.index))))\n",
    "    gene_expressions = np.log(np.array(m_florum.gene_table.loc[genes, 'fpkm']) + ETA)\n",
    "    if len(gene_expressions) > 1:\n",
    "        tu_expression_dict[tu_row.Index] = gene_expressions\n",
    "        gene_counts.append(len(gene_expressions))\n",
    "        \n",
    "print(f'Most genes in single TU: {max(gene_counts)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0.87, 'P=3.89E-06')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD7CAYAAABkO19ZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAc30lEQVR4nO3dfXRU9b3v8fc34TGBQCsa1BSoYjUSCwLHU22qSfHa5RHtvW1d3hRLz7kpFFkrWttTsM09uq4QT7U9nmW1ELUpULWp1MdTLacPMHNbauttEFFhSpUiiK1F5fCQiJCE7/1jJnGCmSETZs+Enc9rrb3I7IfZ3ww7n/zy23v/trk7IiISXgX5LkBERIKloBcRCTkFvYhIyCnoRURCTkEvIhJyQ/JdQG/GjRvnkyZNyncZodDW1kZxcXG+yxDplY7P7NmwYcNb7n5yb8sGZNBPmjSJlpaWfJcRCtFolKqqqnyXIdIrHZ/ZY2Y7Ui1T142ISMgp6EVEQk5BLyIScgp6EZGQU9CLiIScgl5Ecq65uZmKigpmzZpFRUUFzc3N+S4p1Abk5ZUiEl7Nzc3U19fT1NREZ2cnhYWF1NbWAlBTU5Pn6sJJLXoRyamGhgaampqorq5myJAhVFdX09TURENDQ75LCy0FvYjkVCwWo7Kysse8yspKYrFYnioKPwW9iORUeXk569ev7zFv/fr1lJeX56mi8FPQi0hO1dfXU1tbSyQSoaOjg0gkQm1tLfX19fkuLbR0MlZEcqrrhGtdXR2xWIzy8nIaGhp0IjZACnoRybmamhpqamo0qFmOqOtGRCTkchL0ZlZgZs+YmZtZWS72KSIicblq0d8IvJOjfYmISJLAg97MPgIsBP456H2JiMj7BXoy1swKgB8QD/m9x1h3PjAfoLS0lGg0GmRpg0Zra6s+SxmwdHzmiLsHNhHvsnkk8fUkwIGyY203Y8YMl+yIRCL5LqHfLrnkEh82bJgXFxd7SUmJT5s2zR955JGM3+eyyy7z8ePH++jRo72srMxvvPFGf/fdd1Ou39ra6vPmzfPS0lIvKSnxCy64wNetW9djneXLl/tZZ53lxcXFPm3atB6f8/bt2x3woqIiLy4u7p4uvPDCtHWuWbPGzz33XB8xYoRPmTLFf/7zn/dY3t7e7jfffLNPmDDBi4qK/IwzzvCf/exnGX8eA8mJfHwONECLp8riVAuOdwImA7uA8a6gz5sT+Qfpkksu8SVLlrh7PORuv/12Lyws9K1bt2b0Phs3buwO9jfeeMMvvvhi/8Y3vpFy/RtvvNHPO+8837Vrl3d2dvp3v/tdHzVqlO/Zs8fd3VevXu3jxo3zjRs3ekdHhy9fvtyLiop8x44d7v5e0L/22mt9rnHbtm0+cuRIf+CBB/zQoUP+4IMPelFRkW/fvr17ndraWv/Yxz7msVjMjxw54q+//nqP5SeiE/n4HGjSBX2QffSVwMnAS2b2FvBcYv4LZrYwwP1KCA0ZMoSFCxfS2dnJiy++mNG206ZNY/jw4d2vCwoK2Lp1a8r1X3nlFWbPns3pp59OQUEB8+bNo7W1lW3btgHwk5/8hGuvvZZp06ZRWFjIggULOOWUU1i5cmW/vjeAVatWMWPGDK699lqGDRvGnDlzmD59OqtWrQJg69atNDU1sWLFCs455xzMjNNOO41Jkyb1e58yeAQZ9KuBM4FpiekfEvMvA34Y4H4lhA4fPsz3vvc9hg4dytSpU5k9ezZjx45NOf3oRz/qsf3ChQspLi5m/PjxbNq0ia997Wsp93X99dezdu1adu7cSUdHB42NjUyePJmKigqA5L9au7k7zz//fL+/v02bNjFjxowe86ZPn86mTZsAiEQilJSUsHr1ak4//XQmTJjAddddx4EDB/q9Txk8Agt6d3/H3Xd1TcAbiUVvuHtrUPuVcGloaGDs2LGUlZXx5JNP8uijjzJ58mSeeuop9u7dm3L6/Oc/3+N9li1bRmtrKy+++CILFiygrCz17RxTp05l0qRJTJw4kREjRrB06VJWrlzJiBEjAJg9ezYPPvggLS0ttLe3c88997Bz507279/f432mTJnS45fPwoWp/5A9cOAAY8aM6TFv7Nix3e/51ltvsX//fmKxGLFYjGeffZbnn3+er371qxl9njI45ezOWHd/1d0tEfoifVJfX8/evXvZvXs3zzzzDFdeeWW/38vMqKioYNq0aVxzzTUp1/vc5z7H4cOH+dvf/sa7777LihUruOKKK9i8eTMAc+fO5etf/zpz5sxh/PjxbNy4kUsvvZRx48b1eJ/Nmzf3+OWzbNkyABYsWMCoUaMYNWoUl19+OQCjR49m3759Pbbfu3cvJSUl3csBlixZQklJCaeeeiqLFy/mySef7PfnIYOHhkCQE9Lll1/eHZa9TQ899FDKbTs6Onj55ZdTLt+wYQPz58/nlFNOYciQIVx55ZWceeaZ/PKXvwTivzAWL17M1q1befvtt1m+fDmbN2/u85gtjY2NtLa20traypo1a4D4XxHPPfdcj/U2btzI1KlTgfh5hq59Jzv6tUhvFPRyQlqzZk13WPY2zZkzB4A//vGPPPHEE7S2tnLkyBE2btzIrbfe2t2S7s3HP/5xvv/977Nnzx6OHDnC008/zebNm5k+fToA+/btIxaL4e68+eabXHfddYwZM4YvfvGL/f5+5s6dS0tLC83NzbS3t9Pc3MyGDRu63/MTn/gE5513HrfccgttbW3s3r2bb3/723zmM5/p9z5l8FDQS6i5O3fccQdlZWWMGTOGq6++mquuuorGxsbudRYsWNAj+FesWMGwYcMoLy9n7NixLFq0iHvuuYeLL74YgP3793P11VczevRozj77bA4fPkwkEmHkyJE99n322Wf3+Csj3XmBM888k8cee4ylS5dSUlLC0qVLefzxx7uvqikoKOCnP/0pe/bsobS0lPPPP58ZM2bwne98J4ufloSVHX31wEAwc+ZMb2lpyXcZoaBhYGUg0/GZPWa2wd1n9rZMLXoRkZBT0IuIhJyCXkQk5BT0IiIhp6AXEQk5Bb2ISMgp6EVEQk5BLyIScgp6EZGQU9CLiIScgl5EJOQU9CIiIaegFxEJOQW9iEjIKehFREJOQS8iEnIKehGRkFPQh1RzczMVFRXMmjWLiooKmpub812SiOTJkHwXINnX3NxMfX09TU1NdHZ2UlhYSG1tLQA1NTV5rk5Eck0t+hBqaGigqamJ6upqhgwZQnV1NU1NTTQ0NOS7NBHJAwV9CMViMSorK3vMq6ysJBaL5akiEcknBX0IlZeXs379+h7z1q9fT3l5eZ4qEpF8UtCHUH19PbW1tUQiETo6OohEItTW1lJfX5/v0kQkD3QyNoS6TrjW1dURi8UoLy+noaFBJ2JFBikFfUjV1NRQU1NDNBqlqqoq3+WISB6p60ZEJOQU9CIiIaegFxEJOQW9iEjIKehFREJOQS8iEnIKehGRkFPQi4iEXKBBb2YNZrbdzPab2W4ze8TMJgS5T4nTePQi0iXoO2MfAO5w931mVgQsBX4MXBTwfgc1jUcvA11dXR33338/hw4dYvjw4cybN4+7774732WFVqAtenf/o7vvS7w04AhwdpD7FI1HLwNbXV0djY2N3HbbbaxZs4bbbruNxsZG6urq8l1aaAU+1o2ZfR5YDpQAHcBXU6w3H5gPUFpaSjQaDbq00IrFYnR2dhKNRmltbSUajdLZ2UksFtPnKnl37733Mm/ePKZPn05rayvTp09n3rx53HvvvXz2s5/Nd3mhZO6emx2ZjQdqgd+6ezTdujNnzvSWlpac1BVGFRUV3H333VRXV3cPahaJRKirq+Oll17Kd3kyyJkZbW1tFBUVdR+f77zzDsXFxeQqj8LIzDa4+8zeluXsqht3fwO4H3jKzD6Yq/0ORhqPXgay4cOH09jY2GNeY2Mjw4cPz1NF4ZfrYYqHAMXAacCeHO970NB49DKQzZs3j8WLFwNw7rnncuedd7J48WIWLFiQ58rCK7CuGzMrABYCq919t5mVAXcD04Cz3L0j1bbquskejUcvA5Guusm+fnXdmNkH+zCNPca+/wF4yczagGeBd4BL04W8iITfRRddxOTJkykoKGDy5MlcdJGuuA5Suq6bvyQmS7NOIdDrDVDufoR40IuIdNN9HrmX7mRszN3PcPcPp5qAt3NVqIiEg+7zyL10QX9hH7bvyzoiIt1isRiVlZU95lVWVhKLxfJUUfilC/qidP3zAO7+bo7qFJGQKC8vZ/369T3mrV+/nvLy8jxVFH7p+ujfAnYRv5sVevbVO3BGUEWJSHh13efR1UffdZ+Hum6Cky7ovwtUA78FmoH1rtvWROQ46T6P3EsZ9O7+FTMzoAr4AnC3mf0CWO7u23NUn4iEUE1NDTU1NbrPI0fSDoHgcRFgEdAI/BNwaS4KExGR7EjZojezYuDTwDXAycBjwAx335mj2kREJAvS9dHvBl4m/qCQl4mfgJ1pZjMB3P2x4MsTEZHjlS7of0I83M/m/Q8LceItfBERGeDSBf0NSU+H6qGrVS8iIgNfupOxvzKzDxw908z+G/B4cCWJiEg2pQv6+4CImZ3cNSPxWMD7gCuCLkxEwqu5uZmKigpmzZpFRUUFzc3N+S4p1NJdR3+/mb0LrDOzy4hffbMAqHb3V3NUn/RTc3MzDQ0N3Tek1NfX64YUGRA0emXupX3ClLs/kAj7jcBOoNLd38pJZdJv+kGSgSx59MquG6aampqoq6vT8RmQdA8eedHMXgBuBoqAk4i37rvmywClYWBlINPolbmXrkU/O2dVSFbpB0kGsq7RK6urq7vnafTKYKVs0bv7DuB84GrgHHffkTzlrELJmIaBlYGsa/TKSCRCR0dH9+iV9fX1+S4ttNINgbAMmAI8AywxswvcfUnOKpN+0zCwMpBp9Mrcs1QjD5vZS8BUd+80syLgN+4+IxdFzZw501taWnKxq9DSVTdyItDoldljZhvcvdebWdP10R92904Ad38nMWSxnCA0DKyIdEkX9OckXV1jwJmJ10Z8BOOPBl6diIgct3RBrzN3IiIhkC7oT3X33+esEhERCUS6sW6WdX1hZr/LQS2SRRpLRAYKM0s5VVdXp1wm2ZMu6JM/6RFBFyLZ09zczA033EBbWxsAbW1t3HDDDQp7yQt3TzlNXPxUymWSPem6bgoSwxQXJH3dHf7uvifo4qR/Fi1axJAhQ/jBD37QPdbNnDlzWLRokS6xFBmE0rXoxwAbgBagBHgu8bprngxQu3btYtWqVT3Gulm1ahW7du3Kd2kikgfphimelMM6REQkIOlGrxx/rI37so7kXllZGXPnzu0xlsjcuXMpKyvLd2kikgfp+uh/Bkw/xvZ9WUdy7I477uDLX/4yn/rUp2hvb2fo0KGMGDGCe++9N9+liUgepAv6qWa2n8SdsL0sN2B/IFXJcRsxYgQnnXQSO3bs4PTTT+++AkdEBp90wxQXunuJu49O/Hv0NNrdT89lsdI3DQ0NPPzww2zfvp1169axfft2Hn74YY1eKTJIpX2UIICZ9dY1sw/Y4e4d2S9JjpcePCIiydJdXtllGfB74D7g/sTXPwG2Jh4aLgOMHjwiIsmO2aIH/gLUuvtmADM7F7gVWAQ8BvwiuPKkP+rr67nmmmsoLi5mx44dTJw4kba2Nu666658lyYiedCXFv1HukIewN23EH+04J+DK0uyRWOGiEhfgn6zmS03s0sS0zJgi5kNB9pTbWRmt5vZZjPbb2Z/MbP7zeyDWatcUko+Gbt27VqdjBUZ5PoS9P8IvAJ8JTH9OTGvHahOsQ1AJ3AtcBIwFSgDVvazTsmATsaKSLJj9tG7+0Ezu5t4X7wDW929qyXfmma7bya9fNPM7gJWH0+x0jddJ2Orq9/7PayTsSKDV18ur6wCVgGvEr9J6kNm9kV3/3WG+5oFbMpwG+mH5JOxO3fuZMKECToZKzKI9eWqm38DLnP3rQBm9hGgGZjR152Y2WeBBcAladaZD8wHKC0tJRqN9vXt5Shbtmyhvb2dgwcP4u4cPHiQ9vZ2tmzZos9VBhwdkzmQ7qEAicH/X+jLvDTbXw3sAar7us2MGTNc+m/KlCm+bt06d3ePRCLu7r5u3TqfMmVKHqsSeb+Ji5/KdwmhAbR4ikztS4u+xcy+DzyYeD2HPo5Hb2b/RPwvgivd/bcZ/QaSftPJWBFJ1perbq4DtgDXJ6YtiXlpmdn1wHeATynkc0t3xopIsr5cdXMIuDMxZeIuoAOIJN+04+6jMnwfyVB9fT21tbU0NTXR2dlJJBKhtrZW19GLDFIpg97MXqT34YkBcPePpntjd9ctmXnS9VzYuro6YrEY5eXlNDQ06HmxIoNUuhb97JxVIVlXU1NDTU0N0WiUqqqqfJcjInmU7pmxO3JZiIiIBKMvJ2NFROQEpqAXEQm5lEFvZmsT/96eu3JERCTb0p2MPdXMLgKuMrMfEx/nppu7PxdoZZKR/o47H7+hTkTCLF3Q3wz8C/HhhY++ht6BTwZVlGQuVWBPuulpXv3WFTmuRkQGknRX3TwCPGJm/+LuS3JYk4iIZFFf7oxdYmZXARcnZkXd/algyxIRkWw55lU3ZvavwA3Ex7jZAtxgZrcFXZiIiGRHX0avvAKY5u5HAMxsFbAR+GbarUREZEDo63X0Y5O+HhNAHSIiEpC+tOj/FdhoZhHil1heDNwUaFUiIpI1fTkZ22xmUeDvErMWu/sbgVYlIiJZ05cWPe7+V+A/Aq5FREQCoLFuRERCTkEvIhJyx+y6MbPzgHMSL2Pu/lKwJYmISDale5TgGOBJ4EPAC8SvuDnPzHYCn3b3/bkpUUREjke6rpslQAtwlrv/D3f/78BZwB8APWVaROQEka7r5lLgo113xAK4+xEz+ybwYuCViYhIVqRr0R92946jZybmHQquJBERyaZ0LfoRZnY+Rz1wJPF6eHAliYhINqUL+jd4/wNHkpeJiMgJIN2DR6pyWIeIiAQk3cPB/87Mxie9nmtmT5rZd83sg7kpT0REjle6k7H3AocBzOxi4FvAD4F9wH3BlyYiItmQro++0N33JL6+BrjP3R8FHjWz5wOvTEROKFP/zy/Yd7A94+0m3fR0n9cdM3Iom265LON9DHZpg97MhiQup5wFzO/jdiIyCO072M6r37oio22i0ShVVVV9Xj+TXwrynnSB3Qz8XzN7CzgI/AbAzCYT774REZETQLqrbhrMbC1wKvALd/fEogKgLhfFiYjI8UvbBePuv+9l3p+CK0dERLJN49GLiIScgl5EJOQU9CIiIaegFxEJOQW9iEjIKehFREIu0KA3s/9pZr8xs/1m9r6HmIiISPCCHsrgv4BlwEg0EJqISF4EGvTu/nMAM6sKcj8iIpLagBmczMzmkxg4rbS0lGg0mt+CQkSfpeRKpsdaa2trxtvoeM7cgAl6d7+PRPfOzJkzPZMR7SSN/3w6o9EBRfqtH8dapqNX6njuH111IyIScgOmRS9905+HO2Q6hrce7iASLoEGvZkVAkOBYYnXIxKLDiUNeywZyPThDhn/aYwe7iD9M7r8Js5bdVPmG67KZB8AmT3cRIJv0X8BWJH0+mDi3w8Drwa8bxHJoQOxb+kJUwNUoH307r7S3a2X6dUg9ysiIu/RyVgRkZBT0IuIhJyCXkQk5BT0IiIhp6AXEQk5Bb2ISMgp6EVEQk5BLyIScgp6EZGQU9CLiIScgl5EJOQU9CIiIaegFxEJOT145ATTrzG/MxjvO74P0JjfIuGhoD/BZDrmtx48IrnUr2PnP/u+zZiRQzN/f1HQi0h2ZPrQEYj/YujPdpIZ9dGLiIScgl5EJOQU9CIiIaegFxEJOQW9iEjIKehFREJOQS8iEnIKehGRkNMNUyegjO8+zODOQ9DdhyJho6A/wWR6F6HuPBQRdd2IiIScgl5EJOQU9CIiIaegFxEJOQW9iEjIKehFREJOQS8iEnIKehGRkFPQi4iEnIJeRCTkFPQiIiEXaNCbWaGZfdvM3jSzA2b2qJmNC3KfIiLSU9At+puATwN/D5Ql5j0Q8D5FRCRJ0KNXzgdudfc/A5jZIuAVM5vo7jsC3vegYmapl92eejt3D6AaERlIAgt6MxsLTAA2dM1z921mth+YCuw4av35xH8xUFpaSjQaDaq0UIpEIr3Ob21tZdSoUSm30+csQauurk67PFVDJNUxLZkLskU/OvHvvqPm7wVKjl7Z3e8D7gOYOXOmV1VVBVja4BGNRtFnKfmU7q9GHZ+5EWQf/YHEv2OOmj8W2B/gfkVEJElgQe/ue4GdwPSueWZ2BvHW/AtB7VdERHoK+qqb+4DFZvZhMysBbgd+7u6vBrxfERFJCPqqm28BHwD+AAwHfglcG/A+RUQkSaBB7+6dwD8nJhERyQMNgSAiEnIKehGRkFPQi4iEnA3EW+DN7E2OunNW+m0c8Fa+ixBJQcdn9kx095N7WzAgg16yx8xa3H1mvusQ6Y2Oz9xQ142ISMgp6EVEQk5BH3735bsAkTR0fOaA+uhFREJOLXoRkZBT0IuIhJyCfoAyswlm1mpmp6VZp8rMOo7xPpvN7JrsVyiSPWZWaWbqRw6Igj4HzOx6M9t21Lw6M3Mzuzxp3kgze9fMrnL3ne4+yt3/klj2j2b2Sqb7dvcp7v5wP+tuTZraE1P3vMQ6r5rZ+0YkTTVfTgxmFjWzQ4n/631m9ryZXZ3vuqR/FPS5sRY4w8wmJs2bBWwGPpk07+NAIRDNXWmpJX7RjHL3UcAq4KGj5km4LUn8P58ErAR+ZGaT81uS9IeCPgfcfTPwV+LhjpkVApcAt3TNS5gF/MHd95vZpESLv8zMLgQaif+y6GpRV3VtZGbXmNm2RMtrtZmNTlrW3bLu6upJt77I0dy9A7if+LDm0wDMbIWZvWZmB8xsi5l9vmv9vhxnZnZW4q+GA2a2Cehxd6yZFZnZXYl9vGVmT5jZhKTlUTO708weT7zHNjObZWaXmtlLZrY/sUzHNgr6XFrHe6E+A3gD+A/gTDM7KTF/FvCrozd0998BC4A/J7Woo4nFhcBlwFTgI8D5wPVp6sh0fRnkzGwYcF3i5Z8S/64nHvpjgVuBlWZ2btJmKY8zMxsCPEX8L9pTgM8RP76T/TvwscQ0kfh4OD9NNJK6fIH4w43GAg8DDwDzgYuBScDZ6NgGFPS59Cve66aZBaxz93bgGaDazMYQf77u+4K+D25y91Z3/xvwBEe1jrKwvgxO9Wa2FzgILAW+5O4vALh7k7u/7e6d7v5j4s+Brjpq+1TH2d8TD+Kvu/tBd38Z+LeujcysAPgi8L/d/XV3bwO+ApQDFyS9/2p3fzbxgKMHgVOBb7v7HnffQ/yXiY5tFPS5tBYYn2j1fJJ4Cx8gknhdBbwL/C7D9+109zeTXrcB6f5czXT9Y2kHhvYyf2himZy4Gtx9LPERJn8GVEM8iM3sVjPbmuiW2Uu85Z48cmK646wM2O3u7yQt35709cnEHz3aPc/dW4HdwIeS1vtr0tfvpJinrhuCf2asJLj7a2b2J+AK4EKg65LHdcBDQAfw60QrvzdHgq+yX14FepygM7NRwHjgz/koSLLL3f/LzL4EbDOzTwOjgC8R75rZ4u5HzKwFsD6+5evAKWZWlBT2k5KWvwkcSsx7BbqPqVOA147z2xmU1KLPrbXAV4GXE39aAmwkfgBfTfpumzeI/3CUBFtixlYC883sE2ZWaGYfAO4CXiT+vUkIJI7XO4HbiPeJdxAP5AIz+1/EW/R99Xviz5u4PXFJ8ZnEfy669nUE+CGwxMxOM7Mi4l07fwT+Xxa+nUFHQZ9bvyLe0u3qtul6gPqvE/PTBX0E+CWw3cz2mtklQRbaV+7+EPBN4HvAHuAlYCRwZeJqDQmPu4j3gzvwLPHW9uvAucBv+vomiePiKuCjxLtjHuP9g5vdCLQAfwB2JvZ7VeLnRTKkQc1EREJOLXoRkZBT0IuIhJyCXkQk5BT0IiIhp6AXEQk5Bb2ISMgp6EVEQk5BLyIScv8fVw77Qo+COwYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sd_within_tu = [np.std(exps) for exps in tu_expression_dict.values()]\n",
    "\n",
    "# bootstrap by random sample groups of genes and getting their standard deviations; have to random sample \n",
    "# the number of genes too; sample from the same distribution as the empirical one\n",
    "\n",
    "tu_size_kde = gaussian_kde(gene_counts)\n",
    "\n",
    "sds_random = []\n",
    "for i in range(len(sd_within_tu)):\n",
    "    rand_n_genes = int(max(2, np.round(tu_size_kde.resample(1).flatten()[0])))\n",
    "    gene_idxes = np.random.choice(len(sd_within_tu), rand_n_genes)\n",
    "    exps = np.log(np.array(m_florum.gene_table.iloc[gene_idxes, :]['fpkm']) + ETA)\n",
    "    sds_random.append(np.std(exps))\n",
    "    \n",
    "t_stat, p_val = ttest_ind(sd_within_tu, sds_random, equal_var=False)\n",
    "    \n",
    "tu_exp_sd_df = pd.DataFrame(data={'Within TU': sd_within_tu, 'Random': sds_random})\n",
    "ax = tu_exp_sd_df.boxplot()\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_ylabel('SD of log[FPKM]')\n",
    "ax.text(0.5, 0.87, f'P={p_val:.2E}', fontsize=13, transform=ax.transAxes, ha='center')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, so it does seem visually as though variance in detected expression for genes within the same TU is markedly lower than that of randomly selected genes. However, there is still some decent variation there.\n",
    "\n",
    "So we should try:\n",
    "- predicting the average expression of the TU from the constituent genes based on TU's promoter\n",
    "- predicting the expression of the first gene in the TU from the promoter\n",
    "- using all gene expressions individually but with same promoter if in same TU (this seems like it'll be toughest for the model if the genes in the TU have different expressions; it'll be a duplicate sample with a different target value?)\n",
    "- averaging the promoters associated with the same gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Expression Distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'M. florum Gene Expression Distribution')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAdAAAAFUCAYAAACUfG3OAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAiCklEQVR4nO3dfbxcVX3v8c8vJEBCAOVB1FjIpREJGquVVtqqRXlKQu/1AWqttkGxRkMlSKv2gRSCHkQUq4iXViiURFB8qApqiIUrqC3aNrECQjAeuUGIYiEgkBAChNU/1p7jPjszc85ZOSdzTvp5v17zSmY/rd/ae2a+s/bsMxMpJSRJ0shM6nUBkiRNRAaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAN2JRcTSiPh5RKSIeHN1/we9rksTS/X4ObHXdbRExJFVTfuNwbYvj4iv1u7fGBGfGO12qm2PWT+0Yxig40D1pE0RcWmbeedV877abt0u23wBcBbwDuBZwGdHp9reiIjXRMT1EfFARGyOiLURsSwifqPXtcHAC21qc7uq17WNgmcBXxnLBiJiZmO/bYyIH0bEP0TECxuL31TVtGGY2x7JG4DTgD8aQenDEhHrIuLdjckj6ofGHwN0/LgbeH1E7NGaEBGTgQXATwq2N6v698sppXtTSptHoUYiYspobGeEbZ4DfAG4FXgNcCjwB8DtwId3dD1d/CP5BbF+e/tYNhgRu47l9gGqx8+WsW6nMpe83+YApwPPAFZHxBtq9Txe1TRq3wITEZMjIlJKD6WUfjFa2+1mLPqhHSyl5K3HN+By4KvA94C31Ka/Gvj/wDLgqyPY3lIg1W+16T+oLTcJ+BtyeG8hB9Sra/NnVuv/IfANYDPwTuDNwMZGm0dWy+5X3X8zsBGYB9wBPApcA+wNnAj8CHgI+BQwtUtfXlptd3GH+dG4/7+B1cBj1b47B9i1Nn8dsAT4JPAwcA/wnsY29gYuBv4LeAT4JnD4EPv8RuATXeb/cbUPDq1NO7fa90+v1bYUuKLad/cC725sJwF/CnwR2AScP8x+vw64pTqGD1R9OqCa9yvA1dX0R6vj9YZGmyfW7s8Brq9t63Jg7zaP59OA9cCD5DcX07rsn9ZjbZv9DHwa+AXwtA6Ptb2rx9F/Vf2/E3hXbZ/Wnwvr6s8F8uP0x8BWYHqr9sZx/XvggqofD5LftE1qPKaax2ng8VD9v93zcVA/asfpVvLz8W7gDGqPcYbx+PW2426OQMeXS4GTa/dPJr/wjPQd6vnA26r/t0ZC7ZwGvAf4C/KL4peAL0bEixrLnQtcBBwGfHkEdewG/DnwJuAo4HDgn4CTgBPIo8nfA07pso03ksPkonYzW69EABFxHHAl8Ang+eT9dyLwgcZqp5NfpH4dOA/4UET8VrWNAL4GzKhqezHwLeAbEdFpPw4ppfQp8r77dETsGhFHAu8GFqSUHqwt+mfAmqq2s4APRMTrGps7C1hBPmb/d6h+R8QzgavIb8RmA68gB07LRcA04JXV+u8iB9Y2qjMkXycfk98EXgv8NnBZY9GXAy8AjiafLXgt+fFW4nxySB7dYX4feV/8HvA8cv/XV/Nap/jfRn4e1E/5/y/y4+v3gV8jh287byK/2fwt8hmFheR9NFyvIwfd++jyfIyIlwCfJ785mgP8JfBX5DetdR0fv9rBep3g3ga9Y386+V39c4Fnkt+FHkjjXfEwt3kiVb7Upi1l8Ah0PXBmY5kbgSuq/88kh/efN5Z5M8MbgSbgebVlzie/09+v2fcu/bgWuLkx7RTyC3jrdmA1/VvA3zSWfU21TFT31wGfaSzzI2BJ9f9XVctPbSzzfeC9Xeq8EXi8UddG4JTaMntX7V9MHl2c19jGOuC6xrR/AP6ldj8BFzaW6dpv8gttAg7qUPstwFld+jYwAiUH0UPAnm2O/azaMb0b2KW2zCXA9V3aaD3W2o1Ad6/mvbfDY+0a4LLh1N94LjxBNQrv9HisjutaBo8ClwD3NI5bxxFol2Wa/bgS+EabOpttdXz8etuxN0eg40jKI5Evkd9BnwTcmFIq+fxzSBGxF/Bs4F8bs/6FPNKsW1XYzJaU0g9r938O3JtSur8x7Rkj3O6VwIvIF3vswS8/y38JcEZ1AcrGiNhIPv23B/kNScstje39tFbDS8ijsfsa23kB8KtD1PXZqq767crWzJTSQ+Tj+jbgfvILcdN32twf6ngM1e+byadcfxAR/xQRiyJi/9r6FwBLIuI7EdFXjYQ6mQ3cklJ6pDbtJuCpRp23p5S21u7X9/FIRfVvpzMxfwf8QUTcHBHnR8TvDnO796SUfj6M5b6bqqSqfAeYUT2HRtNs2j8fm211e/xqB5rc6wK0jcvIp9o2Amf2qIbmC9Wmxv2n+OWLWku7i4uebLPdJ9pM6/ZGbi3w8oiYklJ6AgaC6KGIeFpj2UnA2eTTYE331f7frYZJ5FB/eZttPNylToCHUkr9QyzzCvIo/BnAXpRdgdk8Hl37nVLaGhHHAkcAxwJvBc6NiN9NKd2cUro0Ir4OzCefJr0pIs5NKS0dYV31x81Ij3M3rWC+s22jKV0bEQeRP28/CvhaRHw+pfSWIbbb3I+lhvt82B5jtW+1Hdzp48//I58K3I+Rfd44Iimlh8nvXH+nMetl5Ktbu7kPmNZ4V/yi0atukM+QR1KnDmPZ75Ev0ulvc2uGebdtHAA81WYb/1XYBwAi4qXki7ZeS973l7RZ7Ig299cMo+au/U7Zd1JKZ5M/B/wp+bNJqvn3pJQuTim9nvzGbWGHttYAcyJiz9q03ya/lgxVZ6l3k08bX99pgZTS/SmlT6WU3kx+g3BSROxWzX4C2GU72n9p9dl4yxHAT6vnEOTnw8DnmhGxO/lK8brHh1HDGto/H+9pjPg1TjgCHWdSSqn6u7dIHf50ICJeS76w56iU0vp2ywzTh4H3RcSPyFdw/hF55PXrQ6z3b+R37+dGxEfJF2B0uxCoWErpuxHxIeDD1SjjC+Q/6zmAX77It04Vvg/4akTcBXyOPAJ+AfCbKaX3DrPJ68mn0a6OiPeSr0h9JvnPK65PKX27y7rTqgt26h5PKT0QEdPJV9f+fUrpKxFxB/CfEfHWlFL973+PiIi/qvp5JPnPmN40RM1d+x0RR5BHll8nj65fTL7y9naAiLiA/FnzWvKoeC6d30RdSR7tLo+IM8mf238S+OIwRt/DsW+1D6eSQ2gReWT5x9WZh21ExPvIbyJuI7+mvQ64s/b8WQccFRHfJH+s8GC77XTxbOBjEXER+eKe95AvXGr5BnByRFxDDtMz2Pa1dR35TMoVVQ33s62PAP8REUvJp+B/g3wR3l+PsF7tII5Ax6GU0iO1d7ft7E2+2nB7TxN9nByiHyJf0v9a4ISU0s1D1PcA+UX9GPLVgAvJI6sxkVL6C+D15Beva4B+8pWK04FXpJTurpb7OnA8+WrSf69uf8kI/o62+qxrPvlF8RLgh+RQeh551NbNW4CfNW7XVPMuII9C3lu18yPyVakXRMSs2jb+Fngh8J/kF+kzU0pfGKLmofr9EHlk81XyBScfAd6fUrqimj8JuJAcmteRQ/akDm09ChxHDtp/J//5y3cYfPX49lhJ3m+3kffZfeQLi7p9IcUW8p/t3Ex+87Mn+c96Wv6cvG/uJu/XkbqSPHr8N/Jj4lLgo7X555IfL1cD/0z+3LLZzpnkNy0/ZvDHCQNSSt8jXxF8Avn5+MHqNibfhKTt17oyUVKPRcQ68pWb5/e6FklDcwQqSVIBA1SSpAKewpUkqYAjUEmSChigkiQVGNHfge63335p5syZY1SKJEnjy+rVq+9PKe3fbt6IAnTmzJmsWlX6taiSJE0s1ReUtOUpXEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUm97qAugsvvJD+/v6uy6xfvx6AGTNmjFkds2bN4tRTTx2z7UuSJr5xFaD9/f18/wdr2Dptn47L7PLoQwDcu2VsSt/l0QfGZLuSpJ3LuApQgK3T9mHzofM7zp96xwqArstsj9b2JUnqxs9AJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSgZ4F6IUXXsiFF17Yq+Z3Wu5XSdoxJveq4f7+/l41vVNzv0rSjuEpXEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigO6FHH32UefPmMXfuXPr7++nv72fevHmcfPLJLFq0iP7+fhYvXkx/fz+nnHIKixYtYsOGDWzYsIGTTz6ZV77yldxwww0sXryYDRs2sGrVKl71qldxzTXXcPzxx3PDDTcwb948FixYwLx58+jv7x9ou7+/n+OPP57Vq1ezePFiVq1axbx58zjuuONYuHAhGzZsGFh2w4YN29Rx9dVXc+SRR3LssccObKNVb2vdDRs2sGjRIk455ZRB01rbaC7f1Gq3XS3tttfaN4sWLWLhwoUD0zptq5t6n7v1aTg1jkedahzp9PFoItVaN1HrLrGj+2qA7oR+8pOfsHnzZh577DH6+vro6+tj8+bN3HnnnaxZs4a+vj5uvfVW+vr6uP3221mzZg3Lly9n2bJl3HnnnaSUOOecc7j11ltZvnw5S5cu5amnnuKjH/0omzZt4pxzzmHz5s0D7fT19Q203dfXx6ZNmzjrrLO49dZbWbp0KZs3b2bLli2sXbuW5cuXDyy7bNmyber42Mc+BsDjjz8+sI1Wva11ly1bxpo1a7j99tsHTWtto7l8U6vddrW0215r36xZs4a1a9cOTOu0rW7qfe7Wp+HUOB51qnGk08ejiVRr3UStu8SO7qsBupN59NFH2bJly8D9devWsW7dukHLrFu3jpTSoOkrVqxgxYoVA/effPJJUkp87WtfY+PGjQCklAbmNbfXGum2trlx40ZSSgPr1ttpjbBWrly5TR2tNurbaNW7cuVK+vv7ufbaaweWufbaa+nv72flypXb9G/lypVtRzytdlvzm9Oa21uxYsWgNpvtdmqrqdnnTn1qtdetxvE4muhU40inj0cTqda6iVp3iV70dfKYt9DB+vXr2bx5M6eddtrAtP7+fiY9nrqsNfYmPfYw/f2PDKprIqmfTh2JJ554ou30Zlh2Uh+FDtXO8uXLSSnx1FNPDbs+gK1bt9LX1zeopieeeIK+vr629W/dupXly5dz+umnD0xbtmzZQLut+fVaWm3Ut9du2612m9uqt9VUb3uoPkVE1xqHaqsX2u3b008/fcTTx6OJVGvdRK27RC/6OuQINCIWRsSqiFh13333jWkx2n71EdyO1G6k28l1113H9ddfP+xwbnnyyScHRm4t9dFcu+Wvu+66QdPq7bbmN6d12l5dq93mtrpp1+dO7dVH++1qHKqtXuhU40inj0cTqda6iVp3iV70dcgRaErpYuBigMMPP3zUXp1nzJgBwAUXXDAw7bTTTmP1nT8frSaKPLX7Xsw6+IBBdU0kxx133KBTuDvKzJkzAYYVoscccwwpJVasWDGiEJ08eTLPec5zuOuuuwYCJiI46KCDBk2rL3/MMccMmnb00UcPtNuaX6+lXRvttNq95557Bm2rm3rb3frU2n5KqWONQ7XVC+32bcn08Wgi1Vo3Uesu0Yu++hnoTubAAw8sWm/KlClMnrzt+6l209pZsmQJS5YsGVY7CxYs4KSTTmLSpJE9/HbZZReWLFkyqKYpU6awZMkSpkyZ0nb5BQsWDJpWb7c1vzmtub0pU6Zss/1Wu81tddOuz5361Lrfqcah2uqFTjWOdPp4NJFqrZuodZfoRV8N0J3MtGnT2G233Qbuz5w5c2B0WJ8WEYOmz58/n/nz5w/cnzx5MhHB8ccfz/Tp0wEGPpdrhurMmTOZNWsWs2bNGtjm9OnTiYiBdevt7Lvvvuy7777MnTt3mzpabdS30ap37ty5zJo1i3nz5g0sM2/ePGbNmsXcuXO36d/cuXPZd999B7Vfb7c1vzmtub358+cParPZbqe2mpp97tSnVnvdahyqrV7oVONIp49HE6nWuolad4le9NUA3QkdeOCBTJ06ld13331gZDh16lQOPvhgZs+ezZIlS5gzZw5LlizhsMMOY/bs2QOjnIMPPpiI4IwzzmDOnDksWLCApUuXMmnSJE4//XT22GMPzjjjDKZOnTrQTn3kuWTJEvbYYw/OPvts5syZw9KlS5k6dSq77bYbhxxyyKB3hSeddNI2dbzrXe8CYNdddx3YRqve+shl9uzZHHbYYYOmtbbRXL6p1W67Wtptr7VvZs+ezSGHHDIwrdO2uqn3uVufhlPjeNSpxpFOH48mUq11E7XuEju6rzGSi04OP/zwtGrVqlFpuHWVa7vPQDcfOr/Taky9I/+pRbdltsfUO1bwkgn8GWi7/SpJKhMRq1NKh7eb5whUkqQCBqgkSQUMUEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSgcm9anjWrFm9anqn5n6VpB2jZwF66qmn9qrpnZr7VZJ2DE/hSpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkApN7XUDTLo8+wNQ7VnSZvwGg6zLb2z4cMCbbliTtPMZVgM6aNWvIZdavfxKAGTPGKuQOGFYdkqT/2cZVgJ566qm9LkGSpGHxM1BJkgoYoJIkFTBAJUkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigkiQVMEAlSSpggEqSVMAAlSSpgAEqSVIBA1SSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAgaoJEkFDFBJkgpESmn4C0fcB9w1iu3vB9w/itvrFfsxvtiP8cV+jC/2Y2QOSint327GiAJ0tEXEqpTS4T0rYJTYj/HFfowv9mN8sR+jx1O4kiQVMEAlSSrQ6wC9uMftjxb7Mb7Yj/HFfowv9mOU9PQzUEmSJqpej0AlSZqQDFBJkgr0JEAjYp+I+FJEbIqIuyLijb2ooykidouIS6uaHomI70fEvNr8oyLijoh4NCJuiIiDGuteFhEPR8S9EfFnjW13XHcM+/PciHgsIq6oTXtj1b9NEfHliNinNq/rcem27hj34w0RsaZq98cR8fJq+oQ5HhExMyJWRMSDVT2fiIjJ1bwXRcTqqpbVEfGi2noREedFxIbqdl5ERG1+x3VHoeZ3RsSqiNgSEZc35o3Jvh9q3dHsR0QcERHXRcQDEXFfRHw+Ip5Vm1+874dadzT70VjmzIhIEXF0bdqEOB7VvGkRcVFE3B8RD0XEt2rzxtXxACCltMNvwGeAzwLTgZcBDwHP70Utjbr2AJYCM8lvLn4PeKS6v19V5+8DuwMfBr5bW/dc4NvA04HZwL3A3Gpe13XHsD//XNV0RXX/+VV/XlHt+08DVw3nuAy17hj24Rjyl3ccUR2TGdVtQh0PYAVwedXeM4FbgcXArlX/Tgd2q6bdBexarfd24IfAc6p+3w68o5rXdd1RqPl1wGuAvwMur00fs33fbd0x6Me8qo69gGnAZcDK2vzifd9t3dHuR23+r1aPq58CR0+041HNuwK4Ctgf2AV4yXg9HimlHR+g5JB6HDikNu1TwAd3dC3DrPcW4ARgIXBTox+bgUOr+z8Fjq3Nfz9VwAy17hjV/Qbgc+Q3BK0A/QDw6doyv1odiz2HOi7d1h3j/X8T8NY20yfa8VgDzK/d/zDwSeBYYD3VBX3VvJ/wyxe4m4CFtXlvpXqBG2rdUay9j8HBM2b7vtu6o92PNvN/HXik8dgr2vfd1h2rfgArgfnAOgYH6IQ4HsChwMPAXh2WH3fHoxencA8Bnkwpra1Nu5k8whlXIuIAcr23keu7uTUvpbQJ+DHw/Ih4OvCs+nwG96njumNU917A+4Dm6ZZmHT+mCk2GPi7d1h0TEbELcDiwf0T0R8Q9kU99Tm1Tz7g9HpWPAW+oTlHNII9+VlZt3pKqZ3Xllk61sm0/uq07VsZk3w9j3bH2CvJzvWV79n23dUddRPw+sCWltKIxfSIdj98kjxrPrk7h3hoRJ9Tmj7vj0YsAnU5+l1H3EHkUNG5ExBTgSmBZSukOct0PNRZr1T29dr85jyHWHQvvBy5NKd3TmD5UH7odlx3dB4ADgCnAicDLgRcBLwaWDFHPeDseAN8iP2EfBu4BVgFfHkYtzfkPAdOrz2960Y92NdXb3Z59P9S6YyYiXgicCbynNnl79n23dUdVROxJPkN0WpvZE+l4PAd4QdXGs4F3AssiYnaHWnt+PHoRoBvJnznU7UX+fG1ciIhJ5NOXj5MPInSve2PtfnPeUOuOquqD86OBj7aZPVQfutXYi+O2ufr3wpTSz1JK9wN/Sz5NNSGOBww8nlYCXySfItuP/JnSecOopTl/L2Bj9U67V8+lsdr3Q607JiJiFnAtcFpK6du1Wduz77utO9qWAp9KKa1rM28iHY/NwBNAX0rp8ZTSN4EbyKdn29Xa8+PRiwBdC0yOiOfWpv0ag0+d9Ez1juRS8ujnhJTSE9Ws28h1tpbbg/w54G0ppQeBn9XnM7hPHdcdgy4cSb7o6ScRcS/wbuCEiPhemzoOJn/gvpahj0u3dcdEtV/vAeoP8tb/J8rxANgHOBD4REppS0ppA/CP5DcCtwEvbLwTfmGnWtm2H93WHStjsu+Hse6oq644vR54f0rpU43Z27Pvu6072o4CFldXyd4L/ArwuYj4iwl2PG5pM63+3B9/x2M0Pgwu+PD4KvIVn3sAv8M4uQq3qu3vge8C0xvT96/qPIF8tdp5DL5a7YPAN8kji0PJD7y5w1l3lOufRr7Ks3U7H/hCVUPrFOLLq31/BYOvwu14XIZadwyPx/uA/wCeUe3bb5NPUU+I41Gr507gL4HJwNOAL5GvZG5dPXga+Q3JOxl89eA7yBcgzSCf1rqNba88bLvuKNQ8udo/55LPyOxeTRuzfd9t3THoxwzy533v7rBe8b7vtu4Y9GNfBj/n7yZfVTt9gh2PKUA/8DfV/d8hjyBbFzSNq+ORUupZgO5D/vxnE/lKqTf2oo42dR1EfsfzGHnI37q9qZp/NHAH+VTDjcDM2rq7kS+Dfxj4OfBnjW13XHeM+7SU6irc6v4bq32+Cbga2Ge4x6XbumNY/xTgIuAX5EvoPw7sPtGOB/nz2xuBB8m/Yfg54IBq3ouB1VUt3wNeXFsvgA8BD1S3DzH4SsOO647SYyc1bkvHct8Pte5o9gM4q/p//bm+cTT2/VDrjvbxaCy3jsFX4U6I41HNez7wHfJrzO3Aa8fr8Ugp+V24kiSV8Kv8JEkqYIBKklTAAJUkqYABKklSAQNUkqQCBqgkSQUMUEmSChigUhsRcWP1w8Svb0x/aTV9XY9KIyKOrGrY2LgdXM1fF/mH1DdG/gHvf42II2vrp4h4We3+CyNifUR8PCImRcTSapmLGu3uHvnHp1NEzNxB3ZXGLQNU6mwN8LbGtLdV03tta0ppeuN2Z23+n6SUppO/tmw18JWI2Lu5kYh4Jfmr2j6eUlqcUnqqmrWW6ufXaoufSP42KEkYoFI3XwReXBvZ7Un+ztB/rC8UEZMj4q8jYm1E/KIa8R1em39URPxbNRq8LyKuiohn1ObfGBEfiYh/iohHIuLHEfHq0ehASmkzcDH555xmNer+A+AaYHFK6bzGqneTvxO6PgJ/G3DJaNQl7QwMUKmzx8i/CfvW6v4fkkdrP2ssdzbwamAu+Yu9LwNWVj9IDLCF/OXW+wNzyKPCCxrbOAn4CLA38Any7yBOYztVv67xdvIXhtd/OecUcrCekLb9FZKWS6hG4BHxPPKXiV+9vTVJOwsDVOruEuAtETEZWEhjBFb9fNJi4D0ppTtTSltTSpeSQ/Z4gJTSv6SU/iOl9GRK6V7yF1kf1Wjnsymlm6pTqBeTg/S5dLZLNdpt3b7cmP/JiPgF+VdgDgOOTynVf8fxeOD75B/67uQrwMER8XxykC4n/0auJPJPxkjqIKX0g4i4i/wTS88g/zD2H9YW2Y98evQrEVH/ZYYpwHMAIuIlwAfIv0E4jfzLENMbTQ2MalNKm6qfNdyzS2lbU0pP6zL/7SmlK7rMfzNwJrAiIv5PSmljc4GU0pMRcTnwp+TPP1/WXEb6n8wRqDS0i8kBellKaWtj3v3kn146OqX0tNptj5TSB6tlriL/vNIhKaW9GBzAvXIf8EpgKnBdRDytw3KXkE8B355SGrMfT5cmIgNUGtpngGPZ9nNLUv49wAuA8yPiuQARMT0ijouIZ1eL7UX+DPKRiDiQ/MPaPZdS+gVwDPmz3m9ExP5tlrkTeAVw8o6tThr/DFBpCCmlx1JK16eUHuywyFnki2uujoiHgR8B7+CXz6+FwJ8Aj5Cv7P38GJc8bNWp23nkU8jfrIV+fZl/bfyJjCTwB7UlSSrhCFSSpAIGqCRJBQxQSZIKGKCSJBUwQCVJKmCASpJUwACVJKmAASpJUgEDVJKkAv8N0MRD9Si+AAsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 576x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "_, ax = plt.subplots(figsize=(8, 5))\n",
    "sns.boxplot(x='fpkm', data=m_florum.gene_table, ax=ax)\n",
    "ax.set_xlabel('Mean FPKM', fontsize=13)\n",
    "ax.tick_params(axis='x', labelsize=12)\n",
    "ax.set_title('M. florum Gene Expression Distribution', fontsize=14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_log = m_florum.gene_table.copy()\n",
    "gene_table_log['log_fpkm'] = gene_table_log['mean_fpkm'].apply(np.log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(8, 5))\n",
    "sns.boxplot(x='log_fpkm', data=gene_table_log, ax=ax)\n",
    "ax.set_xlabel('Log FPKM', fontsize=13)\n",
    "ax.tick_params(axis='x', labelsize=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not sure which is better, so let's prepare a target lookup with 4 possibilities:\n",
    "- raw FPKM\n",
    "- log2-transformed FPKM\n",
    "- standard normalized raw FPKM\n",
    "- log2-transformed, standard normalized FPKM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_df = pd.DataFrame(m_florum.gene_table['mean_fpkm'])\n",
    "target_df['mean_fpkm'] = target_df['mean_fpkm'] + ETA\n",
    "target_df['fpkm_log'] = target_df['mean_fpkm'].apply(np.log)\n",
    "target_df['fpkm_log_stand'] = StandardScaler().fit_transform(target_df['fpkm_log'].values.reshape(-1, 1))\n",
    "target_df['fpkm_stand'] = StandardScaler().fit_transform(target_df['mean_fpkm'].values.reshape(-1, 1))\n",
    "target_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validate these transformations with boxplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_df.drop(columns='mean_fpkm').boxplot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Log stand mean: {target_df['fpkm_log_stand'].mean()}\")\n",
    "print(f\"Log stand SD: {target_df['fpkm_log_stand'].std()}\")\n",
    "print(f\"Stand mean: {target_df['fpkm_stand'].mean()}\")\n",
    "print(f\"Stand SD: {target_df['fpkm_stand'].std()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TSS Alignment Investigation\n",
    "\n",
    "We know that many genes are transcribed from multiple TSSes:\n",
    "- what are the relative strengths of these TSSes? \n",
    "- should we average the one-hot sequence of all TSSes for a gene, or use the strongest?\n",
    "- should we align multiple TSS at the TSS itself, or at the -10 box location?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Signal Height vs RSPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(x='signal_height', y='rspm', data=m_florum.tss_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So these seem to be perfectly correlated metrics; let's use RSPM to assess TSS strength\n",
    "\n",
    "RSPM = number of read starts per million of mapped reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Relative TSS Strength for Multi-TSS Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tu_to_genes = dict(zip(m_florum.tu_table.index, m_florum.tu_table['gene_names']))\n",
    "tu_to_genes = {k: v.split(';') for k, v in tu_to_genes.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_rspm_by_gene = []\n",
    "\n",
    "for gene_row in m_florum.gene_table.itertuples():\n",
    "    tus_for_gene = [tu for tu, genes in tu_to_genes.items() if gene_row.Index in genes]\n",
    "    if len(tus_for_gene) > 1:\n",
    "        gene_tu_df = m_florum.tu_table.loc[tus_for_gene, :]\n",
    "        gene_tsses = gene_tu_df['tss']\n",
    "        gene_tss_df = m_florum.tss_table[m_florum.tss_table['tss'].isin(gene_tsses)]\n",
    "        if gene_tss_df.shape[0] > 1:\n",
    "            tss_rspm_by_gene.append(list(gene_tss_df['rspm']))\n",
    "        \n",
    "rspm_stepdowns = {0: [1]*len(tss_rspm_by_gene)}\n",
    "\n",
    "for g_rspms in tss_rspm_by_gene:\n",
    "    rspm_sort = sorted(g_rspms, reverse=True)\n",
    "    for i in range(1, len(rspm_sort)):\n",
    "        if i in rspm_stepdowns:\n",
    "            rspm_stepdowns[i].append(rspm_sort[i]/rspm_sort[0])\n",
    "        else:\n",
    "            rspm_stepdowns[i] = [rspm_sort[i]/rspm_sort[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(8, 5))\n",
    "\n",
    "ax.boxplot(list(rspm_stepdowns.values()))\n",
    "ax.set_xticklabels([f'{i+1} (n={len(vals)})' for i, vals in enumerate(rspm_stepdowns.values())])\n",
    "ax.set_xlabel('Ordered TSS for Gene', fontsize=13)\n",
    "ax.set_ylabel('RSPM Rel. to Largest', fontsize=13)\n",
    "ax.tick_params(axis='both', labelsize=12)\n",
    "ax.set_title('Relative Strengths of Minor TSSes', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we're seeing that the secondary TSSes have a side range but are mostly much weaker than the primary. Tertiary and quaternary TSS are much less strong. Perhaps we should exclude the 3rd and 4th TSS where applicable (not many)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TSS Distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_min_ranges = []\n",
    "tss_max_ranges = []\n",
    "\n",
    "for gene_row in m_florum.gene_table.itertuples():\n",
    "    tus_for_gene = [tu for tu, genes in tu_to_genes.items() if gene_row.Index in genes]\n",
    "    if len(tus_for_gene) > 1:\n",
    "        gene_tu_df = m_florum.tu_table.loc[tus_for_gene, :]\n",
    "        gene_tsses = sorted(list(gene_tu_df['tss']))\n",
    "        diffs = np.array(gene_tsses[1:]) - np.array(gene_tsses[:-1])\n",
    "        tss_min_ranges.append(min(diffs))\n",
    "        tss_max_ranges.append(max(diffs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(8, 5))\n",
    "\n",
    "ax.boxplot([tss_min_ranges, tss_max_ranges], vert=False)\n",
    "ax.set_yticklabels(['Min', 'Max'])\n",
    "ax.set_xticks(np.arange(0, 10000, 1000))\n",
    "ax.set_xlabel('Inter-TSS Distance for Single Gene (bp)', fontsize=13)\n",
    "ax.tick_params(axis='both', labelsize=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TSS to -10 Box Distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(8, 5))\n",
    "\n",
    "ax.boxplot(m_florum.tss_table['tss_box_10_dist'], vert=False)\n",
    "ax.set_yticklabels([''])\n",
    "ax.set_xlabel('TSS to Start of -10 Box (bp)', fontsize=13)\n",
    "ax.set_title('TSS and -10 Box Relative Locations', fontsize=13)\n",
    "ax.tick_params(axis='both', labelsize=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This suggests that we may want to align promoters at the -10 box and not at the TSS for the same gene. We will do so below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare X and y - One-Hot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N_UP, N_DOWN = 50, 10\n",
    "\n",
    "one_hot_tus = m_florum.one_hot_encode_tu_promoters(N_UP, N_DOWN)\n",
    "one_hot_tus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_y_to_use = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TU Expression-Based\n",
    "\n",
    "Averaged Across all TU Genes (tu_avg)\n",
    "\n",
    "TU Expression of First Gene (tu_first)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tu_names = []\n",
    "tu_avgs = []\n",
    "tu_first_exps = []\n",
    "\n",
    "for tu_tag in one_hot_tus.index:\n",
    "    tu_genes = m_florum.tu_table.loc[tu_tag, 'gene_names'].split(';')\n",
    "    tu_genes_in_gene_table = list(set(tu_genes).intersection(set(list(m_florum.gene_table.index))))\n",
    "    if len(tu_genes_in_gene_table) > 0:\n",
    "        tu_names.append(tu_tag)\n",
    "        tu_gene_df = m_florum.gene_table.loc[tu_genes_in_gene_table, :]\n",
    "        # do the averaging of all tus\n",
    "        tu_avgs.append(np.mean(tu_gene_df['mean_fpkm']))\n",
    "        # figure out the first gene in the TU and just record that\n",
    "        first_gene_exp = tu_gene_df[tu_gene_df['left'] == min(tu_gene_df['left'])].iloc[0, :]['mean_fpkm']\n",
    "        tu_first_exps.append(first_gene_exp)\n",
    "        \n",
    "\n",
    "X_tu_avg = one_hot_tus.loc[tu_names, :]\n",
    "y_tu_avg = pd.Series(tu_avgs, index=tu_names)\n",
    "X_y_to_use['tu_avg'] = (X_tu_avg, y_tu_avg)\n",
    "\n",
    "X_tu_first = one_hot_tus.loc[tu_names, :]\n",
    "y_tu_first = pd.Series(tu_first_exps, index=tu_names)\n",
    "X_y_to_use['tu_first'] = (X_tu_first, y_tu_first)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Expression-Based\n",
    "\n",
    "Using Average of Promoters for All TUs Containing Gene (gene_prom_avg)\n",
    "- also with promoters aligned by -10 box and not TSS (gene_prom_avg_m10)\n",
    "\n",
    "Using strongest TSS for each gene (gene_strong_tss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tu_to_genes = dict(zip(m_florum.tu_table.index, m_florum.tu_table['gene_names']))\n",
    "tu_to_genes = {k: v.split(';') for k, v in tu_to_genes.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M10_UP = 44\n",
    "M10_DOWN = 16\n",
    "\n",
    "gene_names = []\n",
    "gene_names_m10 = []\n",
    "gene_one_hot_promoter_rows = []\n",
    "gene_oh_prom_m10_rows = []\n",
    "gene_one_hot_prom_rows_strong = []\n",
    "\n",
    "gene_exps = []\n",
    "gene_exps_m10 = []\n",
    "\n",
    "for gene_row in m_florum.gene_table.itertuples():\n",
    "    tus_for_gene = [tu for tu, genes in tu_to_genes.items() if gene_row.Index in genes]\n",
    "    if tus_for_gene:\n",
    "        \n",
    "        tu_one_hots = one_hot_tus.loc[tus_for_gene, :]\n",
    "        tu_one_hot_avg = tu_one_hots.mean(axis=0)\n",
    "        gene_one_hot_promoter_rows.append(tu_one_hot_avg)\n",
    "        \n",
    "        gene_tu_df = m_florum.tu_table.loc[tus_for_gene, :]\n",
    "        \n",
    "        m10_one_hots = []\n",
    "        for tu_row in gene_tu_df.itertuples():\n",
    "            if not pd.isna(tu_row.box_10_right):\n",
    "                if tu_row.strand == 1:\n",
    "                    m10_anchor = tu_row.box_10_right\n",
    "                    m10_one_hot = m_florum.one_hot_encode_sequence(m10_anchor-M10_UP, m10_anchor+M10_DOWN, 1)\n",
    "                    m10_one_hots.append(m10_one_hot)\n",
    "                else:\n",
    "                    m10_anchor = tu_row.box_10_left\n",
    "                    m10_one_hot = m_florum.one_hot_encode_sequence(m10_anchor-M10_DOWN, m10_anchor+M10_UP, -1)\n",
    "                    m10_one_hots.append(m10_one_hot)\n",
    "        if m10_one_hots:\n",
    "            gene_names_m10.append(gene_row.Index)\n",
    "            gene_exps_m10.append(gene_row.mean_fpkm)\n",
    "            m10_oh_avg_row = pd.DataFrame(m10_one_hots).mean(axis=0)\n",
    "            gene_oh_prom_m10_rows.append(m10_oh_avg_row)\n",
    "        \n",
    "        def tss_to_rspm(tss):\n",
    "            tss_df = m_florum.tss_table[m_florum.tss_table['tss'] == tss]\n",
    "            if tss_df.empty:\n",
    "                return None\n",
    "            else:\n",
    "                return tss_df.iloc[0, :]['rspm']\n",
    "        gene_tu_df['rspm'] = gene_tu_df['tss'].apply(tss_to_rspm)\n",
    "        gene_tu_df = gene_tu_df.sort_values(by='rspm', ascending=False)\n",
    "        strong_tu = gene_tu_df.index[0]\n",
    "        strong_tu_oh = one_hot_tus.loc[strong_tu, :]\n",
    "        gene_one_hot_prom_rows_strong.append(strong_tu_oh)\n",
    "        \n",
    "        gene_names.append(gene_row.Index)\n",
    "        gene_exps.append(gene_row.mean_fpkm)\n",
    "\n",
    "X_gene_prom_avg = pd.DataFrame(gene_one_hot_promoter_rows, index=gene_names)\n",
    "y_prom_avg = pd.Series(gene_exps, index=gene_names)\n",
    "X_y_to_use['gene_prom_avg'] = (X_gene_prom_avg, y_prom_avg)\n",
    "\n",
    "X_gene_prom_avg_m10 = pd.DataFrame(gene_oh_prom_m10_rows, index=gene_names_m10)\n",
    "y_prom_avg_m10 = pd.Series(gene_exps_m10, index=gene_names_m10)\n",
    "X_y_to_use['gene_prom_avg_m10'] = (X_gene_prom_avg_m10, y_prom_avg_m10)\n",
    "\n",
    "X_strong_tss = pd.DataFrame(gene_one_hot_prom_rows_strong, index=gene_names)\n",
    "y_strong_tss = pd.Series(gene_exps, index=gene_names)\n",
    "X_y_to_use['gene_strong_tss'] = (X_strong_tss, y_strong_tss)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finalize XY to Use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del X_y_to_use['tu_avg']\n",
    "del X_y_to_use['tu_first']\n",
    "del X_y_to_use['gene_prom_avg']\n",
    "del X_y_to_use['gene_prom_avg_m10']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set Aside Lockbox Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Xy_train_to_use = {}\n",
    "Xy_lockbox_to_use = {}\n",
    "\n",
    "for name, (X, y) in X_y_to_use.items():\n",
    "    if X.shape[0] > 500:\n",
    "        test_size = 0.1\n",
    "    else:\n",
    "        test_size = 0.2\n",
    "    X_train, X_lockbox, y_train, y_lockbox = train_test_split(X, y, test_size=test_size, random_state=42)\n",
    "    Xy_train_to_use[name] = (X_train, y_train)\n",
    "    Xy_lockbox_to_use[name] = (X_lockbox, y_lockbox)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Up Models to Try - Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_to_try = {\n",
    "    'ElasticNet': ElasticNet(),\n",
    "    'SVR': SVR(),\n",
    "    'RF': RandomForestRegressor(n_jobs=4),\n",
    "    'NN': MLPRegressor()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initial Model Profiling - One-Hot\n",
    "\n",
    "For each model:\n",
    "- for each Xy pair:\n",
    "    - try all 8 combinations of the following:\n",
    "        - raw and normalized X matrix\n",
    "        - raw, normalized, log-transformed, and log-transformed + normalized targets (expressions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "initial_profile_df = pd.DataFrame(\n",
    "    columns=['model', 'xy', 'train_r2', 'val_r2']\n",
    ")\n",
    "\n",
    "N_CV = 5\n",
    "\n",
    "for model_name, model in models_to_try.items():\n",
    "    for Xy_name, (X_train_raw, y_train_raw) in Xy_train_to_use.items():\n",
    "        \n",
    "        X_train_norm = StandardScaler().fit_transform(X_train_raw)\n",
    "        y_train_norm = StandardScaler().fit_transform(y_train_raw.values.reshape(-1, 1))\n",
    "        y_train_log = np.log(y_train_raw + ETA)\n",
    "        y_train_norm_log = StandardScaler().fit_transform(y_train_log.values.reshape(-1, 1))\n",
    "        \n",
    "        X_list = [('X_raw', X_train_raw), ('X_norm', X_train_norm)]\n",
    "        y_list = [('y_raw', y_train_raw), ('y_norm', y_train_norm),\n",
    "                 ('y_log', y_train_log), ('y_norm_log', y_train_norm_log)]\n",
    "        \n",
    "        Xy_final = {}\n",
    "        for (X_lab, X_mat), (y_lab, y_mat) in product(X_list, y_list):\n",
    "            Xy_lab_final = f'{Xy_name}__{X_lab}__{y_lab}'\n",
    "            Xy_final[Xy_lab_final] = (X_mat, y_mat)\n",
    "            \n",
    "        for Xy_full_name, (X_to_use, y_to_use) in Xy_final.items():\n",
    "\n",
    "            print(f'{model_name}: {Xy_full_name}')\n",
    "\n",
    "            cv_result = cross_validate(\n",
    "                model,\n",
    "                X_to_use,\n",
    "                y=y_to_use,\n",
    "                cv=N_CV,\n",
    "                scoring='r2',\n",
    "                return_train_score=True,\n",
    "                verbose=1,\n",
    "                n_jobs=N_CV\n",
    "            )\n",
    "\n",
    "            sub_result_df = pd.DataFrame(\n",
    "                data={\n",
    "                    'model': [model_name] * N_CV,\n",
    "                    'xy': [Xy_full_name] * N_CV,\n",
    "                    'train_r2': cv_result['train_score'],\n",
    "                    'val_r2': cv_result['test_score']\n",
    "                }\n",
    "            )\n",
    "\n",
    "            initial_profile_df = initial_profile_df.append(sub_result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='model', y='train_r2', data=initial_profile_df, hue='xy', dodge=True, ax=ax)\n",
    "ax.legend(loc='center left', bbox_to_anchor=(1.05, 0.5))\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Training R-Squared')\n",
    "ax.set_ylim(bottom=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='model', y='val_r2', data=initial_profile_df, hue='xy', dodge=True, ax=ax)\n",
    "ax.legend(loc='center left', bbox_to_anchor=(1.05, 0.5))\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Validation R-Squared')\n",
    "ax.set_ylim(bottom=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Error Analysis - One-Hot\n",
    "\n",
    "It's clear that the models using each gene as a training sample are better; let's analyze these in more detail. Specifically, let's use the \"X_raw_y_norm\" training set.\n",
    "\n",
    "Also, the SVR and RF models are better too. So let's focus on these as well. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Predicted vs Actual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_error, y_error = Xy_train_to_use['gene_strong_tss']\n",
    "\n",
    "# need to quickly fix the feature names to make sense\n",
    "new_cols = []\n",
    "\n",
    "for col in X_error.columns:\n",
    "    ind, base = col.split('_')\n",
    "    new_ind = int(ind)-N_UP\n",
    "    new_col = f'{new_ind}_{base}'\n",
    "    new_cols.append(new_col)\n",
    "\n",
    "X_error.columns = new_cols\n",
    "\n",
    "X_error = pd.DataFrame(StandardScaler().fit_transform(X_error), columns=X_error.columns, index=X_error.index)\n",
    "y_error = StandardScaler().fit_transform(y_error.values.reshape(-1, 1)).flatten()\n",
    "\n",
    "X_error_train, X_error_val, y_error_train, y_error_val = train_test_split(\n",
    "    X_error, y_error, test_size=0.15, random_state=35)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_error = models_to_try['RF'].fit(X_error_train, y_error_train)\n",
    "rf_y_pred = rf_error.predict(X_error_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax_rf = plt.subplots()\n",
    "\n",
    "ax_rf.scatter(y_error_val, rf_y_pred)\n",
    "ax_rf.set_xlabel('Actual FPKM (norm)', fontsize=13)\n",
    "ax_rf.set_ylabel('Predicted FPKM (norm)', fontsize=13)\n",
    "r2_rf = r2_score(y_error_val, rf_y_pred)\n",
    "ax_rf.set_title(f'Random Forest (R2={r2_rf:.2f})', fontsize=14)\n",
    "xmin, xmax = ax_rf.get_xlim()\n",
    "ymin, ymax = ax_rf.get_ylim()\n",
    "ax_rf.set_xlim(min(xmin, ymin), max(xmax, ymax))\n",
    "ax_rf.set_ylim(min(xmin, ymin), max(xmax, ymax))\n",
    "ax_rf.set_aspect('equal')\n",
    "ax_rf.plot([0, 1], [0, 1], transform=ax_rf.transAxes, color='gray', linestyle='--')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Residual Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax_rf = plt.subplots()\n",
    "\n",
    "ax_rf.scatter(y_error_val, abs(rf_y_pred - y_error_val))\n",
    "ax_rf.set_xlabel('Actual FPKM (norm)', fontsize=13)\n",
    "ax_rf.set_ylabel('Raw Error (norm[FPKM])', fontsize=13)\n",
    "ax_rf.set_title('Random Forest', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare error percentages as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax_rf = plt.subplots()\n",
    "\n",
    "ax_rf.scatter(y_error_val, ((rf_y_pred - y_error_val)/(y_error_val))*100)\n",
    "ax_rf.set_xlabel('Actual FPKM (norm)', fontsize=13)\n",
    "ax_rf.set_ylabel('Error (% of Actual)', fontsize=13)\n",
    "ax_rf.set_title('Random Forest', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we actually do quite poorly on the lower-expressed genes, a bit better on highly expressed. This definitely motivates a multiple model approach. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Error for Genes with Strong Secondary TSS\n",
    "\n",
    "In the strong-TSS case, we do have some genes that have multiple TSS with comparable signal intensities; are we doing worse on those than the ones with a clear strong TSS?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "val_genes_tss2_ratio = []\n",
    "val_genes_err = []\n",
    "val_genes_err_pct = []\n",
    "\n",
    "for val_gene, y_pred, y_true in zip(X_error_val.index, rf_y_pred, y_error_val):\n",
    "    tus_for_gene = [tu for tu, genes in tu_to_genes.items() if val_gene in genes]\n",
    "    if len(tus_for_gene) > 1:\n",
    "        gene_tu_df = m_florum.tu_table.loc[tus_for_gene, :]\n",
    "        gene_tsses = gene_tu_df['tss']\n",
    "        gene_tss_df = m_florum.tss_table[m_florum.tss_table['tss'].isin(gene_tsses)]\n",
    "        if gene_tss_df.shape[0] > 1:\n",
    "            gene_tss_df = gene_tss_df.sort_values(by='rspm', ascending=False)\n",
    "            ratio = gene_tss_df['rspm'][1]/gene_tss_df['rspm'][0]\n",
    "            val_genes_tss2_ratio.append(ratio)\n",
    "            val_genes_err.append(abs(y_pred - y_true))\n",
    "            val_genes_err_pct.append((abs(y_pred - y_true)/y_true)*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "_, ax = plt.subplots()\n",
    "\n",
    "ax.scatter(val_genes_tss2_ratio, val_genes_err_pct)\n",
    "ax.set_xlabel('Strength of 2nd TSS Relative to 1st', fontsize=13)\n",
    "ax.set_ylabel('Error (% of Actual)', fontsize=13)\n",
    "ax.set_title('Error by Secondary TSS (Validation Set)', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, so seems to be nothing doing here. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Importances - One-Hot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cv_featimp_rf = cross_validate(\n",
    "    models_to_try['RF'],\n",
    "    X_error,\n",
    "    y=y_error,\n",
    "    cv=5,\n",
    "    scoring='r2',\n",
    "    return_train_score=True,\n",
    "    return_estimator=True,\n",
    "    verbose=1,\n",
    "    n_jobs=4\n",
    ")\n",
    "\n",
    "estimators = cv_featimp_rf['estimator']\n",
    "feat_imps = [est.feature_importances_ for est in estimators]\n",
    "\n",
    "feat_bps = []\n",
    "bases = []\n",
    "imps = []\n",
    "\n",
    "for feat, feat_imps in zip(X_error.columns, zip(*feat_imps)):\n",
    "    bp, base = feat.split('_')\n",
    "\n",
    "    feat_bps += [int(bp)] * len(feat_imps)\n",
    "    bases += [base] * len(feat_imps)\n",
    "    imps += list(feat_imps)\n",
    "    \n",
    "feat_imp_cv_df = pd.DataFrame(data={'bp': feat_bps, 'imp': imps, 'base': bases})\n",
    "\n",
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.lineplot(x='bp', y='imp', hue='base', data=feat_imp_cv_df, ax=ax)\n",
    "ax.set_xlabel('bp from TSS')\n",
    "ax.set_ylabel('RF Importance')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shapley Values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "explainer = shap.TreeExplainer(rf_error)\n",
    "shap_values = explainer.shap_values(X_error_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shap.summary_plot(shap_values, X_error_val, plot_type=\"bar\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shap.summary_plot(shap_values, X_error_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classifying Expression Level - One-Hot\n",
    "\n",
    "Let's remind ourselves of the raw distribution of expression values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Re-examining Expression Distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(8, 5))\n",
    "sns.histplot(x='mean_fpkm', data=m_florum.gene_table, ax=ax)\n",
    "ax.set_xlabel('Mean FPKM', fontsize=13)\n",
    "ax.tick_params(axis='x', labelsize=12)\n",
    "ax.set_title('M. florum Gene Expression Distribution', fontsize=14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(8, 5))\n",
    "sns.stripplot(x='log_fpkm', data=gene_table_log, ax=ax)\n",
    "ax.set_xlabel('Log FPKM', fontsize=13)\n",
    "ax.tick_params(axis='x', labelsize=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perhaps we can try:\n",
    "- Classifying the cutoff point from the main blob (call it 4 Log FPKM)\n",
    "- Regressing ONLY the main distribution (drop data points below the cutoff)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare Classification Features and Targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Xy_to_use_class = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene with Strongest TSS - Classify Around Cutoff\n",
    "\n",
    "Let's just use the strongest TSS dataset, which seems to perform best at regression. Try a few different cutoff points for defining the classes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_cutoffs = []\n",
    "for log_cutoff in [5, 6, 6.25, 6.5, 6.75, 7, 8]:\n",
    "    raw_cutoff = np.exp(log_cutoff) - 1\n",
    "    raw_cutoffs.append(raw_cutoff)\n",
    "    \n",
    "for raw_cutoff in raw_cutoffs:\n",
    "\n",
    "    y_strong_tss_class = (y_strong_tss < raw_cutoff).apply(int)\n",
    "\n",
    "    Xy_to_use_class[f'gene_strong_tss_{raw_cutoff:.2f}'] = (X_strong_tss, y_strong_tss_class)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Aside Lockbox Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Xy_train_to_use_class = {}\n",
    "Xy_lockbox_to_use_class = {}\n",
    "\n",
    "for name, (X, y) in Xy_to_use_class.items():\n",
    "    if X.shape[0] > 500:\n",
    "        test_size = 0.1\n",
    "    else:\n",
    "        test_size = 0.2\n",
    "    X_train, X_lockbox, y_train, y_lockbox = train_test_split(X, y, test_size=test_size, random_state=42,\n",
    "                                                             stratify=y)\n",
    "    Xy_train_to_use_class[name] = (X_train, y_train)\n",
    "    Xy_lockbox_to_use_class[name] = (X_lockbox, y_lockbox)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set Up Models to Try"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_to_try_class = {\n",
    "    'Logistic': LogisticRegression(class_weight='balanced'),\n",
    "    'SVM': SVC(),\n",
    "    'RF': RandomForestClassifier(n_jobs=4),\n",
    "    'NN': MLPClassifier()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initial Model Profiling\n",
    "\n",
    "For each model:\n",
    "- for each Xy pair:\n",
    "    - try all 8 combinations of the following:\n",
    "        - raw and normalized X matrix\n",
    "        - raw, normalized, log-transformed, and log-transformed + normalized targets (expressions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "initial_profile_df = pd.DataFrame(\n",
    "    columns=['model', 'xy', 'train_auc', 'val_auc']\n",
    ")\n",
    "\n",
    "N_CV = 5\n",
    "\n",
    "for model_name, model in models_to_try_class.items():\n",
    "    for Xy_name, (X_train_raw, y_train_raw) in Xy_train_to_use_class.items():\n",
    "        \n",
    "        X_train_norm = StandardScaler().fit_transform(X_train_raw)\n",
    "        \n",
    "        # ONLY USE THE RAW DATASETS\n",
    "        X_list = [('X_raw', X_train_raw)]\n",
    "        y_list = [('y_raw', y_train_raw)]\n",
    "        \n",
    "        Xy_final = {}\n",
    "        for (X_lab, X_mat), (y_lab, y_mat) in product(X_list, y_list):\n",
    "            Xy_lab_final = f'{Xy_name}'\n",
    "            Xy_final[Xy_lab_final] = (X_mat, y_mat)\n",
    "            \n",
    "        for Xy_full_name, (X_to_use, y_to_use) in Xy_final.items():\n",
    "\n",
    "            print(f'{model_name}: {Xy_full_name}')\n",
    "\n",
    "            cv_result = cross_validate(\n",
    "                model,\n",
    "                X_to_use,\n",
    "                y=y_to_use,\n",
    "                cv=N_CV,\n",
    "                scoring='roc_auc',\n",
    "                return_train_score=True,\n",
    "                verbose=1,\n",
    "                n_jobs=N_CV\n",
    "            )\n",
    "\n",
    "            sub_result_df = pd.DataFrame(\n",
    "                data={\n",
    "                    'model': [model_name] * N_CV,\n",
    "                    'xy': [Xy_full_name] * N_CV,\n",
    "                    'train_auc': cv_result['train_score'],\n",
    "                    'val_auc': cv_result['test_score']\n",
    "                }\n",
    "            )\n",
    "\n",
    "            initial_profile_df = initial_profile_df.append(sub_result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='model', y='train_auc', data=initial_profile_df, hue='xy', dodge=True, ax=ax)\n",
    "ax.legend(loc='center left', bbox_to_anchor=(1.05, 0.5))\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Training AUC')\n",
    "ax.set_ylim(bottom=0, top=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='model', y='val_auc', data=initial_profile_df, hue='xy', dodge=True, ax=ax)\n",
    "ax.legend(loc='center left', bbox_to_anchor=(1.05, 0.5))\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Validation AUC')\n",
    "ax.set_ylim(bottom=0, top=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we're actually getting a pretty nice score with the random forest for the ~666 cutoff. Let's investigate that further, along with a general overview of error states"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FPKM_CUTOFF = 664.14"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Error Analysis\n",
    "\n",
    "Re-train a random forest cohort at the ~1000 cutoff as an exemplar of errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_error_class, y_error_class = Xy_train_to_use_class[f'gene_strong_tss_{FPKM_CUTOFF}']\n",
    "\n",
    "# need to quickly fix the feature names to make sense\n",
    "new_cols = []\n",
    "\n",
    "for col in X_error_class.columns:\n",
    "    ind, base = col.split('_')\n",
    "    new_ind = int(ind)-N_UP\n",
    "    new_col = f'{new_ind}_{base}'\n",
    "    new_cols.append(new_col)\n",
    "\n",
    "X_error_class.columns = new_cols\n",
    "\n",
    "X_error_class_train, X_error_class_val, y_error_class_train, y_error_class_val = train_test_split(\n",
    "    X_error_class, y_error_class, test_size=0.15, random_state=35, stratify=y_error_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_error_class = models_to_try_class['RF'].fit(X_error_class_train, y_error_class_train)\n",
    "rf_y_pred_class = rf_error_class.predict(X_error_class_val)\n",
    "\n",
    "conf_mat = confusion_matrix(y_error_class_val, rf_y_pred_class, normalize='true')\n",
    "\n",
    "_, ax = plt.subplots()\n",
    "sns.heatmap(\n",
    "    conf_mat,\n",
    "    cmap=\"PuBu\",\n",
    "    annot=True,\n",
    "    annot_kws={'fontsize': 13},\n",
    "    cbar=False,\n",
    "    square=True,\n",
    "    xticklabels=['High Exp', 'Low Exp'],\n",
    "    yticklabels=['High Exp', 'Low Exp'],\n",
    "    ax=ax\n",
    ")\n",
    "ax.tick_params(axis='both', labelsize=12)\n",
    "ax.tick_params(axis='y', rotation=0)\n",
    "ax.spines['right'].set_visible(True)\n",
    "ax.spines['left'].set_visible(True)\n",
    "ax.spines['top'].set_visible(True)\n",
    "ax.spines['bottom'].set_visible(True)\n",
    "ax.set_title(f'Random Forest (FPKM cut={FPKM_CUTOFF})')\n",
    "ax.set_xlabel('Predicted Class', fontsize=12)\n",
    "ax.set_ylabel('True Class', fontsize=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature Importances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cv_featimp_rf = cross_validate(\n",
    "    models_to_try_class['RF'],\n",
    "    X_error_class,\n",
    "    y=y_error_class,\n",
    "    cv=5,\n",
    "    scoring='roc_auc',\n",
    "    return_train_score=True,\n",
    "    return_estimator=True,\n",
    "    verbose=1,\n",
    "    n_jobs=4\n",
    ")\n",
    "\n",
    "estimators = cv_featimp_rf['estimator']\n",
    "feat_imps = [est.feature_importances_ for est in estimators]\n",
    "\n",
    "feat_bps = []\n",
    "bases = []\n",
    "imps = []\n",
    "\n",
    "for feat, feat_imps in zip(X_error_class.columns, zip(*feat_imps)):\n",
    "    bp, base = feat.split('_')\n",
    "\n",
    "    feat_bps += [int(bp)] * len(feat_imps)\n",
    "    bases += [base] * len(feat_imps)\n",
    "    imps += list(feat_imps)\n",
    "    \n",
    "feat_imp_cv_df = pd.DataFrame(data={'bp': feat_bps, 'imp': imps, 'base': bases})\n",
    "\n",
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.lineplot(x='bp', y='imp', hue='base', data=feat_imp_cv_df, ax=ax)\n",
    "ax.set_xlabel('bp from TSS')\n",
    "ax.set_ylabel('RF Importance')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The expected ranges (-35, -10, TSS) seem to contribute most to typing the classes.\n",
    "\n",
    "Can we use this classifier as part of model stacking to improve regression?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Shapley Values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "explainer = shap.TreeExplainer(rf_error_class)\n",
    "shap_values = explainer.shap_values(X_error_class_val)\n",
    "shap.summary_plot(shap_values, X_error_class_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model Stacking - One-Hot\n",
    "\n",
    "So we're going to try to use a classifier to decide whether we have high or low expression, then train models on each subset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Regression on Expression Subsets\n",
    "\n",
    "First let's see if we actually do better on regression when we look at the subsets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare Feature Matrices and Targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_y_to_use_exp_split = {}\n",
    "\n",
    "X_strong_tss_lowexp = X_strong_tss[y_strong_tss < FPKM_CUTOFF]\n",
    "y_strong_tss_lowexp = y_strong_tss[y_strong_tss < FPKM_CUTOFF]\n",
    "\n",
    "X_y_to_use_exp_split['gene_strong_tss_low'] = (X_strong_tss_lowexp, y_strong_tss_lowexp)\n",
    "\n",
    "X_strong_tss_highexp = X_strong_tss[y_strong_tss > FPKM_CUTOFF]\n",
    "y_strong_tss_highexp = y_strong_tss[y_strong_tss > FPKM_CUTOFF]\n",
    "\n",
    "X_y_to_use_exp_split['gene_strong_tss_high'] = (X_strong_tss_highexp, y_strong_tss_highexp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Xy_train_to_use_split = {}\n",
    "Xy_lockbox_to_use_split = {}\n",
    "\n",
    "for name, (X, y) in X_y_to_use_exp_split.items():\n",
    "    if X.shape[0] > 500:\n",
    "        test_size = 0.1\n",
    "    else:\n",
    "        test_size = 0.2\n",
    "    X_train, X_lockbox, y_train, y_lockbox = train_test_split(X, y, test_size=test_size, random_state=42)\n",
    "    Xy_train_to_use_split[name] = (X_train, y_train)\n",
    "    Xy_lockbox_to_use_split[name] = (X_lockbox, y_lockbox)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initial Model Profiling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "initial_profile_df_split = pd.DataFrame(\n",
    "    columns=['model', 'xy', 'train_r2', 'val_r2']\n",
    ")\n",
    "\n",
    "N_CV = 5\n",
    "\n",
    "for model_name, model in models_to_try.items():\n",
    "    for Xy_name, (X_train_raw, y_train_raw) in Xy_train_to_use_split.items():\n",
    "        \n",
    "        X_train_norm = StandardScaler().fit_transform(X_train_raw)\n",
    "        y_train_norm = StandardScaler().fit_transform(y_train_raw.values.reshape(-1, 1))\n",
    "        y_train_log = np.log(y_train_raw + ETA)\n",
    "        y_train_norm_log = StandardScaler().fit_transform(y_train_log.values.reshape(-1, 1))\n",
    "        \n",
    "        X_list = [('X_raw', X_train_raw)]\n",
    "        y_list = [('y_raw', y_train_raw), ('y_norm', y_train_norm),\n",
    "                 ('y_log', y_train_log), ('y_norm_log', y_train_norm_log)]\n",
    "        \n",
    "        Xy_final = {}\n",
    "        for (X_lab, X_mat), (y_lab, y_mat) in product(X_list, y_list):\n",
    "            Xy_lab_final = f'{Xy_name}__{X_lab}__{y_lab}'\n",
    "            Xy_final[Xy_lab_final] = (X_mat, y_mat)\n",
    "            \n",
    "        for Xy_full_name, (X_to_use, y_to_use) in Xy_final.items():\n",
    "\n",
    "            print(f'{model_name}: {Xy_full_name}')\n",
    "\n",
    "            cv_result = cross_validate(\n",
    "                model,\n",
    "                X_to_use,\n",
    "                y=y_to_use,\n",
    "                cv=N_CV,\n",
    "                scoring='r2',\n",
    "                return_train_score=True,\n",
    "                verbose=1,\n",
    "                n_jobs=N_CV\n",
    "            )\n",
    "\n",
    "            sub_result_df = pd.DataFrame(\n",
    "                data={\n",
    "                    'model': [model_name] * N_CV,\n",
    "                    'xy': [Xy_full_name] * N_CV,\n",
    "                    'train_r2': cv_result['train_score'],\n",
    "                    'val_r2': cv_result['test_score']\n",
    "                }\n",
    "            )\n",
    "\n",
    "            initial_profile_df_split = initial_profile_df_split.append(sub_result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='model', y='train_r2', data=initial_profile_df_split, hue='xy', dodge=True, ax=ax)\n",
    "ax.legend(loc='center left', bbox_to_anchor=(1.05, 0.5))\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Training R-Squared')\n",
    "ax.set_ylim(bottom=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='model', y='val_r2', data=initial_profile_df_split, hue='xy', dodge=True, ax=ax)\n",
    "ax.legend(loc='center left', bbox_to_anchor=(1.05, 0.5))\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Validation R-Squared')\n",
    "ax.set_ylim(bottom=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This doesn't really work well at all...let's table this for now"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome Organization Features\n",
    "\n",
    "Switching gears a bit, we want to integrate our genome organization-style features into this one-hot model. Let's start with a baseline model that just uses these genome organization features at first, to see where that's at"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare Feature Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot expression on log scale as well for all feature plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Replication Region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>strand</th>\n",
       "      <th>KO</th>\n",
       "      <th>category</th>\n",
       "      <th>subcategory</th>\n",
       "      <th>subsubcategory</th>\n",
       "      <th>product_name</th>\n",
       "      <th>name</th>\n",
       "      <th>fpkm</th>\n",
       "      <th>primary_tu</th>\n",
       "      <th>log_fpkm</th>\n",
       "      <th>rep_strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>peg.1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>1</td>\n",
       "      <td>K02313</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>DNA Maintenance</td>\n",
       "      <td>DNA replication and partition</td>\n",
       "      <td>dnaA, chromosomal replication initiator protein</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>192.793077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.266791</td>\n",
       "      <td>leading</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peg.2</th>\n",
       "      <td>1560.0</td>\n",
       "      <td>2681.0</td>\n",
       "      <td>1</td>\n",
       "      <td>K02338</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>DNA Maintenance</td>\n",
       "      <td>DNA replication and partition</td>\n",
       "      <td>dnaN, DNA polymerase III subunit beta [EC:2.7....</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>642.952373</td>\n",
       "      <td>TU_002</td>\n",
       "      <td>6.467625</td>\n",
       "      <td>leading</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peg.3</th>\n",
       "      <td>2708.0</td>\n",
       "      <td>3682.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>K03569</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cytoskeleton</td>\n",
       "      <td>Cytoskeleton proteins</td>\n",
       "      <td>mreB, rod shape-determining protein MreB and r...</td>\n",
       "      <td>mreB</td>\n",
       "      <td>2046.205311</td>\n",
       "      <td>TU_004</td>\n",
       "      <td>7.624231</td>\n",
       "      <td>lagging</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peg.4</th>\n",
       "      <td>3751.0</td>\n",
       "      <td>4278.0</td>\n",
       "      <td>1</td>\n",
       "      <td>K05985</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>Folding, Sorting and Degradation</td>\n",
       "      <td>Nucleases</td>\n",
       "      <td>rnmV, ribonuclease M5 [EC:3.1.26.8]</td>\n",
       "      <td>rnmV</td>\n",
       "      <td>8.156475</td>\n",
       "      <td>TU_005</td>\n",
       "      <td>2.214461</td>\n",
       "      <td>leading</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peg.5</th>\n",
       "      <td>4280.0</td>\n",
       "      <td>5083.0</td>\n",
       "      <td>1</td>\n",
       "      <td>K02528</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Ribosome biogenesis</td>\n",
       "      <td>ksgA, 16S rRNA (adenine1518-N6/adenine1519-N6)...</td>\n",
       "      <td>ksgA</td>\n",
       "      <td>215.862583</td>\n",
       "      <td>TU_005</td>\n",
       "      <td>5.379264</td>\n",
       "      <td>leading</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             left   right  strand      KO                        category  \\\n",
       "locus_tag                                                                   \n",
       "peg.1         1.0  1332.0       1  K02313  Genetic Information Processing   \n",
       "peg.2      1560.0  2681.0       1  K02338  Genetic Information Processing   \n",
       "peg.3      2708.0  3682.0      -1  K03569              Cellular Processes   \n",
       "peg.4      3751.0  4278.0       1  K05985  Genetic Information Processing   \n",
       "peg.5      4280.0  5083.0       1  K02528  Genetic Information Processing   \n",
       "\n",
       "                                subcategory                 subsubcategory  \\\n",
       "locus_tag                                                                    \n",
       "peg.1                       DNA Maintenance  DNA replication and partition   \n",
       "peg.2                       DNA Maintenance  DNA replication and partition   \n",
       "peg.3                          Cytoskeleton          Cytoskeleton proteins   \n",
       "peg.4      Folding, Sorting and Degradation                      Nucleases   \n",
       "peg.5                           Translation            Ribosome biogenesis   \n",
       "\n",
       "                                                product_name  name  \\\n",
       "locus_tag                                                            \n",
       "peg.1        dnaA, chromosomal replication initiator protein  dnaA   \n",
       "peg.2      dnaN, DNA polymerase III subunit beta [EC:2.7....  dnaN   \n",
       "peg.3      mreB, rod shape-determining protein MreB and r...  mreB   \n",
       "peg.4                    rnmV, ribonuclease M5 [EC:3.1.26.8]  rnmV   \n",
       "peg.5      ksgA, 16S rRNA (adenine1518-N6/adenine1519-N6)...  ksgA   \n",
       "\n",
       "                  fpkm primary_tu  log_fpkm rep_strand  \n",
       "locus_tag                                               \n",
       "peg.1       192.793077        NaN  5.266791    leading  \n",
       "peg.2       642.952373     TU_002  6.467625    leading  \n",
       "peg.3      2046.205311     TU_004  7.624231    lagging  \n",
       "peg.4         8.156475     TU_005  2.214461    leading  \n",
       "peg.5       215.862583     TU_005  5.379264    leading  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table_ml = m_florum.gene_table.copy()\n",
    "\n",
    "gene_table_ml['rep_strand'] = gene_table_ml.apply(m_florum.replication_region, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.025, 0.05, 'P = 0.008')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfUAAAFBCAYAAABuEzZyAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAeGElEQVR4nO3de5RddX338feXhJB7ImQKbTAEmCLeiIR4oRUeLCpKiwjooymgvRlYtGmsKQWiWNsHgSDgypPCg6EqoJBqBVSoBrxAQYqXEEwERRgiCUSpA+YKlEj8Pn/snXgyTDIzmcs+Z+f9Wuus4ezfPnt/J4s9n7P3/u3fLzITSZLU+vaougBJkjQwDHVJkmrCUJckqSYMdUmSasJQlySpJgx1SZJqYnjVBfTXpEmTcurUqVWXIUnSkLjvvvueysy27tqGLNQj4r3AXwPTgNGZObxL+9uAy4CDgEeBD2Xm7T1td+rUqSxdunQQKpYkqflExKodtQ3l5fe1wJXAB7s2RMRBwE3ARcCE8ufNETF1COuTJKmlDVmoZ+ZtmbkYWNlN8/uB+zLz85m5OTOvB5aVyyVJUi80S0e5acB9XZYtK5dLkqReaJZQHwes77JsHTC+u5UjYlZELI2IpZ2dnYNdmyRJLaFZQn0jxb30RhOBDd2tnJmLMnNGZs5oa+u2A6AkSbudZgn15cD0LssOL5dLkqReGLJQj4hhETESGFG+H1m+ArgOmBERMyNiz4iYCRwBXDtU9UmS1OqGcvCZ04HPNrx/rvx5YGY+GhEnUzyn/hmKHvInZeZjQ1ifJEktbchCPTOvAa7ZSfsSYMlQ1SNJUt20/DCxKixcuJCOjo6qy+i1NWvWADB58uSKK+mb9vZ2Zs+eXXUZktQtQ12VeO6553peSZLUJ4Z6TbTa2eOcOXMAWLBgQcWVSFJ9NMsjbZIkqZ8MdUmSasJQlySpJgx1SZJqwlCXJKkm7P0uSf3QamNEQGuOE+EYEb1jqEvSbsZxIurLUJekfmjFs0fHiagv76lLklQThrokSTVhqEuSVBOGuiRJNWGoS5JUE4a6JEk1YahLklQThrokSTVhqEuSVBOGuiRJNWGoS5JUE4a6JEk1YahLklQThrokSTVhqEuSVBOGuiRJNWGoS5JUE4a6JEk1YahLklQThrokSTVhqEuSVBOGuiRJNWGoS5JUE4a6JEk1YahLklQThrokSTVhqEuSVBOGuiRJNWGoS5JUE4a6JEk10TShHhH7RcQXIqIzItZGxLcjYlrVdUmS1CqaJtSBK4G9gUOAfYGlwK0REZVWJUlSi2imUG8H/j0z12bmZuDTwP7APtWWJUlSa2imUP8EcEpEtEXESGAW8J3MfKriuiRJagnNFOr3AMOAXwKbgJOBD3S3YkTMioilEbG0s7NzCEuUJKl5NUWoR8QewDeBh4EJwGjg48DdEbFv1/Uzc1FmzsjMGW1tbUNbrCRJTaopQp2ig9yBwMLM3JCZmzPzXynqO7La0iRJag1NEerlffOHgbMiYkxEDI+IvwDGASuqrU6SpNYwvOoCGryTorPcKmBPoAN4d2aurLIoSZJaRdOEemb+BPiTquuQJKlVNcXld0mS1H+GuiRJNWGoS5JUE4a6JEk1YahLklQThrokSTVhqEuSVBOGuiRJNWGoS5JUE00zopwkLVy4kI6OjqrLqL2t/8Zz5sypuJJ6a29vZ/bs2UO6T0O9G/5hGXz+URk6Vfxh2VUdHR088uD9TBm7pepSam3Er4uLtM+vWlpxJfW1etOwSvZrqHejo6ODHz7wE7aM3rvqUmprj80JwH0r/7viSupt2LO/qrqEPpsydgvzpm+ougypXy5cNr6S/RrqO7Bl9N48d+jxVZch9cuoh75WdQmShpAd5SRJqglDXZKkmjDUJUmqCUNdkqSaMNQlSaoJQ12SpJow1CVJqglDXZKkmjDUJUmqCUNdkqSaMNQlSaoJQ12SpJow1CVJqglDXZKkmjDUJUmqCUNdkqSaMNQlSaoJQ12SpJow1CVJqglDXZKkmhhedQHNaM2aNQx7dj2jHvpa1aVI/TLs2adZs+aFqsuQNEQ8U5ckqSY8U+/G5MmTefL54Tx36PFVlyL1y6iHvsbkyftWXYakIeKZuiRJNWGoS5JUE4a6JEk1YahLklQThrokSTXRVKEeEW+OiO9GxKaIeCoirqy6JkmSWkXTPNIWEccAXwL+CrgFCOAVFZYkSVJLaZpQBy4CrsrMLzUsW1ZVMZIktZqmuPweEWOA1wHDI2JZeen9zoiYUXVtkiS1iqYIdeAlFLXMBP4M+D3gduBrETGx68oRMSsilkbE0s7OzqGsU5KkprXTUI+IlT29BqiOjeXPz2bmiszcTHE5fk/gD7qunJmLMnNGZs5oa2sboBIkSWptPd1Tnwr8GPgs8ORgFZGZ6yPiMSC7NnWzTJIkdaOnUH8D8AHgw8CdwNXAkswcjKC9EpgTEYuBh4EPAc8D/zUI+5IkqXZ2evk9M7+fmR8ApgBfB/4Z+FlEnB8REwa4lkuBzwDfBp4C3g68PTPXD/B+JEmqpV490paZm4CrI+IzwPnAPwL3UATwgCjP/j9aviTthtasWcMzG4dx4bLxVZci9cuqjcMYs2bNkO+3V73fI2JqRFwArALeQjFAzD2DWZgkSeqbnZ6pR8S7KO6pHw5cDxyXmQ8ORWGSdj+TJ0/m+Rd+wbzpG6ouReqXC5eNZ6/Jk4d8vz1dfv8iRe/3q4D/AU6MiBMbV8jMCwepNkmS1Ac9hfpdFI+UHbWD9gQMdUmSmsBOQz0zj9lZe0Q009jxkiTt1noaUe69O2nbg+I+uyRJagI99X5fFBFv2kHbp4HpA1yPJEnaRT2F+hnAjRHx6saFEXEFcGz5kiRJTaCne+qLI+L3gK9HxJGZ+XhEzAdOAY7KzNVDUmUFhj37K0Y99LWqy6itPf6neGTpNyMdZGQwDXv2V8C+VZchaYj02NEtMy+LiMnAbRHxVYqBZ96UmY8MenUVaW9vr7qE2uvoKCbmaz/IwBlc+/r/s7Qb6e0wsR+KiN8FzgTekpkrBresas2ePbvqEmpvzpw5ACxYsKDiSiSpPnoaUe7X/Hbq06C4B39PRGxbJzNHDFp1kiSp13o6U3/zkFQhSZL6raeOcv85VIVIkqT+6WnwmQMi4s6IWB8R3yh7wkuSpCbU03Pql5XrzANGApcMekWSJGmX9HRP/Y3AqzOzMyJuBu4dgpokSdIu6OlMfXRmdgJk5s+BcYNfkiRJ2hU9nalH+Xx67OD91rCXJEkV6ynUxwBPNLyPhvdB8Qz7sEGoS5Ik9VFPoX7gkFQhSZL6rafn1FcNVSGSJKl/enpO/Sdd3n9ocMuRJEm7qqfe7/t3ef+RwSpEkiT1T0+hnl3eR7drSZKkyvUU6l11DXlJktQkeur9Pioibm94P7bLezLzrQNfliRJ6queQv2CLu/vGaxCJElS//T0SNs/DVUhkiSpf/p6T12SJDUpQ12SpJow1CVJqomeOspJ0pBavWkYFy4bX3UZtfbfzxbnc/uO/k3FldTX6k3D+P0K9muoS2oa7e3tVZewW9jc0QHAXgf47z1Yfp9q/n/uVahHxG/ofuCZzcAq4Abg4szcPIC1SdrNzJ49u+oSdgtz5swBYMGCBRVXooHW2zP1DwIfAD5JEeIHAHOA64BNwNkUc6+fM/AlSpKk3uhtqP85cEJmPrZ1QUTcAdyUmYdHxL3AVzDUJUmqTG97vx8E/LzLsp8DBwNk5gqgbQDrkiRJfdTbUL8fmB8RewGUPy8qlxMRBwFPD0qFkiSpV3ob6h8A3g6si4hVwFrg+HI5wO/ipXdJkirVq3vqmflIRLwSOBL4PWAN8N3M3FK2O9GLJEkV6/Vz6mWAfyciJmXmU4NVUETsAXyH4gvESzPzicHalyRJddKry+8RMTIi/iUingH+OyKeiYiFETFyEGr6O+DZQdiuJEm11tt76hcBrwNOAg4pf762XD5gIuIQ4Czg7wdyu5Ik7Q56e/n9ZOANmfmL8v2jEfEA8F2KM+t+Ky+7f4Yi0NcNxDYlSdqd9PZMfTRFj/dGa4FRA1jLHODJzLy5pxUjYlZELI2IpZ2dnQNYgiRJrau3oX4PcPnWe+jlz0uBeweiiIhoB+YCf9Ob9TNzUWbOyMwZbW2OeSNJEvQ+1P8WOApYGxGrKc7SjwYGavaFN1KMSPdARDwFLCuXr4iIswZoH5Ik1Vpvn1NfHRGvoegs91LgceD7wD4DVMcXgW82vN+f4irAW4GHBmgfkiTVWl+fU7+3fG0dKvYXwLD+FpGZz9LwGFtEbK3ryczc1N/tS5K0O+h1qO9ADEgVXZSzwQ3KtiVJqqve3lPfkRyQKiRJUr/1N9QlSVKT2Onl94iYt6uflSRJQ6unYH5LD+13DVQhkiSpf3Ya6pn5pqEqRJIk9Y/31CVJqglDXZKkmjDUJUmqCUNdkqSaMNQlSaoJQ12SpJow1CVJqglDXZKkmjDUJUmqCUNdkqSaMNQlSaoJQ12SpJow1CVJqglDXZKkmjDUJUmqCUNdkqSaMNQlSaoJQ12SpJow1CVJqglDXZKkmjDUJUmqCUNdkqSaMNQlSaqJ4VUXoIGxcOFCOjo6qi6j17bWOmfOnIor6Zv29nZmz55ddRmS1C1DXZUYNWpU1SVIUu0Y6jXh2aMkyXvqkiTVhKEuSVJNGOqSJNWEoS5JUk0Y6pIk1YShLklSTRjqkiTVhKEuSVJNGOqSJNWEoS5JUk00TahHxPyIeDAiNkTEzyPi6ojYu+q6JElqFU0T6sAW4DRgH2AasD9wTZUFSZLUSppmQpfMnNfwtjMiFgBfrKoeSZJaTTOdqXd1LLC8u4aImBURSyNiaWdn5xCXJUlSc2rKUI+IU4AzgTndtWfmosyckZkz2trahrY4SZKaVNOFekS8G7gaeEdmLqu6HkmSWkVThXpE/DnwKeCEzLyj6nokSWolTRPqEfG3wKXAcZl5T9X1SJLUapqm9zuwAHgBuCMiti3MzLGVVSRJUgtpmlDPzOh5LUmStCNNc/ldkiT1j6EuSVJNGOqSJNWEoS5JUk0Y6pIk1YShLklSTRjqkiTVhKEuSVJNGOqSJNWEoS5JUk0Y6pIk1YShLklSTRjqkiTVhKEuSVJNGOqSJNWEoS5JUk0Y6pIk1YShLklSTRjqkiTVhKEuSVJNGOqSJNWEoS5JUk0Y6pIk1YShLklSTRjqkiTVhKEuSVJNGOqSJNWEoS5JUk0Y6pIk1YShLklSTRjqkiTVhKEuSVJNGOqSJNWEoS5JUk0Y6pIk1cRuF+rHHHMMe+21F2PHjmXChAkcfvjh3HjjjYOyry1btnD22WfT1tbGuHHjOOWUU3jqqad2+pklS5bwyle+klGjRvGqV72K22+/fbv2jo4O3vzmNzNmzBj2339/Lrvssu3af/rTn3L88cezzz77MGnSJE488UQee+yxgf7VJElNaLcLdYDzzz+fTZs28fTTTzNz5kze85738PDDDw/4fi6++GK+8pWv8L3vfY8nnngCgNNPP32H669cuZKTTz6Z8847j/Xr13Peeedx0kknbQvlLVu2cMIJJ/Dyl7+czs5OvvrVrzJ//ny+8IUvbNvGzJkzmTRpEo8//jirVq1i3LhxnHrqqQP+u0mSms9uGepbDR8+nLPOOostW7bwox/9aMC3v2jRIs455xwOOuggJkyYwCWXXMKSJUtYtWpVt+tfe+21HHHEEZx22mmMGDGCU089lenTp3PttdcCcNddd7Fq1SouuugiRo8ezfTp0znjjDO46qqrtm2jo6OD0047jdGjRzNmzBhOP/10li9fPuC/mySp+QyvuoAqbd68mSuuuII999yTadOmdbvOWWedxQ033LDDbZx77rmce+65L1q+bt06Vq9ezRFHHLFt2cEHH8z48eNZvnw5BxxwwIs+s3z58u3WB5g+ffq2UF6+fDmHHHIIY8eO3a79iiuu2K6e6667jiOPPJLM5JprruGkk07aYf2S+mfhwoV0dHRUXUafbK13zpw5FVfSe+3t7cyePbvqMprebhnqH//4x7n00ksZMWIE7e3t3HjjjbS3t3e77pVXXsmVV17Z531s3LgRgAkTJmy3fOLEiWzYsGGHn+lu/QcffHCn7Y3be9vb3sZNN93ExIkTyUwOO+wwbrvttj7XL6m+Ro0aVXUJGiS7Zah/+MMf5iMf+cig7mPcuHEArF+/frvl69atY/z48Tv8zM7W76l97dq1HHvsscydO5e7776bzOSSSy7hqKOOYsWKFYwcOXJAfjdJv+XZo5pJ09xTj4hhEfGJiOiMiI0RcWNETKq6rjPPPJOxY8fu8HXhhRd2+7mJEycyZcoUli1btm3ZypUr2bBhA4cddli3n5k2bdp26wPcf//9224NTJs2jYcffphnnnmm2/ZHH32UdevWMXfuXEaNGsXo0aOZO3cujzzyCA899FC//h0kSc2vaUIdOBc4EXg9sH+57HPVlVO46qqr2LRp0w5f8+bN2+FnZ82axfz58/nZz37Ghg0bOOecczjuuOOYOnVqt+u/733vY+nSpSxevJhf//rXLF68mPvuu4/3v//9ABx99NEccMABzJs3j+eee44f/vCHfOpTn+KMM84A4NBDD2XvvfdmwYIFbN68meeff57LL7+c8ePH7/D2giSpPpop1GcB8zNzZWauB/4BeFtEvLhHWYs499xzOeGEE3jta1/L5MmT2bJlC5///Oe3tV9//fXbdXo7+OCDuemmm7jgggsYP348F1xwATfffPO2LwHDhg3jlltu4YEHHmCfffbh+OOP5+yzz+a9730vAGPHjuXWW29lyZIl7Lfffuy3335885vf5NZbb91uP5KkeorMrLoGImIisBY4PDN/2LB8PXB6Zn61y/qzKL4EMGXKlCN29IiYJEl1ExH3ZeaM7tqa5Ux9XPlzfZfl64AX9SrLzEWZOSMzZ7S1tQ12bZIktYRmCfWN5c8JXZZPBLp//kuSJG2nKUI9M9cBq4HpW5dFxEEUZ+krKipLkqSW0hShXloEnBMRB0bEeGA+cFtmPlZtWZIktYZmGnzmYuAlwA+AvYBvAKdVWpEkSS2kaUI9M7cAf1++JElSHzXT5XdJktQPhrokSTVhqEuSVBOGuiRJNWGoS5JUE4a6JEk10RQTuvRHRHQCzujSmiYBT1VdhLSb8vhrXQdkZrcTn7R8qKt1RcTSHc00JGlwefzVk5ffJUmqCUNdkqSaMNRVpUVVFyDtxjz+ash76pIk1YRn6pIk1YShrh5FxJ0R8ZFB2vZpEfFYw/uvR8Q/DMa+pFYzmMdeN/u6KiL+ZSj2pcHTNFOvSgCZ+faqa5B2R5l5ZtU1qP88U5ckqSYMdfVJREyJiC9FxJMR8YuIWBQR4xraL4yIlRGxKSIejYgPdvn86yJiadn+HeCgLu3bLjdGxNSIyIg4PSJ+HBEbI+L2iPjdhvX3i4hbImJ9RDwcEX9ZfmbqoP5DSEMsIj4bEY+Xx8GPI+JPu7T/cbl8U0TcGhGfjIg7G9oPiYj/jIgNEbE8IuZERDa0XxMR/9rwPiPirIj4QbnP70bEoQ3t4yLiuoj4VUSsioj3RcQLEXHMoP5DaKcMdfVaRIwEvg38GDgQeAWwP7CgYbUfA28ExgEfAC6KiOPKz08Avg58Cdgb+DvgrF7s+j3A0cBkYAzwzw1t1wObgZeW+z191347qel9B3gNMJHiGLgmIl4BEBEHAzcB/6ds/yTwl1s/GBHDgVuA5cC+wEkUx2dP/gw4hWJI2ceBhQ1tCyi+lB8KvBr4Y2DYLv1mGjCGuvriTygeg/xoZj6XmWuB84FTI2IYQGZ+PjN/noVvA/8BHNvw+WeA+Zm5OTN/AHy6F/v9p8x8KjM3ADcAMwAiYn/gj4CzM3NDZv6S4o+aVDuZ+enMfDozt2TmvwErgGPK5pnA9zJzcWa+kJnfAr7S8PE3AFOBc8pjdyVF8PfkE5m5OjOfB67ht8feMOBU4KOZ+cvy2JzX/99S/WVHOfXFgcCUiFjXZXkC+wFrIuJvKc4A9gcCGEURxJTLVuX2gyP8rBf7/UXDfz9DcRUAijN3gNUN7U7uo9qJiD2Aj1FctdqP4pgbA2yd1GMyL/5/fxXFFayt7b/MzOe6tPdkR8feJGBEl2147DUBz9TVF6uAhzNzYpfXyMxcExF/CMwHzgAmZeZEikt+UX5+DXBARETDNqf2o5415c8pDcumdLei1OJmAn9FcSn8JeWxtZwux1aXzzQeC2uAtogYtYP2vnqK4rZX4z499pqAoa6+uBUYERHzyk4yERGTI+Kksn08sAXoBDIi/hh4e5fPjwXOjog9I2I6Dff9+ioznwDuBC4u62kDhuSZXmmIjQdeoDi29oiIvwCmNbT/G/D6iPjfETEsIt4EvLOh/bsUV7QuioiREXEg8MFdLSYzt1BcgftYRLSVnWU/vqvb08Ax1NVrmfksxT3sVwAPAeuBb1F03gG4DbgO+D7FN/l3ATc3fH4dRWea9wBrgf8L/L9+lvWnwGjgCeAe4N/L5c/3c7tSM7kW+B7QQXHW/Qrg7q2NmdkBvBv4J4rjci7wOcrjIDNfAN4BTKf4YvDlsn1zP2qaQ/FF4WHgAeAbFLcFPPYq5NjvqpWyp/1XgFHp/9zajUXEYmBjZs7aQfsZwNzMPGSA9vcyii/7kzPz5wOxTfWdZ+pqaRHxmog4rLwVcBBwAfAFA127m4h4R0TsHRHDI+JEivvvixva3xgRB5fHymHAPzS278L+DoqIPygv9+9L0Zv+LgO9Woa6Wt1LKJ7P3UTxHO8KisuC0u7maIpL4euAi4AzM/OOhvaXAndQ9GK/heLW2EX92N9Iiulb1wM/Ap6luB2mCnn5XZKkmvBMXZKkmjDU1a2IOC4i7u55zeYTEf8WEbv8qJzUbJr1eIyIvSKio3FMeFXLUNeLlIPDfBL4x0Ha/sERcXM5Ccv6cqKIPRvaZ0TE9yPi2SgmhTmtoe13ykkkVpUTV3RExHldBrT5GHBhl4E2pJY0mMdj2cltfvx2opgfRcS7GtpfVh6fT0cxEcyPI2Jbb/py+NhPlC81AUNd3XkrxRCQd/S0Yl+VA8TcTTEa1hSKiV3+hmLQmsZJX26k6AR3JnBVRBxZbmIsxaQxx1AMWflOihHs/m7rPjLzIYrneWcOdP1SBQbteAT+mmISpDdTDHBzPnBDw5n3k2X772TmeOC9wAUR8daGbSwG/igi2gehPvWRoa7uvBP45tbHwqKYDvWyiLix/Db/aPnIzK74ELA6Mz+WmevLySmWZuZvyvaTKXrRXpKZz2fmNyh66c4CyMyVmXlxZv6snDTmAYrRtI7psp9vsP2IWlKreieDdzy2A3dm5k/L4+nLwNPAqwDKY/SRcgQ5KAaXSeBlWzdQTubyA4rBbVQxQ13dmU5xNtzo/cBlwATgX4BrI2I0bHv+dd1OXisatvMm4PGI+I8o5mFeERGnNrRPA+7v8pz5MrYfEnObKCa6OIbizL/Rj8rfQ2p1g3k8Xg28KiJeUV6KfxfFRF93Ne6sPE6fp3hk9Je8+Pl2j7cm4Sxt6s5LgA1dln0hM/8LICIWAZcDvw8sz8zvUMzh3BuTgNdSDBV7IkXI3xIRq8rtjKN47rXROopLg925vKz30i7LN1Bc2pda3WAejyspboc9APyGYojX08tpjLfJzMPKfi9Hl69numxnA3Bwb38hDR7P1NWdtbw4RLdNwZiZWw/ocfTdRuDezPxSOe/zN4Al/PbS3UaKs49GE3nxHzUi4nKKCWOOzcyuXwTGA7/ahfqkZjOYx+OVwOEU0yqPAN5C0YflrV1XzMxfl/O0twEf7dLs8dYkDHV1536KCSN6JSKOKnui7+j1YMPqP6S4J9fV1mXL+e0EMVsdTsPl9YjYIyKupuhA9L/K2dq6elX5e0itbjCPxyOAz2Xmqsz8TXn2fzdw/E52MZziqkAjj7cmYairO18Gju3typl5d2aO3cnrlQ2rfwp4Q0S8swznN1GE85fL9puBMRFxdkSMiIhjKTrPLQKIiOHA9cAM4JjMfHIHZb2lYZtSK/syg3c83gOcGhGTASLi9RR9VO4r3x8XEa8vj8U9yw55p1E8oUK5zjjgdcBX+/l7agAY6urObcALEXHMQG84M79LMT70fIpL7QuB92fmvWX7OoqzhHdT3Fu/mmIM63vLTfwhxWM1Lwceazj7aPwj8zKKM4kbBrp+qQKDdjwCZ1PcT/9+RGyk+MJ8WWZ+rmwfD3yG4tJ6J8UYEHMz89MN25gJ3JGZjwxCfeojx35XtyLibcC8zDy66lr6KoopJ7+Vmf9adS3SQGjW4zEi9qL4UvCOzPxJ1fXIUJckqTa8/C5JUk0Y6pIk1YShLklSTRjqkiTVhKEuSVJNGOqSJNWEoS5JUk0Y6pIk1cT/B8GKR4LQ1XPzAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 576x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "rep_strand_df = gene_table_ml[['rep_strand', 'fpkm']].dropna()\n",
    "rep_strand_df['log_fpkm'] = np.log(rep_strand_df['fpkm'] + ETA)\n",
    "\n",
    "# get an actual significance comparison here\n",
    "leading = rep_strand_df[rep_strand_df['rep_strand'] == 'leading']['log_fpkm']\n",
    "lagging = rep_strand_df[rep_strand_df['rep_strand'] == 'lagging']['log_fpkm']\n",
    "\n",
    "t_stat, p_value = ttest_ind(leading, lagging, equal_var=False)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(8, 5))\n",
    "\n",
    "sns.boxplot(x='rep_strand', y='log_fpkm',\n",
    "            data=rep_strand_df, fliersize=0, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels([f\"{lab.get_text()}\\n(n={rep_strand_df[rep_strand_df['rep_strand'] == lab.get_text()].shape[0]})\"\n",
    "                       for lab in ax.get_xticklabels()])    \n",
    "ax.set_ylabel('Log FPKM', fontsize=13)\n",
    "ax.text(0.025, 0.05, f'P = {p_value:.3f}', fontsize=13, transform=ax.transAxes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Order in TU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def strongest_tu_for_gene(gene_row):\n",
    "    \n",
    "    # get TUs associated with gene\n",
    "    tus_for_gene = [tu_id for tu_id, gene_ids in tu_to_genes.items() if gene_row.name in gene_ids]\n",
    "    tus_for_gene_df = m_florum.tu_table.loc[tus_for_gene]\n",
    "    if tus_for_gene_df.empty:\n",
    "        return None\n",
    "    \n",
    "    # get the strongest TU specifically; previous work indicates that this helps\n",
    "    def tss_to_rspm(tss):\n",
    "        tss_df = m_florum.tss_table[m_florum.tss_table['tss'] == tss]\n",
    "        if tss_df.empty:\n",
    "            return None\n",
    "        else:\n",
    "            return tss_df.iloc[0, :]['rspm']\n",
    "    tus_for_gene_df['rspm'] = tus_for_gene_df['tss'].apply(tss_to_rspm)\n",
    "    gene_strong_tu = tus_for_gene_df.sort_values(by='rspm', ascending=False).iloc[0, :]\n",
    "    \n",
    "    return gene_strong_tu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def order_in_tu(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    all_tu_gene_ids = [gid for gid in gene_strong_tu.gene_names.split(';') if 'peg' in gid]\n",
    "    gene_df = m_florum.gene_table[m_florum.gene_table.index.isin(all_tu_gene_ids)]\n",
    "    if gene_strong_tu.strand == 1:\n",
    "        gene_starts = gene_df['left']\n",
    "        sorted_genes, _ = list(zip(*sorted(zip(all_tu_gene_ids, gene_starts),\n",
    "                                          key=lambda tup: tup[1])))\n",
    "    else:\n",
    "        gene_starts = gene_df['right']\n",
    "        # reverse the sort on the reverse strand so the highest gene start is first (that's closest to TSS)\n",
    "        sorted_genes, _ = list(zip(*sorted(zip(all_tu_gene_ids, gene_starts),\n",
    "                                          key=lambda tup: tup[1], reverse=True)))\n",
    "\n",
    "    gene_order_dict = {gene_id: i+1 for i, gene_id in enumerate(sorted_genes)}\n",
    "\n",
    "    return gene_order_dict[gene_row.name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'tu_to_genes' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-17775b78cac5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgene_table_ml\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tu_order'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgene_table_ml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0morder_in_tu\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mgene_table_ml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/strainware/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, axis, raw, result_type, args, **kwds)\u001b[0m\n\u001b[1;32m   7546\u001b[0m             \u001b[0mkwds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7547\u001b[0m         )\n\u001b[0;32m-> 7548\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7549\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7550\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapplymap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;34m\"DataFrame\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/strainware/lib/python3.8/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mget_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    178\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    179\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 180\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    181\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    182\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_empty_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/strainware/lib/python3.8/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    269\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    270\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 271\u001b[0;31m         \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    272\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    273\u001b[0m         \u001b[0;31m# wrap results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/strainware/lib/python3.8/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    298\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseries_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m                     \u001b[0;31m# ignore SettingWithCopy here in case the user mutates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 300\u001b[0;31m                     \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    301\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m                         \u001b[0;31m# If we have a view on v, we need to make a copy because\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-13-e1c77b4f8a1d>\u001b[0m in \u001b[0;36morder_in_tu\u001b[0;34m(gene_row)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0morder_in_tu\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_row\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mgene_strong_tu\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstrongest_tu_for_gene\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_row\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mgene_strong_tu\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-12-235bf432e53b>\u001b[0m in \u001b[0;36mstrongest_tu_for_gene\u001b[0;34m(gene_row)\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;31m# get TUs associated with gene\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mtus_for_gene\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mtu_id\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mtu_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgene_ids\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtu_to_genes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mgene_row\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgene_ids\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mtus_for_gene_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mm_florum\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtu_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtus_for_gene\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mtus_for_gene_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'tu_to_genes' is not defined"
     ]
    }
   ],
   "source": [
    "gene_table_ml['tu_order'] = gene_table_ml.apply(order_in_tu, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tu_rel_exps = {}\n",
    "\n",
    "for tu_row in m_florum.tu_table.itertuples():\n",
    "    tu_genes = tu_row.gene_names.split(';')\n",
    "    tu_gene_df = m_florum.gene_table[m_florum.gene_table.index.isin(tu_genes)]\n",
    "    if tu_gene_df.shape[0] > 1:\n",
    "        \n",
    "        if tu_row.strand == 1:\n",
    "            sorted_gene_df = tu_gene_df.sort_values(by='left')\n",
    "        else:\n",
    "            sorted_gene_df = tu_gene_df.sort_values(by='right', ascending=False)\n",
    "        \n",
    "        tu_gene_exps_rel = np.log(np.array(sorted_gene_df['mean_fpkm']) + ETA) / \\\n",
    "            np.log(sorted_gene_df['mean_fpkm'][0] + ETA)\n",
    "        \n",
    "        for i, rel_exp in enumerate(tu_gene_exps_rel):\n",
    "            if i in tu_rel_exps:\n",
    "                tu_rel_exps[i].append(rel_exp)\n",
    "            else:\n",
    "                tu_rel_exps[i] = [rel_exp]\n",
    "                \n",
    "tu_rel_exps = {k: v for k, v in tu_rel_exps.items() if len(v) > 10}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(15, 5))\n",
    "\n",
    "ax.boxplot(list(tu_rel_exps.values()))\n",
    "ax.set_xticklabels([f'{i+1}\\n(n={len(vals)})' for i, vals in enumerate(tu_rel_exps.values())])\n",
    "ax.set_xlabel('Order in TU', fontsize=13)\n",
    "ax.set_ylabel('log[FPKM] Relative to 1st Gene', fontsize=13)\n",
    "ax.tick_params(axis='both', labelsize=12)\n",
    "ax.set_ylim(bottom=0, top=3)\n",
    "for i, vals in enumerate(tu_rel_exps.values()):\n",
    "    med = np.median(vals)\n",
    "    ax.text(i+1, med+0.05, f'{med:.2f}x', ha='center')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This comparison would be more informative at a TU-specific level (i.e. relative expression within multi-gene TUs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TSS to Translation Start Distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tss_to_gene_start(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "        \n",
    "    strand = gene_strong_tu.strand\n",
    "    if strand == 1:\n",
    "        tu_start = gene_strong_tu.left\n",
    "        gene_start = gene_row.left\n",
    "    else:\n",
    "        tu_start = gene_strong_tu.right\n",
    "        gene_start = gene_row.right\n",
    "\n",
    "    return abs(tu_start - gene_start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['tss_dist'] = gene_table_ml.apply(tss_to_gene_start, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_tss_dists = []\n",
    "all_rel_exps_tss = []\n",
    "\n",
    "for tu_row in m_florum.tu_table.itertuples():\n",
    "    tu_genes = tu_row.gene_names.split(';')\n",
    "    tu_gene_df = m_florum.gene_table[m_florum.gene_table.index.isin(tu_genes)]\n",
    "    if tu_gene_df.shape[0] > 1:\n",
    "        \n",
    "        if tu_row.strand == 1:\n",
    "            sorted_gene_df = tu_gene_df.sort_values(by='left')\n",
    "            tss_dists = sorted_gene_df['left'].values - tu_row.tss\n",
    "        else:\n",
    "            sorted_gene_df = tu_gene_df.sort_values(by='right', ascending=False)\n",
    "            tss_dists = tu_row.tss - sorted_gene_df['right']\n",
    "        \n",
    "        tu_gene_exps_rel = np.log(np.array(sorted_gene_df['mean_fpkm']) + ETA) / \\\n",
    "            np.log(sorted_gene_df['mean_fpkm'][0] + ETA)\n",
    "        \n",
    "        all_tss_dists += list(tss_dists[1:])\n",
    "        all_rel_exps_tss += list(tu_gene_exps_rel[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(15, 5))\n",
    "\n",
    "ax.scatter(all_tss_dists, all_rel_exps_tss)\n",
    "ax.set_xlabel('Distance to TSS (bp)', fontsize=13)\n",
    "ax.set_ylabel('log[FPKM] Relative to 1st Gene in TU', fontsize=13)\n",
    "ax.tick_params(axis='both', labelsize=12)\n",
    "ax.set_ylim(0, 3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This comparison would be more informative at a TU-specific level (i.e. relative expression within multi-gene TUs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distance to Origin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dist_to_origin(gene_row):\n",
    "    l, r = gene_row.left, gene_row.right\n",
    "    seq_max = len(m_florum.sequence)\n",
    "    return np.min([\n",
    "        genome_point_to_point(m_florum.origin[0], l, seq_max),\n",
    "        genome_point_to_point(m_florum.origin[0], r, seq_max),\n",
    "        genome_point_to_point(m_florum.origin[1], l, seq_max),\n",
    "        genome_point_to_point(m_florum.origin[1], r, seq_max)\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['origin_dist'] = gene_table_ml.apply(dist_to_origin, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "origin_dist_df = gene_table_ml[['origin_dist', 'mean_fpkm', 'tu_order', 'rep_strand']].dropna()\n",
    "origin_dist_df['log_fpkm'] = np.log(origin_dist_df['mean_fpkm'] + ETA)\n",
    "origin_dist_df = origin_dist_df[origin_dist_df['tu_order'] == 1]\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.scatterplot(x='origin_dist', y='log_fpkm', hue='rep_strand', data=origin_dist_df, ax=ax,\n",
    "               hue_order=['lagging', 'leading'])\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('ORI to Gene Start (bp)', fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5'-UTR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def utr_len(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    all_tu_gene_ids = [gid for gid in gene_strong_tu.gene_names.split(';') if 'peg' in gid]\n",
    "    gene_df = m_florum.gene_table[m_florum.gene_table.index.isin(all_tu_gene_ids)]\n",
    "\n",
    "    if gene_strong_tu.strand == 1:\n",
    "        gene_starts = gene_df['left']\n",
    "        _, sorted_starts = list(zip(*sorted(zip(all_tu_gene_ids, gene_starts),\n",
    "                                          key=lambda tup: tup[1])))\n",
    "        utr_len = sorted_starts[0] - gene_strong_tu.tss\n",
    "    else:\n",
    "        gene_starts = gene_df['right']\n",
    "        # reverse the sort on the reverse strand so the highest gene start is first (that's closest to TSS)\n",
    "        _, sorted_starts = list(zip(*sorted(zip(all_tu_gene_ids, gene_starts),\n",
    "                                          key=lambda tup: tup[1], reverse=True)))\n",
    "        utr_len = gene_strong_tu.tss - sorted_starts[0]\n",
    "\n",
    "    return utr_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['utr_len'] = gene_table_ml.apply(utr_len, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "utr_len_df = gene_table_ml[['utr_len', 'mean_fpkm']].dropna()\n",
    "utr_len_df['log_fpkm'] = np.log(utr_len_df['mean_fpkm'] + ETA)\n",
    "utr_len_df = utr_len_df[utr_len_df['utr_len'] > 0]\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.scatterplot(x='utr_len', y='log_fpkm', data=utr_len_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel(\"5' UTR Len (bp) \", fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GC Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def utr_gc(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    all_tu_gene_ids = [gid for gid in gene_strong_tu.gene_names.split(';') if 'peg' in gid]\n",
    "    gene_df = m_florum.gene_table[m_florum.gene_table.index.isin(all_tu_gene_ids)]\n",
    "    if gene_strong_tu.strand == 1:\n",
    "        gene_starts = gene_df['left']\n",
    "        _, sorted_starts = list(zip(*sorted(zip(all_tu_gene_ids, gene_starts),\n",
    "                                          key=lambda tup: tup[1])))\n",
    "        utr_left, utr_right = gene_strong_tu.tss, sorted_starts[0]\n",
    "    else:\n",
    "        gene_starts = gene_df['right']\n",
    "        # reverse the sort on the reverse strand so the highest gene start is first (that's closest to TSS)\n",
    "        _, sorted_starts = list(zip(*sorted(zip(all_tu_gene_ids, gene_starts),\n",
    "                                          key=lambda tup: tup[1], reverse=True)))\n",
    "\n",
    "        utr_left, utr_right = sorted_starts[0], gene_strong_tu.tss\n",
    "        \n",
    "    # handle weird case with TU_052 and peg.78...double check data later\n",
    "    if utr_left > utr_right:\n",
    "        return None\n",
    "    else:\n",
    "        return GC(m_florum.get_sequence(utr_left, utr_right, gene_strong_tu.strand))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['utr_gc'] = gene_table_ml.apply(utr_gc, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "utr_gc_df = gene_table_ml[['utr_gc', 'mean_fpkm']].dropna()\n",
    "utr_gc_df['log_fpkm'] = np.log(utr_gc_df['mean_fpkm'] + ETA)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.scatterplot(x='utr_gc', y='log_fpkm', data=utr_gc_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel(\"5' UTR GC %\", fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transcript Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tu_len(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    else:\n",
    "        return gene_strong_tu.right - gene_strong_tu.left"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['tu_len'] = gene_table_ml.apply(tu_len, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tu_len_df = gene_table_ml[['tu_len', 'mean_fpkm']].dropna()\n",
    "tu_len_df['log_fpkm'] = np.log(tu_len_df['mean_fpkm'] + ETA)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.scatterplot(x='tu_len', y='log_fpkm', data=tu_len_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel(\"TU Len (bp) \", fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TU GC Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tu_gc(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    else:\n",
    "        return GC(m_florum.get_feature_sequence(gene_strong_tu))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['tu_gc'] = gene_table_ml.apply(tu_gc, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tu_gc_df = gene_table_ml[['tu_gc', 'mean_fpkm']].dropna()\n",
    "tu_gc_df['log_fpkm'] = np.log(tu_gc_df['mean_fpkm'] + ETA)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.scatterplot(x='tu_gc', y='log_fpkm', data=tu_gc_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel(\"TU GC %\", fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genome Location\n",
    "\n",
    "Meant to be a crude approximation of the chromosome interacting domains discovered in M. pneumo (Cell Systems 2019)\n",
    "\n",
    "They discovered 44 CIDs; let's replicate that by just splitting the genome evenly into 44 pieces (yes very crude)\n",
    "\n",
    "Also to capture the circularity, do this in the sin of the radians"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N_CID_BINS = 44\n",
    "\n",
    "def cid(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if gene_strong_tu.strand == 1:\n",
    "        tu_start = gene_strong_tu.left\n",
    "    else:\n",
    "        tu_start = gene_strong_tu.right\n",
    "        \n",
    "    cid = np.floor(tu_start / len(m_florum.sequence) * N_CID_BINS)\n",
    "\n",
    "    return cid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['cid'] = gene_table_ml.apply(cid, axis=1)\n",
    "gene_table_ml['cid_sin'] = gene_table_ml['cid'].apply(\n",
    "    lambda cid: np.round(np.sin(2 * np.pi * cid / N_CID_BINS), 4))\n",
    "gene_table_ml['cid_cos'] = gene_table_ml['cid'].apply(\n",
    "    lambda cid: np.round(np.cos(2 * np.pi * cid / N_CID_BINS), 4))\n",
    "gene_table_ml['cid_cat'] = gene_table_ml['cid'].apply(str)\n",
    "gene_table_ml = gene_table_ml.drop(columns=['cid'])\n",
    "gene_table_ml.plot.scatter('cid_sin','cid_cos').set_aspect('equal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cid_df = gene_table_ml[['cid_sin', 'cid_cos', 'cid_cat', 'mean_fpkm']].dropna()\n",
    "cid_df['log_fpkm'] = np.log(cid_df['mean_fpkm'] + ETA)\n",
    "cid_df['sincos'] = cid_df.apply(lambda row: str(row.cid_sin) + str(row.cid_cos), axis=1)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(15, 7))\n",
    "sns.boxplot(x='cid_cat', y='log_fpkm', data=cid_df, ax=ax, dodge=True)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.tick_params(axis='x', rotation=90)\n",
    "ax.set_xlabel(\"CID\", fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ANOVA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_groups = []\n",
    "for sincos, sincos_df in cid_df.groupby('sincos'):\n",
    "    exp_groups.append(sincos_df['log_fpkm'])\n",
    "    \n",
    "f, p = f_oneway(*exp_groups)\n",
    "\n",
    "print(f'P value of ANOVA: {p:.2E}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### -10 Box\n",
    "\n",
    "Let's try something here; can we make these into categories?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -10 Box Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def box_10_seq(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if not pd.isna(gene_strong_tu.box_10_left):\n",
    "        box_10_seq = str(m_florum.get_sequence(gene_strong_tu.box_10_left,\n",
    "                                           gene_strong_tu.box_10_right,\n",
    "                                           gene_strong_tu.strand))\n",
    "    else:\n",
    "        box_10_seq = None\n",
    "\n",
    "    return box_10_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['box_10_seq'] = gene_table_ml.apply(box_10_seq, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "box_df = gene_table_ml[['box_10_seq', 'mean_fpkm']].dropna()\n",
    "box_df['log_fpkm'] = np.log(box_df['mean_fpkm'] + ETA)\n",
    "\n",
    "box_10_dfs = []\n",
    "\n",
    "for _, b10seq_df in box_df.groupby('box_10_seq'):\n",
    "    if b10seq_df.shape[0] > 5:\n",
    "        box_10_dfs.append(b10seq_df)\n",
    "        \n",
    "box_10_dfs = sorted(box_10_dfs, key=lambda df: df['log_fpkm'].median(), reverse=True)\n",
    "box_10_df = pd.concat(box_10_dfs)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.boxplot(x='box_10_seq', y='log_fpkm', data=box_10_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='box_10_seq', y='log_fpkm', data=box_10_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_xticklabels([f\"{lab.get_text()}\\n(n={box_10_df[box_10_df['box_10_seq'] == lab.get_text()].shape[0]})\"\n",
    "                   for lab in ax.get_xticklabels()])\n",
    "ax.set_title(f\"-10 Box Sequence (n={len(box_df['box_10_seq'].unique())}; 11 with >5 genes shown)\", fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -10 Box Distance to TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def box_10_dist(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if not pd.isna(gene_strong_tu.box_10_left):\n",
    "        box_10_tss_dist = int(min(\n",
    "            abs(gene_strong_tu.box_10_left - gene_strong_tu.tss),\n",
    "            abs(gene_strong_tu.box_10_right - gene_strong_tu.tss)\n",
    "        ))\n",
    "    else:\n",
    "        box_10_tss_dist = None\n",
    "\n",
    "    return box_10_tss_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['box_10_dist'] = gene_table_ml.apply(box_10_dist, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "box_dist_df = gene_table_ml[['box_10_dist', 'mean_fpkm']].dropna()\n",
    "box_dist_df['log_fpkm'] = np.log(box_dist_df['mean_fpkm'] + ETA)\n",
    "box_dist_df = box_dist_df[box_dist_df['box_10_dist'] < 9]\n",
    "\n",
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.boxplot(x='box_10_dist', y='log_fpkm', data=box_dist_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='box_10_dist', y='log_fpkm', data=box_dist_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('-10 Box Distance to TSS (bp)', fontsize=14)\n",
    "ax.set_xticklabels([\n",
    "    f\"{int(float(lab.get_text()))}\\n (n={box_dist_df[box_dist_df['box_10_dist'] == float(lab.get_text())].shape[0]})\"\n",
    "    for lab in ax.get_xticklabels()\n",
    "])\n",
    "ax.set_ylabel('log[FPKM]', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -10 Box Motif Strength"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's recreate the -10 box motif using all known TUs\n",
    "box_10_seqs = []\n",
    "\n",
    "for tu_row in m_florum.tu_table.itertuples():\n",
    "    if not pd.isna(tu_row.box_10_left):\n",
    "        box_10_seq = m_florum.get_sequence(tu_row.box_10_left,\n",
    "                                           tu_row.box_10_right,\n",
    "                                           tu_row.strand)\n",
    "        box_10_seqs.append(box_10_seq)\n",
    "        \n",
    "box_10_motif = create_motif(box_10_seqs)\n",
    "box_10_pssm = box_10_motif.pssm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('temp.fa', 'w') as file_motif:\n",
    "    for i, seq in enumerate(box_10_seqs):\n",
    "        file_motif.write(f'>{i}\\n{str(seq)}\\n')\n",
    "        \n",
    "with open(\"temp.fa\") as file_motif:\n",
    "    seqs = read_seq_data(file_motif, alphabet=unambiguous_dna_alphabet)\n",
    "    \n",
    "os.remove('temp.fa')\n",
    "\n",
    "logodata = LogoData.from_seqs(seqs)\n",
    "logooptions = LogoOptions(\n",
    "    title='-10 Box Motif',\n",
    "    show_fineprint=False\n",
    ")\n",
    "logoformat = LogoFormat(logodata, logooptions)\n",
    "png = png_formatter(logodata, logoformat)\n",
    "Image(png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def box_10_score(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if not pd.isna(gene_strong_tu.box_10_left):\n",
    "        box_10_seq = m_florum.get_sequence(gene_strong_tu.box_10_left,\n",
    "                                          gene_strong_tu.box_10_right,\n",
    "                                          gene_strong_tu.strand)\n",
    "        box_10_score = score_motif_match(box_10_seq, box_10_pssm)\n",
    "    else:\n",
    "        box_10_score = None\n",
    "\n",
    "    return box_10_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['box_10_score'] = gene_table_ml.apply(box_10_score, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "box_10_score_df = gene_table_ml[['box_10_score', 'mean_fpkm']].dropna()\n",
    "box_10_score_df['log_fpkm'] = np.log(box_10_score_df['mean_fpkm'] + ETA)\n",
    "\n",
    "_, ax = plt.subplots()\n",
    "sns.scatterplot(x='box_10_score', y='log_fpkm', data=box_10_score_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('-10 Box Motif Score', fontsize=14)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extended -10 Box AT Content\n",
    "\n",
    "AT content of 3 bp upstream of -10 box"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ext_10_at(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if not pd.isna(gene_strong_tu.box_10_left):\n",
    "        if gene_strong_tu.strand == 1:\n",
    "            ext_10_seq = str(m_florum.get_sequence(gene_strong_tu.box_10_left - 3,\n",
    "                                           gene_strong_tu.box_10_left - 1,\n",
    "                                           gene_strong_tu.strand))\n",
    "        else:\n",
    "            ext_10_seq = str(m_florum.get_sequence(gene_strong_tu.box_10_right + 1,\n",
    "                                           gene_strong_tu.box_10_right + 3,\n",
    "                                           gene_strong_tu.strand))\n",
    "        ext_10_at = np.round((100 - GC(ext_10_seq)) / 100, decimals=2)\n",
    "    else:\n",
    "        ext_10_at = None\n",
    "\n",
    "    return ext_10_at"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['ext_10_at'] = gene_table_ml.apply(ext_10_at, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ext_10_at_df = gene_table_ml[['ext_10_at', 'mean_fpkm']].dropna()\n",
    "ext_10_at_df['log_fpkm'] = np.log(ext_10_at_df['mean_fpkm'] + ETA)\n",
    "\n",
    "_, ax = plt.subplots()\n",
    "sns.boxplot(x='ext_10_at', y='log_fpkm', data=ext_10_at_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('Extended -10 Box AT Content', fontsize=14)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=14)\n",
    "ax.set_xticklabels([\n",
    "    f\"{float(lab.get_text())}\\n (n={ext_10_at_df[ext_10_at_df['ext_10_at'] == float(lab.get_text())].shape[0]})\"\n",
    "    for lab in ax.get_xticklabels()\n",
    "])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CRE Sequence\n",
    "\n",
    "Let's try this again with the CRE (-4 to +2, core RNAP binding site)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cre_seq(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if not pd.isna(gene_strong_tu.tss):\n",
    "        if gene_strong_tu.strand == 1:\n",
    "            cre_seq = str(m_florum.get_sequence(gene_strong_tu.tss - 4,\n",
    "                                               gene_strong_tu.tss + 1,\n",
    "                                               gene_strong_tu.strand))\n",
    "        else:\n",
    "            cre_seq = str(m_florum.get_sequence(gene_strong_tu.tss - 1,\n",
    "                                               gene_strong_tu.tss + 4,\n",
    "                                               gene_strong_tu.strand))\n",
    "    else:\n",
    "        cre_seq = None\n",
    "\n",
    "    return cre_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['cre_seq'] = gene_table_ml.apply(cre_seq, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df = gene_table_ml[['cre_seq', 'mean_fpkm']].dropna()\n",
    "\n",
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.boxplot(x='cre_seq', y='mean_fpkm', data=cre_df, fliersize=0, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.tick_params(axis='x', rotation=90)\n",
    "ax.set_ylabel('Mean FPKM', fontsize=13)\n",
    "ax.set_title(f\"CRE Sequence (n={len(cre_df['cre_seq'].unique())})\", fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TSS Nucleotide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tss_base(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if not pd.isna(gene_strong_tu.tss):\n",
    "        tss_base = str(m_florum.get_sequence(\n",
    "            gene_strong_tu.tss,\n",
    "            gene_strong_tu.tss,\n",
    "            gene_strong_tu.strand)\n",
    "                      )\n",
    "    else:\n",
    "        tss_base = None\n",
    "\n",
    "    return tss_base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['tss_base'] = gene_table_ml.apply(tss_base, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_base_df = gene_table_ml[['tss_base', 'mean_fpkm']].dropna()\n",
    "tss_base_df['log_fpkm'] = np.log(tss_base_df['mean_fpkm'] + ETA)\n",
    "\n",
    "_, ax = plt.subplots()\n",
    "sns.boxplot(x='tss_base', y='log_fpkm', data=tss_base_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='tss_base', y='log_fpkm', data=tss_base_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_title(f\"TSS Base\", fontsize=14)\n",
    "ax.set_xticklabels([\n",
    "    f\"{lab.get_text()}\\n (n={tss_base_df[tss_base_df['tss_base'] == lab.get_text()].shape[0]})\"\n",
    "    for lab in ax.get_xticklabels()\n",
    "])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TSS Dinucleotide (-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tss_dinuc(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if not pd.isna(gene_strong_tu.tss):\n",
    "        if gene_strong_tu.strand == 1:\n",
    "            tss_dinuc = str(m_florum.get_sequence(\n",
    "                gene_strong_tu.tss-1,\n",
    "                gene_strong_tu.tss,\n",
    "                gene_strong_tu.strand)\n",
    "                          )\n",
    "        else:\n",
    "            tss_dinuc = str(m_florum.get_sequence(\n",
    "                gene_strong_tu.tss,\n",
    "                gene_strong_tu.tss+1,\n",
    "                gene_strong_tu.strand)\n",
    "                          )\n",
    "    else:\n",
    "        tss_dinuc = None\n",
    "\n",
    "    return tss_dinuc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['tss_dinuc'] = gene_table_ml.apply(tss_dinuc, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_dinuc_df = gene_table_ml[['tss_dinuc', 'mean_fpkm']].dropna()\n",
    "tss_dinuc_df['log_fpkm'] = np.log(tss_dinuc_df['mean_fpkm'] + ETA)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 6))\n",
    "sns.boxplot(x='tss_dinuc', y='log_fpkm', data=tss_dinuc_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='tss_dinuc', y='log_fpkm', data=tss_dinuc_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_title(f\"TSS Dinucleotide (-1/+1)\", fontsize=14)\n",
    "ax.set_xticklabels([\n",
    "    f\"{lab.get_text()}\\n (n={tss_dinuc_df[tss_dinuc_df['tss_dinuc'] == lab.get_text()].shape[0]})\"\n",
    "    for lab in ax.get_xticklabels()\n",
    "]);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TSS Dinucleotide (+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tss_dinuc_plus(gene_row):\n",
    "\n",
    "    gene_strong_tu = strongest_tu_for_gene(gene_row)\n",
    "    if gene_strong_tu is None:\n",
    "        return None\n",
    "    \n",
    "    if not pd.isna(gene_strong_tu.tss):\n",
    "        if gene_strong_tu.strand == 1:\n",
    "            tss_dinuc = str(m_florum.get_sequence(\n",
    "                gene_strong_tu.tss,\n",
    "                gene_strong_tu.tss+1,\n",
    "                gene_strong_tu.strand)\n",
    "                          )\n",
    "        else:\n",
    "            tss_dinuc = str(m_florum.get_sequence(\n",
    "                gene_strong_tu.tss-1,\n",
    "                gene_strong_tu.tss,\n",
    "                gene_strong_tu.strand)\n",
    "                          )\n",
    "    else:\n",
    "        tss_dinuc = None\n",
    "\n",
    "    return tss_dinuc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_ml['tss_dinuc_plus'] = gene_table_ml.apply(tss_dinuc_plus, axis=1)\n",
    "gene_table_ml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_dinucp_df = gene_table_ml[['tss_dinuc_plus', 'mean_fpkm']].dropna()\n",
    "tss_dinucp_df['log_fpkm'] = np.log(tss_dinucp_df['mean_fpkm'] + ETA)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 6))\n",
    "sns.boxplot(x='tss_dinuc_plus', y='log_fpkm', data=tss_dinucp_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='tss_dinuc_plus', y='log_fpkm', data=tss_dinucp_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_title(f\"TSS Dinucleotide (+1/+2)\", fontsize=14)\n",
    "ax.set_xticklabels([\n",
    "    f\"{lab.get_text()}\\n (n={tss_dinucp_df[tss_dinucp_df['tss_dinuc_plus'] == lab.get_text()].shape[0]})\"\n",
    "    for lab in ax.get_xticklabels()\n",
    "]);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Diverging Promoters\n",
    "\n",
    "Find and visualize instances where TUs are \"back-to-back\" (i.e. their TSSes are next to each other, going opposite ways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_tu_table = m_florum.tu_table.copy().sort_values(by='tss')\n",
    "\n",
    "i = 0\n",
    "\n",
    "ratios = []\n",
    "rand_ratios = []\n",
    "\n",
    "while i < sorted_tu_table.shape[0] - 1:\n",
    "    \n",
    "    tu_row_a = sorted_tu_table.iloc[i, :]\n",
    "    tu_row_b = sorted_tu_table.iloc[i+1, :]\n",
    "    \n",
    "    # this is the only way to have back-to-back, if we go along the positive strand\n",
    "    if tu_row_a.strand == -1 and tu_row_b.strand == 1:\n",
    "\n",
    "        tu_row_a_genes = tu_row_a.gene_names.split(';')\n",
    "        tu_a_gene_df = m_florum.gene_table[m_florum.gene_table.index.isin(tu_row_a_genes)]\n",
    "        tu_row_b_genes = tu_row_b.gene_names.split(';')\n",
    "        tu_b_gene_df = m_florum.gene_table[m_florum.gene_table.index.isin(tu_row_b_genes)]\n",
    "        \n",
    "        if not tu_a_gene_df.empty and not tu_b_gene_df.empty:\n",
    "            tu_a_exp = np.log(tu_a_gene_df.sort_values(by='right', ascending=False).iloc[0, :]['mean_fpkm'] + ETA)\n",
    "            tu_b_exp = np.log(tu_b_gene_df.sort_values(by='left').iloc[0, :]['mean_fpkm'] + ETA)\n",
    "        \n",
    "            ratio = min(tu_a_exp, tu_b_exp) / max(tu_a_exp, tu_b_exp)\n",
    "            ratios.append(ratio)\n",
    "            rand_exps = np.log(np.random.choice(m_florum.gene_table['mean_fpkm'], size=2) + ETA)\n",
    "            rand_ratio = min(rand_exps) / max(rand_exps)\n",
    "            rand_ratios.append(rand_ratio)\n",
    "            \n",
    "            i += 2\n",
    "        else:\n",
    "            i += 1\n",
    "    else:\n",
    "        i += 1\n",
    "        \n",
    "diverge_df = pd.DataFrame(data={'log[FPKM] Ratio': ratios + rand_ratios,\n",
    "                               'source': ['diverging'] * len(ratios) + ['random'] * len(rand_ratios)})\n",
    "\n",
    "_, ax = plt.subplots()\n",
    "\n",
    "t_stat, p_value = ttest_ind(ratios, rand_ratios, equal_var=False)\n",
    "\n",
    "sns.boxplot(x='source', y='log[FPKM] Ratio', data=diverge_df, ax=ax)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM] Ratio', fontsize=13)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_title('Effect of Diverging TUs on 1st Gene Expression', fontsize=14)\n",
    "ax.text(0.8, 0.2, f'P={p_value:.3f}', transform=ax.transAxes, fontsize=13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots()\n",
    "\n",
    "t_stat, p_value = ttest_ind(ratios, rand_ratios, equal_var=False)\n",
    "\n",
    "sns.boxplot(x='source', y='log[FPKM] Ratio', data=diverge_df, ax=ax)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM] Ratio', fontsize=13)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_title('Effect of Diverging TUs on 1st Gene Expression', fontsize=14)\n",
    "ax.text(0.8, 0.2, f'P={p_value:.3f}', transform=ax.transAxes, fontsize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finalize Feature Matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = gene_table_ml.drop(columns=['left', 'right', 'strand', 'KO', 'category', 'subcategory', 'subsubcategory',\n",
    "                               'product_name', 'name', 'sd_fpkm']).dropna()\n",
    "y = X['mean_fpkm']\n",
    "X = X.drop(columns=['mean_fpkm'])\n",
    "\n",
    "X_go_all = pd.get_dummies(X.copy())\n",
    "X_no_cre = pd.get_dummies(X.copy().drop(columns=['cre_seq']))\n",
    "X_no_box10 = pd.get_dummies(X.copy().drop(columns=['box_10_seq']))\n",
    "X_no_seqs = pd.get_dummies(X.copy().drop(columns=['cre_seq', 'box_10_seq']))\n",
    "\n",
    "Xy_to_try_go = {\n",
    "    'all': (X_go_all, y),\n",
    "    'no_cre': (X_no_cre, y),\n",
    "    'no_box10': (X_no_box10, y),\n",
    "    'no_seqs': (X_no_seqs, y)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Aside Lockbox Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Xy_train_to_use_go = {}\n",
    "Xy_lockbox_to_use_go = {}\n",
    "\n",
    "for name, (X, y) in Xy_to_try_go.items():\n",
    "    if X.shape[0] > 500:\n",
    "        test_size = 0.1\n",
    "    else:\n",
    "        test_size = 0.2\n",
    "    X_train, X_lockbox, y_train, y_lockbox = train_test_split(X, y, test_size=test_size, random_state=42)\n",
    "    Xy_train_to_use_go[name] = (X_train, y_train)\n",
    "    Xy_lockbox_to_use_go[name] = (X_lockbox, y_lockbox)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initial Model Profiling - Genome Organization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "initial_profile_df_go = pd.DataFrame(\n",
    "    columns=['model', 'xy', 'train_r2', 'val_r2']\n",
    ")\n",
    "\n",
    "N_CV = 5\n",
    "\n",
    "for model_name, model in models_to_try.items():\n",
    "    for Xy_name, (X_train_raw, y_train_raw) in Xy_train_to_use_go.items():\n",
    "        \n",
    "        X_train_norm = StandardScaler().fit_transform(X_train_raw)\n",
    "        y_train_norm = StandardScaler().fit_transform(y_train_raw.values.reshape(-1, 1))\n",
    "        y_train_log = np.log(y_train_raw + ETA)\n",
    "        y_train_norm_log = StandardScaler().fit_transform(y_train_log.values.reshape(-1, 1))\n",
    "        \n",
    "        X_list = [('X_raw', X_train_raw), ('X_norm', X_train_norm)]\n",
    "        y_list = [('y_raw', y_train_raw), ('y_norm', y_train_norm),\n",
    "                 ('y_log', y_train_log), ('y_norm_log', y_train_norm_log)]\n",
    "        \n",
    "        Xy_final = {}\n",
    "        for (X_lab, X_mat), (y_lab, y_mat) in product(X_list, y_list):\n",
    "            Xy_lab_final = f'{Xy_name}__{X_lab}__{y_lab}'\n",
    "            Xy_final[Xy_lab_final] = (X_mat, y_mat)\n",
    "            \n",
    "        for Xy_full_name, (X_to_use, y_to_use) in Xy_final.items():\n",
    "\n",
    "            print(f'{model_name}: {Xy_full_name}')\n",
    "\n",
    "            cv_result = cross_validate(\n",
    "                model,\n",
    "                X_to_use,\n",
    "                y=y_to_use,\n",
    "                cv=N_CV,\n",
    "                scoring='r2',\n",
    "                return_train_score=True,\n",
    "                verbose=1,\n",
    "                n_jobs=N_CV\n",
    "            )\n",
    "\n",
    "            sub_result_df = pd.DataFrame(\n",
    "                data={\n",
    "                    'model': [model_name] * N_CV,\n",
    "                    'xy': [Xy_full_name] * N_CV,\n",
    "                    'train_r2': cv_result['train_score'],\n",
    "                    'val_r2': cv_result['test_score']\n",
    "                }\n",
    "            )\n",
    "\n",
    "            initial_profile_df_go = initial_profile_df_go.append(sub_result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='model', y='train_r2', data=initial_profile_df_go, hue='xy', dodge=True, ax=ax)\n",
    "ax.legend(loc='center left', bbox_to_anchor=(1.05, 0.5))\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Training R-Squared')\n",
    "ax.set_ylim(bottom=0, top=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "to_plot_df = initial_profile_df_go[initial_profile_df_go['model'] == 'RF']\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='model', y='val_r2', data=initial_profile_df_go, hue='xy', dodge=True, ax=ax)\n",
    "ax.legend(loc='center left', bbox_to_anchor=(1.05, 0.5))\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Validation R-Squared')\n",
    "ax.set_ylim(0, 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Importances - Genome Organization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_fi_go, y_fi_go = Xy_train_to_use_go['no_cre']\n",
    "\n",
    "X_fi_go = pd.DataFrame(StandardScaler().fit_transform(X_fi_go), columns=X_fi_go.columns, index=X_fi_go.index)\n",
    "y_fi_go = np.log(y_fi_go + ETA)\n",
    "\n",
    "X_fi_go_train, X_fi_go_val, y_fi_go_train, y_fi_go_val = train_test_split(\n",
    "    X_fi_go, y_fi_go, test_size=0.15, random_state=35)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_fi_go = models_to_try['RF'].fit(X_fi_go_train, y_fi_go_train)\n",
    "explainer = shap.TreeExplainer(rf_fi_go)\n",
    "shap_values = explainer.shap_values(X_fi_go_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shap.summary_plot(shap_values, X_fi_go_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hyperoptimization - Genome Organization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_go_hyperopt, y_go_hyperopt = Xy_train_to_use_go['no_cre']\n",
    "\n",
    "X_go_hyperopt = pd.DataFrame(\n",
    "    StandardScaler().fit_transform(X_go_hyperopt), columns=X_go_hyperopt.columns, index=X_go_hyperopt.index)\n",
    "y_go_hyperopt = np.log(y_go_hyperopt + ETA)\n",
    "\n",
    "hyperopt_XY_go = {\n",
    "    'no_cre': (X_go_hyperopt, y_go_hyperopt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperopt_models = {\n",
    "    'RF': {\n",
    "        'model': RandomForestRegressor(),\n",
    "        'param_dists': {\n",
    "            'n_estimators': [25, 50, 100, 250],\n",
    "            'max_depth': np.arange(2, 20)\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperopt_results = {}\n",
    "\n",
    "for model_name, model_hyperopt_dict in hyperopt_models.items():\n",
    "\n",
    "    hyperopt_results[model_name] = {}\n",
    "    \n",
    "    for X_name, (X, y) in hyperopt_XY_go.items():\n",
    "        print(f'{model_name}: {X_name}')\n",
    "\n",
    "        random_search_hyperopt = RandomizedSearchCV(\n",
    "            model_hyperopt_dict['model'],\n",
    "            model_hyperopt_dict['param_dists'],\n",
    "            n_iter=25,\n",
    "            scoring='r2',\n",
    "            n_jobs=4,\n",
    "            cv=5,\n",
    "            verbose=3,\n",
    "            return_train_score=True\n",
    "        )\n",
    "\n",
    "        random_search_hyperopt.fit(X, y)\n",
    "\n",
    "        hyperopt_results_df = pd.DataFrame(random_search_hyperopt.cv_results_).sort_values(\n",
    "            by='mean_test_score',\n",
    "            ascending=False\n",
    "        )\n",
    "        hyperopt_results[model_name][X_name] = hyperopt_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "hyperopt_results['RF']['no_cre'].iloc[:10, :][['params', 'mean_train_score', 'mean_test_score']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TATAAT Group Variance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TSS Base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_df = gene_table_ml[gene_table_ml['box_10_seq'] == 'TATAAT'].copy()\n",
    "tata_df['log_fpkm'] = np.log(tata_df['mean_fpkm'] + ETA)\n",
    "# tata_df = tata_df[tata_df['tu_order'] == 1]\n",
    "\n",
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.boxplot(x='tss_base', y='log_fpkm', data=tata_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='tss_base', y='log_fpkm', data=tata_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_xticklabels([f\"{lab.get_text()}\\n(n={tata_df[tata_df['tss_base'] == lab.get_text()].shape[0]})\"\n",
    "                   for lab in ax.get_xticklabels()])\n",
    "ax.set_title('TSS Bases within TATAAT -10 Box Group', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## High-Expressed TATA/TSS A Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tata_a_df = tata_df[tata_df['tss_base'] == 'A']\n",
    "tata_a_df['length'] = (tata_a_df['right'] - tata_a_df['left']).copy()\n",
    "\n",
    "tata_a_df[tata_a_df['log_fpkm'] > 8].sort_values(by='log_fpkm', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Many involved in central processes (central carbon metabolism, ribosome)\n",
    "- almost all on leading strand"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Low-Expressed TATA/TSS A Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tata_a_df[tata_a_df['log_fpkm'] < 4].sort_values(by='log_fpkm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Origin Distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.scatterplot(x='origin_dist', y='log_fpkm', data=tata_a_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('ORI to Gene Start (bp)', fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_title('ORI to Gene Start for TATAAT/TSS A Group', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TU GC Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.scatterplot(x='tu_gc', y='log_fpkm', data=tata_a_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('TU GC %', fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_title('TU GC Content for TATAAT/TSS A Group', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5'-UTR Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.scatterplot(x='utr_len', y='log_fpkm', data=tata_a_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel(\"5' UTR Len(bp)\", fontsize=13)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_title(\"5' UTR Length for TATAAT/TSS A Group\", fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TSS Dinucleotide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.boxplot(x='tss_dinuc', y='log_fpkm', data=tata_a_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='tss_dinuc', y='log_fpkm', data=tata_a_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_xticklabels([f\"{lab.get_text()}\\n(n={tata_a_df[tata_a_df['tss_dinuc'] == lab.get_text()].shape[0]})\"\n",
    "                   for lab in ax.get_xticklabels()])\n",
    "ax.set_title('TSS Dinucleotide (-1/+1) within TATAAT -10 Box Group', fontsize=14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.boxplot(x='tss_dinuc_plus', y='log_fpkm', data=tata_a_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='tss_dinuc_plus', y='log_fpkm', data=tata_a_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_xticklabels([f\"{lab.get_text()}\\n(n={tata_a_df[tata_a_df['tss_dinuc_plus'] == lab.get_text()].shape[0]})\"\n",
    "                   for lab in ax.get_xticklabels()])\n",
    "ax.set_title('TSS Dinucleotide (+1/+2) within TATAAT -10 Box Group', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genome Location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_a_df['cid'] = tata_a_df.apply(cid, axis=1)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(15, 6))\n",
    "sns.boxplot(x='cid', y='log_fpkm', data=tata_a_df, fliersize=0, ax=ax)\n",
    "sns.swarmplot(x='cid', y='log_fpkm', data=tata_a_df, ax=ax)\n",
    "ax.tick_params(axis='both', labelsize=13)\n",
    "ax.set_xlabel('')\n",
    "ax.tick_params(axis='x', rotation=90)\n",
    "ax.set_ylabel('log[FPKM]', fontsize=13)\n",
    "ax.set_xticklabels([f\"{lab.get_text()} (n={tata_a_df[tata_a_df['cid'] == lab.get_text()].shape[0]})\"\n",
    "                   for lab in ax.get_xticklabels()])\n",
    "ax.set_title('CID within TATAAT -10 Box Group', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modeling Expression within TATA/TSS A Group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initial Model Profiling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_no_cre_copy = Xy_to_try_go['no_cre'][0].copy()\n",
    "y_no_cre_copy = Xy_to_try_go['no_cre'][1].copy()\n",
    "\n",
    "X_tata_tss_a = X_no_cre_copy[\n",
    "    X_no_cre_copy.index.isin(tata_a_df.index)\n",
    "].drop(columns=[col for col in X_no_cre_copy.columns if ('tss_base' in col) or\n",
    "                (col == 'box_10_score') or ('box_10_seq' in col)])\n",
    "y_tata_tss_a = y_no_cre_copy[\n",
    "    X_no_cre_copy.index.isin(tata_a_df.index)\n",
    "]\n",
    "\n",
    "Xy_to_try_tata = {\n",
    "    'tata': (X_tata_tss_a, y_tata_tss_a)\n",
    "}\n",
    "\n",
    "Xy_train_to_use_tata = {}\n",
    "Xy_lockbox_to_use_tata = {}\n",
    "\n",
    "for name, (X, y) in Xy_to_try_tata.items():\n",
    "    if X.shape[0] > 500:\n",
    "        test_size = 0.1\n",
    "    else:\n",
    "        test_size = 0.2\n",
    "    X_train, X_lockbox, y_train, y_lockbox = train_test_split(X, y, test_size=test_size, random_state=42)\n",
    "    Xy_train_to_use_tata[name] = (X_train, y_train)\n",
    "    Xy_lockbox_to_use_tata[name] = (X_lockbox, y_lockbox)\n",
    "\n",
    "models_to_try_tata = {\n",
    "    'RF': models_to_try['RF']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "initial_profile_df_tata = pd.DataFrame(\n",
    "    columns=['model', 'xy', 'train_r2', 'val_r2']\n",
    ")\n",
    "\n",
    "N_CV = 5\n",
    "\n",
    "for model_name, model in models_to_try_tata.items():\n",
    "    for Xy_name, (X_train_raw, y_train_raw) in Xy_train_to_use_tata.items():\n",
    "        \n",
    "        X_train_norm = StandardScaler().fit_transform(X_train_raw)\n",
    "        y_train_norm = StandardScaler().fit_transform(y_train_raw.values.reshape(-1, 1))\n",
    "        y_train_log = np.log(y_train_raw + ETA)\n",
    "        y_train_norm_log = StandardScaler().fit_transform(y_train_log.values.reshape(-1, 1))\n",
    "        \n",
    "        X_list = [('X_raw', X_train_raw), ('X_norm', X_train_norm)]\n",
    "        y_list = [('y_raw', y_train_raw), ('y_norm', y_train_norm),\n",
    "                 ('y_log', y_train_log), ('y_norm_log', y_train_norm_log)]\n",
    "        \n",
    "        Xy_final = {}\n",
    "        for (X_lab, X_mat), (y_lab, y_mat) in product(X_list, y_list):\n",
    "            Xy_lab_final = f'{Xy_name}__{X_lab}__{y_lab}'\n",
    "            Xy_final[Xy_lab_final] = (X_mat, y_mat)\n",
    "            \n",
    "        for Xy_full_name, (X_to_use, y_to_use) in Xy_final.items():\n",
    "\n",
    "            print(f'{model_name}: {Xy_full_name}')\n",
    "\n",
    "            cv_result = cross_validate(\n",
    "                model,\n",
    "                X_to_use,\n",
    "                y=y_to_use,\n",
    "                cv=N_CV,\n",
    "                scoring='r2',\n",
    "                return_train_score=True,\n",
    "                verbose=1,\n",
    "                n_jobs=N_CV\n",
    "            )\n",
    "\n",
    "            sub_result_df = pd.DataFrame(\n",
    "                data={\n",
    "                    'model': [model_name] * N_CV,\n",
    "                    'xy': [Xy_full_name] * N_CV,\n",
    "                    'train_r2': cv_result['train_score'],\n",
    "                    'val_r2': cv_result['test_score']\n",
    "                }\n",
    "            )\n",
    "\n",
    "            initial_profile_df_tata = initial_profile_df_tata.append(sub_result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='xy', y='train_r2', data=initial_profile_df_tata, dodge=True, ax=ax)\n",
    "ax.set_xlabel('')\n",
    "ax.tick_params(axis='x', rotation=90)\n",
    "ax.set_ylabel('Training R-Squared')\n",
    "ax.set_ylim(bottom=0, top=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "_, ax = plt.subplots(figsize=(10, 5))\n",
    "sns.boxplot(x='xy', y='val_r2', data=initial_profile_df_tata, dodge=True, ax=ax)\n",
    "ax.set_xlabel('')\n",
    "ax.tick_params(axis='x', rotation=90)\n",
    "ax.set_ylabel('Validation R-Squared')\n",
    "ax.set_ylim(0, 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Feature Importances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_fi_tata, y_fi_tata = Xy_train_to_use_tata['tata']\n",
    "\n",
    "X_fi_tata = pd.DataFrame(StandardScaler().fit_transform(X_fi_tata), columns=X_fi_tata.columns,\n",
    "                         index=X_fi_tata.index)\n",
    "y_fi_tata = np.log(y_fi_tata + ETA)\n",
    "\n",
    "X_fi_tata_train, X_fi_tata_val, y_fi_tata_train, y_tata_go_val = train_test_split(\n",
    "    X_fi_tata, y_fi_tata, test_size=0.15, random_state=35)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_fi_tata = models_to_try_tata['RF'].fit(X_fi_tata_train, y_fi_tata_train)\n",
    "explainer = shap.TreeExplainer(rf_fi_tata)\n",
    "shap_values = explainer.shap_values(X_fi_tata_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shap.summary_plot(shap_values, X_fi_tata_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alt -10 Box Distance Metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Significance of -10 Box Effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "box_10_counts = gene_table_ml['box_10_seq'].value_counts()\n",
    "box_10_common = box_10_counts[box_10_counts > 10]\n",
    "\n",
    "box_10_u_test_df = pd.DataFrame(columns=box_10_common.index, index=box_10_common.index)\n",
    "box_10_u_test_df = box_10_u_test_df.astype(float)\n",
    "\n",
    "for box_10_a, box_10_b in product(box_10_common.index, box_10_common.index):\n",
    "    box_10_a_exps = gene_table_ml[gene_table_ml['box_10_seq'] == box_10_a]['mean_fpkm']\n",
    "    box_10_b_exps = gene_table_ml[gene_table_ml['box_10_seq'] == box_10_b]['mean_fpkm']\n",
    "    u_stat, p_val = mannwhitneyu(box_10_a_exps, box_10_b_exps, alternative='two-sided')\n",
    "    box_10_u_test_df.loc[box_10_a, box_10_b] = p_val\n",
    "\n",
    "reject, pval_corr = multicomp(box_10_u_test_df.values, method='fdr_bh')\n",
    "sig_df = pd.DataFrame(reject, columns=box_10_common.index, index=box_10_common.index)\n",
    "\n",
    "_, ax = plt.subplots(figsize=(10, 7))\n",
    "sns.heatmap(sig_df, ax=ax)\n",
    "ax.tick_params(axis='y', rotation=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Next Steps - 2/2/21"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Redo the above analysis with just the first gene in each TU that has a TATAAT box and TSS A\n",
    "- Ask Anand about estimating noise levels (technical, amplification) from raw RNAseq data\n",
    "- Add 5'UTR length as feature\n",
    "- Add transcript length as feature\n",
    "- Add transcript GC content\n",
    "- Investigate better distance metric for -10 box scores\n",
    "    - Wilcoxon test for significance of expression difference for different -10 box scores\n",
    "    - does DNA shape describe differences between these scores?\n",
    "    - Sizhe's linear discriminant method?\n",
    "    \n",
    "- try the one-hot model that centers some features on the pre-discovered locations of features\n",
    "    - i.e. +/-10 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Next Steps - 2/9/21\n",
    "\n",
    "- port this workflow over to E. coli to see where we're at there\n",
    "- try to get TRN/regulatory information for M. florum\n",
    "    - from ICA of M. pneumo\n",
    "    - from BLAST of putative M. florum regulators against M. pneumo (or maybe even E coli)\n",
    "- get the shape function added and actually look at -10 box shape\n",
    "    - with the goal of a distance metric for motifs that has something to do with expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "12px",
    "width": "160px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "459px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
