{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "697a14e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports and loadings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "sys.path.insert(0, '../functions/')\n",
    "import interface_GAMS as iG\n",
    "import conversion_equations as ce\n",
    "sys.path.append('/home/chris/github/Strainware-s/')\n",
    "sys.path.append('/home/chris/github/Strainware-s/bitome2/')\n",
    "import bitome\n",
    "from bitome.core import Bitome\n",
    "from workflows_CAD.feature_functions import *\n",
    "import pickle\n",
    "import os\n",
    "\n",
    "# settings\n",
    "testing = True\n",
    "    \n",
    "# load in settings flags\n",
    "if testing:\n",
    "    flags_filepath = '../options/test_gene_flags.csv'\n",
    "    TF_flags_filepath = '../options/test_TF_flags.csv'\n",
    "else:\n",
    "    flags_filepath = '../options/gene_flags.csv'\n",
    "    TF_flags_filepath = '../options/TF_flags.csv'\n",
    "flags_df = pd.read_csv(flags_filepath, index_col = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd5d4d1f",
   "metadata": {},
   "source": [
    "# RNAP Concentration per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "25edc637",
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# settings\n",
    "bott_pct = 0.01\n",
    "top_pct = 0.99\n",
    "\n",
    "# setup\n",
    "# merge together log_tpm_df files\n",
    "log_tpm_df = pd.read_csv('../data/external/imodulon_info/log_tpm.csv', index_col = 0)\n",
    "starve_log_tpm = pd.read_csv('../data/external/validation_data_sets/stationary_phase/cleaned_log_tpm_qc.csv', index_col = 0)\n",
    "to_blank_inds = list(set(log_tpm_df.index) - set(starve_log_tpm.index))\n",
    "# need to create zero rows for missing values\n",
    "zeros_data = {col : 0 for col in starve_log_tpm.columns}\n",
    "zeros_df = pd.DataFrame(zeros_data, index = to_blank_inds)\n",
    "starve_log_tpm = pd.concat([starve_log_tpm, zeros_df])\n",
    "starve_log_tpm = starve_log_tpm.loc[log_tpm_df.index]\n",
    "log_tpm_df = pd.concat([starve_log_tpm, log_tpm_df], axis = 1)\n",
    "tpm_df = 2**log_tpm_df\n",
    "\n",
    "# let's find the fold change in the middle 95% percentile of this distribution, make the current RNAP the center, and scale thigns to both sides\n",
    "# do this for both genes, each sample gets the RNAP value that's the average of both\n",
    "RNAP_default = 10**-6\n",
    "fold_modify = 2 # modify the fold amount by this factor (divided by if below, multiplied by if above)\n",
    "\n",
    "genes = ['b3987', 'b3988']\n",
    "RNAP_df = pd.DataFrame(index = tpm_df.columns)\n",
    "for gene in genes:\n",
    "    # calculations\n",
    "    bott = tpm_df.loc[gene].sort_values().iloc[int(bott_pct*len(tpm_df.columns))]\n",
    "    top = tpm_df.loc[gene].sort_values().iloc[int(top_pct*len(tpm_df.columns))]\n",
    "    fold_change = top/bott\n",
    "    middle = np.median(tpm_df.loc[gene])\n",
    "\n",
    "    # scale everything\n",
    "    new_col = []\n",
    "    for sample in tpm_df.columns:\n",
    "        # set outleirs to the min/max\n",
    "        if tpm_df.loc[gene][sample] > top:\n",
    "            tpm = top\n",
    "        elif tpm_df.loc[gene][sample] < bott:\n",
    "            tpm = bott\n",
    "        else:\n",
    "            tpm = tpm_df.loc[gene][sample]\n",
    "        \n",
    "        # scale to fold change, then by RNAP value\n",
    "        fold_change = tpm / middle\n",
    "        if tpm < middle:\n",
    "            fold_change /= fold_modify\n",
    "        else:\n",
    "            fold_change *= fold_modify\n",
    "        new_val = RNAP_default * fold_change\n",
    "        \n",
    "        new_col.append(new_val)\n",
    "    RNAP_df[gene] = new_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d59c58b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save it off\n",
    "RNAP_mean_df = pd.DataFrame(RNAP_df.mean(axis = 1))\n",
    "RNAP_mean_df.columns = ['RNAP']\n",
    "RNAP_mean_df.to_csv('../data/interim/sample_constants/RNAP_conc.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76a4e1b6",
   "metadata": {},
   "source": [
    "# Sequence based motif scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f4eef02c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Alias location.position is deprecated and will be removed in a future release. Use location directly, or int(location). However, that will fail for UnknownPosition, and for OneOfPosition and WithinPosition will give the default rather than left-most value.\n"
     ]
    }
   ],
   "source": [
    "# setup\n",
    "bitome_genbank = Bitome('../data/external/sequence_info/NC_000913.3.gb')\n",
    "bitome_fasta = Bitome('../data/external/sequence_info/NC_000913.3.fasta')\n",
    "tu_table = pd.read_csv('../data/external/sequence_info/tu.csv')\n",
    "tf_pwm_db = rpwm('../data/external/sequence_info/motif_pwm_db.txt')\n",
    "tfbs = pd.read_csv('../data/external/sequence_info/TFBS_regulondb.csv')\n",
    "TF_flags_df = pd.read_csv(TF_flags_filepath, index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1f1da8b6",
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# runs pipeline for all of our relevant genes\n",
    "gene_to_operon_to_motif_scores_dist_to_TSS = {}\n",
    "    \n",
    "genes_test = TF_flags_df.TF\n",
    "for gene in genes_test:\n",
    "    gene = gene[0].upper()+gene[1:]\n",
    "    if gene not in tf_pwm_db:\n",
    "        continue\n",
    "    gene_to_operon_to_motif_scores_dist_to_TSS.update({gene : {}})\n",
    "    gene_pwm_old = tf_pwm_db[gene]\n",
    "    gene_pssm = {base: [pos_dict[base] for pos_dict in gene_pwm_old] for base in 'ATCG'}\n",
    "\n",
    "    # pull out data\n",
    "    actual_val = []\n",
    "    gene_rows = tfbs[[val.upper() == gene.upper() for val in tfbs['TF_name']]]\n",
    "    left_end_positions = gene_rows['TF_bs_left_end_position'].tolist()\n",
    "    right_end_positions = gene_rows['TF_bs_right_end_position'].tolist()\n",
    "    for index, row in gene_rows.iterrows():\n",
    "        name = row['Transcription_unit_name_regulated_by_TF']\n",
    "        left = row['TF_bs_left_end_position']\n",
    "        right = row['TF_bs_right_end_position']\n",
    "        try:\n",
    "            actual_val.append([name, int(left), int(right)])\n",
    "        except:\n",
    "            continue\n",
    "    actual_df = pd.DataFrame(actual_val, columns=['name', 'left', 'right'])\n",
    "    \n",
    "    # tu_dict?\n",
    "    tu = list(actual_df['name'])\n",
    "    tu_dict = {item: tu.count(item) for item in tu}\n",
    "    \n",
    "    tu_table = pd.read_csv('../data/external/sequence_info/tu.csv')\n",
    "    tu_name=[]\n",
    "    locus_tags = []\n",
    "    score_gene=[]\n",
    "    location_left =[]\n",
    "    location_right = []\n",
    "    matches = []\n",
    "    for index, row in tu_table.iterrows():\n",
    "        i = row['strand']\n",
    "        j = row['tss']\n",
    "        k = row['name']\n",
    "        #i,j,k in zip(tu_table['strand'],tu_table['tss'],tu_table['name']):\n",
    "        for m,n in (tu_dict.items()):\n",
    "            if i == 1.0 and j.is_integer():\n",
    "                if m==k:\n",
    "                    #print(m)\n",
    "                    tu_name.append([m] * n)  \n",
    "                    locus_tags.append([row['locus_tag']] * n)\n",
    "                    score_gene.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['log_odds'].tolist())\n",
    "                    location_left.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['left'].tolist())\n",
    "                    location_right.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['right'].tolist())\n",
    "                    matches.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['match_sequence'].tolist())\n",
    "            if i == -1.0 and j.is_integer():\n",
    "                if m==k:\n",
    "                    tu_name.append([m] * n)\n",
    "                    locus_tags.append([row['locus_tag']] * n)\n",
    "                    score_gene.append(bitome_fasta.motif_search(j-50, j+100, -1, gene_pssm, n_best_matches=n)['log_odds'].tolist())\n",
    "                    location_left.append(bitome_fasta.motif_search(j-50, j+100, -1, gene_pssm, n_best_matches=n)['left'].tolist())\n",
    "                    location_right.append(bitome_fasta.motif_search(j-50, j+100, -1, gene_pssm, n_best_matches=n)['right'].tolist())\n",
    "                    matches.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['match_sequence'].tolist())\n",
    "            if not j.is_integer():\n",
    "                if m==k:\n",
    "                    tu_name.append([m] * 1)\n",
    "                    locus_tags.append([row['locus_tag']] * 1)\n",
    "                    score_gene.append([0])\n",
    "                    location_left.append([0])\n",
    "                    location_right.append([0])\n",
    "                    matches.append([''])\n",
    "                    \n",
    "    tu_name_flat_list = [item for sublist in tu_name for item in sublist]\n",
    "    score_gene_flat_list = [item for sublist in score_gene for item in sublist]\n",
    "    location_left_flat_list = [item for sublist in location_left for item in sublist]\n",
    "    location_right_flat_list = [item for sublist in location_right for item in sublist]\n",
    "    matches_list = [item for sublist in matches for item in sublist]\n",
    "    locus_tags = [item for sublist in locus_tags for item in sublist]\n",
    "\n",
    "    data = {\n",
    "        'TU': tu_name_flat_list,\n",
    "        'locus_tag' : locus_tags,\n",
    "        'Score': score_gene_flat_list,\n",
    "        'Location_left': location_left_flat_list,\n",
    "        'Location_right': location_right_flat_list,\n",
    "        'Match_sequence' : matches_list,\n",
    "    }\n",
    "    predicted_df = pd.DataFrame(data)\n",
    "    #predicted_df = predicted_df.drop([2, 3])\n",
    "    predicted_df = predicted_df.reset_index(drop=True)\n",
    "    predicted_df = predicted_df.drop_duplicates()\n",
    "    predicted_df.to_pickle('../data/interim/bitome_results/'+gene+'.pkl')\n",
    "    \n",
    "    for index, row in predicted_df.iterrows():\n",
    "        if row['TU'] not in gene_to_operon_to_motif_scores_dist_to_TSS[gene]:\n",
    "            gene_to_operon_to_motif_scores_dist_to_TSS[gene].update({row['TU'] : []})\n",
    "        bby_TU = tu_table[tu_table['locus_tag'] == row['locus_tag']]\n",
    "        if bby_TU['strand'].values[0] == -1:\n",
    "            # aimed to the left, so TSS - motif\n",
    "            dist_to_TSS = bby_TU['tss'].values[0] - ((row['Location_left'] + row['Location_right']) / 2)\n",
    "        else:\n",
    "            # aimed to the right, so motif - TSS\n",
    "            dist_to_TSS = ((row['Location_left'] + row['Location_right']) / 2) - bby_TU['tss'].values[0]\n",
    "        #dist_to_TSS = min([abs(bby_TU['tss'] - row['Location_left']).values[0], abs(bby_TU['tss'] - row['Location_right']).values[0]])\n",
    "        next_val = (row['Score'], dist_to_TSS)\n",
    "        gene_to_operon_to_motif_scores_dist_to_TSS[gene][row['TU']].append(next_val)\n",
    "\n",
    "pickle_out = open('../data/interim/misc_dictionaries/gene_to_operon_to_motif_scores_dist_to_TSS.pkl', 'wb')\n",
    "pickle.dump(gene_to_operon_to_motif_scores_dist_to_TSS, pickle_out)\n",
    "pickle_out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b58a112",
   "metadata": {},
   "source": [
    "# correlation matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4903a399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first we need to make a ratio df\n",
    "\n",
    "# merge together log_tpm_df files\n",
    "log_tpm_df = pd.read_csv('../data/external/imodulon_info/log_tpm.csv', index_col = 0)\n",
    "starve_log_tpm = pd.read_csv('../data/external/validation_data_sets/stationary_phase/cleaned_log_tpm_qc.csv', index_col = 0)\n",
    "to_blank_inds = list(set(log_tpm_df.index) - set(starve_log_tpm.index))\n",
    "# need to create zero rows for missing values\n",
    "zeros_data = {col : 0 for col in starve_log_tpm.columns}\n",
    "zeros_df = pd.DataFrame(zeros_data, index = to_blank_inds)\n",
    "starve_log_tpm = pd.concat([starve_log_tpm, zeros_df])\n",
    "starve_log_tpm = starve_log_tpm.loc[log_tpm_df.index]\n",
    "log_tpm_df = pd.concat([starve_log_tpm, log_tpm_df], axis = 1)\n",
    "\n",
    "cases = set([(row['act_iM'], row['inh_iM']) for _, row in flags_df.iterrows()])\n",
    "collect_actuals = []\n",
    "collect_predictions = []\n",
    "for iMs_run in cases:\n",
    "    clean = [x for x in iMs_run if str(x) != 'nan']\n",
    "    case = '__'.join([str(iM) for iM in iMs_run]).replace(' ', '_').replace('/', '_')\n",
    "    if not os.path.exists('../GAMS/runs/'+case):\n",
    "        continue\n",
    "\n",
    "    # let's find our run_dir\n",
    "    try:\n",
    "        GAMs_run_dir = '../GAMS/runs/'+case\n",
    "        mRNA_df, GAMS_calc_cAct, cAct_kd_df, act_metab_df, act_kd_metab_df, GAMS_calc_cInh, cInh_kd_df, inh_metab_df, inh_kd_metab_df = iG.read_multi_GAMs(GAMs_run_dir)\n",
    "    except:\n",
    "        continue # Usually means it hasn't been run yet\n",
    "    actual_ratio = ce.log_tpm_df_to_mRNA_ratio_df(log_tpm_df.loc[mRNA_df.columns], flags_df.loc[mRNA_df.columns])\n",
    "\n",
    "    collect_actuals.append(actual_ratio)\n",
    "    collect_predictions.append(mRNA_df)\n",
    "\n",
    "merge_actual = pd.concat(collect_actuals, axis = 1)\n",
    "merge_pred = pd.concat(collect_predictions, axis = 1)\n",
    "overlap = list(set(merge_actual.index).intersection(set(merge_pred.index)))\n",
    "actual_mRNA_df = merge_actual.loc[:,~merge_actual.columns.duplicated()].astype(float)\n",
    "pred_mRNA_df = merge_pred.loc[:,~merge_pred.columns.duplicated()].astype(float)\n",
    "\n",
    "# create sample-sample and gene-gene correlation dataframe\n",
    "sample_corr_df = pd.DataFrame(index = actual_mRNA_df.index, columns = actual_mRNA_df.index)\n",
    "for sample1 in pred_mRNA_df.index:\n",
    "    for sample2 in pred_mRNA_df.index:\n",
    "        sample_corr_df.at[sample1, sample2] = actual_mRNA_df.loc[sample1].corr(pred_mRNA_df.loc[sample2])\n",
    "gene_corr_df = pd.DataFrame(index = actual_mRNA_df.columns, columns = actual_mRNA_df.columns)\n",
    "for sample1 in pred_mRNA_df.columns:\n",
    "    for sample2 in pred_mRNA_df.columns:\n",
    "        gene_corr_df.at[sample1, sample2] = actual_mRNA_df[sample1].corr(pred_mRNA_df[sample2])\n",
    "\n",
    "# these things take a while to run, let's save them off\n",
    "sample_corr_df.to_csv('../data/interim/samples_correlation_df.csv')\n",
    "gene_corr_df.to_csv('../data/interim/genes_correlation_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "878dcfdb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
