{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "697a14e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports and loadings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "sys.path.insert(0, '../functions/')\n",
    "import interface_GAMS as iG\n",
    "import conversion_equations as ce\n",
    "sys.path.append('/home/chris/github/Strainware-s/')\n",
    "sys.path.append('/home/chris/github/Strainware-s/bitome2/')\n",
    "import bitome\n",
    "from bitome.core import Bitome\n",
    "from workflows_CAD.feature_functions import *\n",
    "import pickle\n",
    "import os\n",
    "import ast\n",
    "\n",
    "# load in settings flags\n",
    "settings_df = pd.read_csv('../options/settings.csv', index_col = 0)\n",
    "flags_filepath = settings_df.loc['gene_flags_filepath']['Setting']\n",
    "TF_flags_filepath = settings_df.loc['TF_flags_filepath']['Setting']\n",
    "\n",
    "flags_df = pd.read_csv(flags_filepath, index_col = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd5d4d1f",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# RNAP Concentration per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "25edc637",
   "metadata": {
    "code_folding": [],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# settings\n",
    "bott_pct = 0.01\n",
    "top_pct = 0.99\n",
    "\n",
    "# setup\n",
    "# merge together log_tpm_df files\n",
    "log_tpm_df = pd.read_csv('../data/external/imodulon_info/log_tpm.csv', index_col = 0)\n",
    "starve_log_tpm = pd.read_csv('../data/external/validation_data_sets/stationary_phase/cleaned_log_tpm_qc.csv', index_col = 0)\n",
    "to_blank_inds = list(set(log_tpm_df.index) - set(starve_log_tpm.index))\n",
    "# need to create zero rows for missing values\n",
    "zeros_data = {col : 0 for col in starve_log_tpm.columns}\n",
    "zeros_df = pd.DataFrame(zeros_data, index = to_blank_inds)\n",
    "starve_log_tpm = pd.concat([starve_log_tpm, zeros_df])\n",
    "starve_log_tpm = starve_log_tpm.loc[log_tpm_df.index]\n",
    "log_tpm_df = pd.concat([starve_log_tpm, log_tpm_df], axis = 1)\n",
    "tpm_df = 2**log_tpm_df\n",
    "\n",
    "# let's find the fold change in the middle 95% percentile of this distribution, make the current RNAP the center, and scale thigns to both sides\n",
    "# do this for both genes, each sample gets the RNAP value that's the average of both\n",
    "RNAP_default = 10**-6\n",
    "fold_modify = 2 # modify the fold amount by this factor (divided by if below, multiplied by if above)\n",
    "\n",
    "genes = ['b3987', 'b3988']\n",
    "RNAP_df = pd.DataFrame(index = tpm_df.columns)\n",
    "for gene in genes:\n",
    "    # calculations\n",
    "    bott = tpm_df.loc[gene].sort_values().iloc[int(bott_pct*len(tpm_df.columns))]\n",
    "    top = tpm_df.loc[gene].sort_values().iloc[int(top_pct*len(tpm_df.columns))]\n",
    "    fold_change = top/bott\n",
    "    middle = np.median(tpm_df.loc[gene])\n",
    "\n",
    "    # scale everything\n",
    "    new_col = []\n",
    "    for sample in tpm_df.columns:\n",
    "        # set outleirs to the min/max\n",
    "        if tpm_df.loc[gene][sample] > top:\n",
    "            tpm = top\n",
    "        elif tpm_df.loc[gene][sample] < bott:\n",
    "            tpm = bott\n",
    "        else:\n",
    "            tpm = tpm_df.loc[gene][sample]\n",
    "        \n",
    "        # scale to fold change, then by RNAP value\n",
    "        fold_change = tpm / middle\n",
    "        if tpm < middle:\n",
    "            fold_change /= fold_modify\n",
    "        else:\n",
    "            fold_change *= fold_modify\n",
    "        new_val = RNAP_default * fold_change\n",
    "        \n",
    "        new_col.append(new_val)\n",
    "    RNAP_df[gene] = new_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d59c58b2",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# save it off\n",
    "RNAP_mean_df = pd.DataFrame(RNAP_df.mean(axis = 1))\n",
    "RNAP_mean_df.columns = ['RNAP']\n",
    "RNAP_mean_df.to_csv('../data/interim/sample_constants/RNAP_conc.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76a4e1b6",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Sequence based motif scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f4eef02c",
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Alias location.position is deprecated and will be removed in a future release. Use location directly, or int(location). However, that will fail for UnknownPosition, and for OneOfPosition and WithinPosition will give the default rather than left-most value.\n"
     ]
    }
   ],
   "source": [
    "# setup\n",
    "bitome_genbank = Bitome('../data/external/sequence_info/NC_000913.3.gb')\n",
    "bitome_fasta = Bitome('../data/external/sequence_info/NC_000913.3.fasta')\n",
    "tu_table = pd.read_csv('../data/external/sequence_info/tu.csv')\n",
    "tf_pwm_db = rpwm('../data/external/sequence_info/motif_pwm_db.txt')\n",
    "tfbs = pd.read_csv('../data/external/sequence_info/TFBS_regulondb.csv')\n",
    "TF_flags_df = pd.read_csv(TF_flags_filepath, index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1f1da8b6",
   "metadata": {
    "code_folding": [
     0
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# runs pipeline for all of our relevant genes\n",
    "gene_to_operon_to_motif_scores_dist_to_TSS = {}\n",
    "    \n",
    "genes_test = TF_flags_df.TF\n",
    "for gene in genes_test:\n",
    "    gene = gene[0].upper()+gene[1:]\n",
    "    if gene not in tf_pwm_db:\n",
    "        continue\n",
    "    gene_to_operon_to_motif_scores_dist_to_TSS.update({gene : {}})\n",
    "    gene_pwm_old = tf_pwm_db[gene]\n",
    "    gene_pssm = {base: [pos_dict[base] for pos_dict in gene_pwm_old] for base in 'ATCG'}\n",
    "\n",
    "    # pull out data\n",
    "    actual_val = []\n",
    "    gene_rows = tfbs[[val.upper() == gene.upper() for val in tfbs['TF_name']]]\n",
    "    left_end_positions = gene_rows['TF_bs_left_end_position'].tolist()\n",
    "    right_end_positions = gene_rows['TF_bs_right_end_position'].tolist()\n",
    "    for index, row in gene_rows.iterrows():\n",
    "        name = row['Transcription_unit_name_regulated_by_TF']\n",
    "        left = row['TF_bs_left_end_position']\n",
    "        right = row['TF_bs_right_end_position']\n",
    "        try:\n",
    "            actual_val.append([name, int(left), int(right)])\n",
    "        except:\n",
    "            continue\n",
    "    actual_df = pd.DataFrame(actual_val, columns=['name', 'left', 'right'])\n",
    "    \n",
    "    # tu_dict?\n",
    "    tu = list(actual_df['name'])\n",
    "    tu_dict = {item: tu.count(item) for item in tu}\n",
    "    \n",
    "    tu_table = pd.read_csv('../data/external/sequence_info/tu.csv')\n",
    "    tu_name=[]\n",
    "    locus_tags = []\n",
    "    score_gene=[]\n",
    "    location_left =[]\n",
    "    location_right = []\n",
    "    matches = []\n",
    "    for index, row in tu_table.iterrows():\n",
    "        i = row['strand']\n",
    "        j = row['tss']\n",
    "        k = row['name']\n",
    "        #i,j,k in zip(tu_table['strand'],tu_table['tss'],tu_table['name']):\n",
    "        for m,n in (tu_dict.items()):\n",
    "            if i == 1.0 and j.is_integer():\n",
    "                if m==k:\n",
    "                    #print(m)\n",
    "                    tu_name.append([m] * n)  \n",
    "                    locus_tags.append([row['locus_tag']] * n)\n",
    "                    score_gene.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['log_odds'].tolist())\n",
    "                    location_left.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['left'].tolist())\n",
    "                    location_right.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['right'].tolist())\n",
    "                    matches.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['match_sequence'].tolist())\n",
    "            if i == -1.0 and j.is_integer():\n",
    "                if m==k:\n",
    "                    tu_name.append([m] * n)\n",
    "                    locus_tags.append([row['locus_tag']] * n)\n",
    "                    score_gene.append(bitome_fasta.motif_search(j-50, j+100, -1, gene_pssm, n_best_matches=n)['log_odds'].tolist())\n",
    "                    location_left.append(bitome_fasta.motif_search(j-50, j+100, -1, gene_pssm, n_best_matches=n)['left'].tolist())\n",
    "                    location_right.append(bitome_fasta.motif_search(j-50, j+100, -1, gene_pssm, n_best_matches=n)['right'].tolist())\n",
    "                    matches.append(bitome_fasta.motif_search(j-100, j+50, 1, gene_pssm, n_best_matches=n)['match_sequence'].tolist())\n",
    "            if not j.is_integer():\n",
    "                if m==k:\n",
    "                    tu_name.append([m] * 1)\n",
    "                    locus_tags.append([row['locus_tag']] * 1)\n",
    "                    score_gene.append([0])\n",
    "                    location_left.append([0])\n",
    "                    location_right.append([0])\n",
    "                    matches.append([''])\n",
    "                    \n",
    "    tu_name_flat_list = [item for sublist in tu_name for item in sublist]\n",
    "    score_gene_flat_list = [item for sublist in score_gene for item in sublist]\n",
    "    location_left_flat_list = [item for sublist in location_left for item in sublist]\n",
    "    location_right_flat_list = [item for sublist in location_right for item in sublist]\n",
    "    matches_list = [item for sublist in matches for item in sublist]\n",
    "    locus_tags = [item for sublist in locus_tags for item in sublist]\n",
    "\n",
    "    data = {\n",
    "        'TU': tu_name_flat_list,\n",
    "        'locus_tag' : locus_tags,\n",
    "        'Score': score_gene_flat_list,\n",
    "        'Location_left': location_left_flat_list,\n",
    "        'Location_right': location_right_flat_list,\n",
    "        'Match_sequence' : matches_list,\n",
    "    }\n",
    "    predicted_df = pd.DataFrame(data)\n",
    "    #predicted_df = predicted_df.drop([2, 3])\n",
    "    predicted_df = predicted_df.reset_index(drop=True)\n",
    "    predicted_df = predicted_df.drop_duplicates()\n",
    "    predicted_df.to_pickle('../data/interim/bitome_results/'+gene+'.pkl')\n",
    "    \n",
    "    for index, row in predicted_df.iterrows():\n",
    "        if row['TU'] not in gene_to_operon_to_motif_scores_dist_to_TSS[gene]:\n",
    "            gene_to_operon_to_motif_scores_dist_to_TSS[gene].update({row['TU'] : []})\n",
    "        bby_TU = tu_table[tu_table['locus_tag'] == row['locus_tag']]\n",
    "        if bby_TU['strand'].values[0] == -1:\n",
    "            # aimed to the left, so TSS - motif\n",
    "            dist_to_TSS = bby_TU['tss'].values[0] - ((row['Location_left'] + row['Location_right']) / 2)\n",
    "        else:\n",
    "            # aimed to the right, so motif - TSS\n",
    "            dist_to_TSS = ((row['Location_left'] + row['Location_right']) / 2) - bby_TU['tss'].values[0]\n",
    "        #dist_to_TSS = min([abs(bby_TU['tss'] - row['Location_left']).values[0], abs(bby_TU['tss'] - row['Location_right']).values[0]])\n",
    "        next_val = (row['Score'], dist_to_TSS)\n",
    "        gene_to_operon_to_motif_scores_dist_to_TSS[gene][row['TU']].append(next_val)\n",
    "\n",
    "pickle_out = open('../data/interim/misc_dictionaries/gene_to_operon_to_motif_scores_dist_to_TSS.pkl', 'wb')\n",
    "pickle.dump(gene_to_operon_to_motif_scores_dist_to_TSS, pickle_out)\n",
    "pickle_out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee439695",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# select Kd for metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a0d15a44",
   "metadata": {
    "code_folding": [
     45
    ],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# setting for KdMetabolite\n",
    "min_max_fold_change = 100 # min will be min metab divided by this, max will be max metab times this\n",
    "\n",
    "# loading\n",
    "starve_metabs = pd.read_csv('../data/external/validation_data_sets/stationary_phase/cleaned_metabolites_data.csv', index_col = 0)\n",
    "starve_metabs[starve_metabs.columns[1:]] *= 1e-6\n",
    "\n",
    "# KdMetabolite needs to change more varied condition by condition, maybe let it be a variable and just set its upper and lower limits here?\n",
    "# it shouldn't change between samples though, just one hard number\n",
    "TF_flags_df = pd.read_csv(TF_flags_filepath, index_col = 0)\n",
    "\n",
    "\n",
    "# first let's find min and max to set for those that don't have anything\n",
    "min_kd = 9e99\n",
    "max_kd = 0\n",
    "for index, row in TF_flags_df.iterrows():\n",
    "    act_ct = 0\n",
    "    inh_ct = 0\n",
    "    for col in TF_flags_df.columns:\n",
    "        if 'cAct' in col:\n",
    "            act_ct += TF_flags_df.at[index, col]\n",
    "        elif 'cInh' in col:\n",
    "            inh_ct += TF_flags_df.at[index, col]\n",
    "    if act_ct > 0:\n",
    "        change_col = 'kd_act_metab'\n",
    "        other = 'kd_inh_metab'\n",
    "    if inh_ct > 0:\n",
    "        change_col = 'kd_inh_metab'\n",
    "        other = 'kd_act_metab'\n",
    "    try:\n",
    "        metabolites = ast.literal_eval(row['effectors'])\n",
    "    except:\n",
    "        metabolites = []\n",
    "    metabolites = list(set(metabolites).intersection(starve_metabs.index))\n",
    "    if len(metabolites) == 0:\n",
    "        continue\n",
    "    \n",
    "    KdArg_max = max(starve_metabs.loc[metabolites][starve_metabs.columns[1:]].sum().values.flatten())\n",
    "    KdArg_min = min(starve_metabs.loc[metabolites][starve_metabs.columns[1:]].sum().values.flatten())\n",
    "    if KdArg_min < min_kd:\n",
    "        min_kd = KdArg_min\n",
    "    if KdArg_max > max_kd:\n",
    "        max_kd = KdArg_max\n",
    "\n",
    "for index, row in TF_flags_df.iterrows():\n",
    "    act_ct = 0\n",
    "    inh_ct = 0\n",
    "    for col in TF_flags_df.columns:\n",
    "        if 'cAct' in col:\n",
    "            act_ct += TF_flags_df.at[index, col]\n",
    "        elif 'cInh' in col:\n",
    "            inh_ct += TF_flags_df.at[index, col]\n",
    "    if act_ct > 0:\n",
    "        change_col = 'kd_act_metab'\n",
    "        other = 'kd_inh_metab'\n",
    "    if inh_ct > 0:\n",
    "        change_col = 'kd_inh_metab'\n",
    "        other = 'kd_act_metab'\n",
    "    try:\n",
    "        metabolites = ast.literal_eval(row['effectors'])\n",
    "    except:\n",
    "        metabolites = []\n",
    "    metabolites = list(set(metabolites).intersection(starve_metabs.index))\n",
    "    if len(metabolites) == 0:\n",
    "        # set dummies and continue\n",
    "        TF_flags_df.at[index, 'kd_act_metab_up'] = max_kd * min_max_fold_change\n",
    "        TF_flags_df.at[index, 'kd_act_metab_lo'] = min_kd / min_max_fold_change\n",
    "        TF_flags_df.at[index, 'kd_inh_metab_up'] = max_kd * min_max_fold_change\n",
    "        TF_flags_df.at[index, 'kd_inh_metab_lo'] = min_kd / min_max_fold_change\n",
    "        continue\n",
    "    \n",
    "    KdArg_max = max(starve_metabs.loc[metabolites][starve_metabs.columns[1:]].sum().values.flatten())\n",
    "    KdArg_min = min(starve_metabs.loc[metabolites][starve_metabs.columns[1:]].sum().values.flatten())\n",
    "    TF_flags_df.at[index, change_col+'_up'] = KdArg_max * min_max_fold_change\n",
    "    TF_flags_df.at[index, change_col+'_lo'] = KdArg_min / min_max_fold_change\n",
    "    \n",
    "    # need to set dummy values to the other (won't be used, but will error out GAMS if I don't)\n",
    "    TF_flags_df.at[index, other+'_up'] = KdArg_max * min_max_fold_change\n",
    "    TF_flags_df.at[index, other+'_lo'] = KdArg_min / min_max_fold_change\n",
    "    \n",
    "TF_flags_df.fillna(0).to_csv(TF_flags_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7930e506",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# proteomics conversions for constraints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e0654e23",
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "invalid value encountered in scalar divide\n",
      "invalid value encountered in scalar divide\n",
      "invalid value encountered in scalar divide\n",
      "invalid value encountered in scalar divide\n",
      "invalid value encountered in scalar divide\n",
      "invalid value encountered in scalar divide\n",
      "invalid value encountered in scalar divide\n"
     ]
    }
   ],
   "source": [
    "# loading\n",
    "metab_df = pd.read_excel('../data/external/validation_data_sets/heinemann_protein_conc.xlsx', index_col = 0)\n",
    "metab_cols = ['Glucose', 'Acetate', 'Glycerol', 'Pyruvate', 'Galactose ', 'Succinate', 'Fructose']\n",
    "\n",
    "# load in un-logged tpm\n",
    "# merge together log_tpm_df files\n",
    "log_tpm_df = pd.read_csv('../data/external/imodulon_info/log_tpm.csv', index_col = 0)\n",
    "starve_log_tpm = pd.read_csv('../data/external/validation_data_sets/stationary_phase/cleaned_log_tpm_qc.csv', index_col = 0)\n",
    "to_blank_inds = list(set(log_tpm_df.index) - set(starve_log_tpm.index))\n",
    "# need to create zero rows for missing values\n",
    "zeros_data = {col : 0 for col in starve_log_tpm.columns}\n",
    "zeros_df = pd.DataFrame(zeros_data, index = to_blank_inds)\n",
    "starve_log_tpm = pd.concat([starve_log_tpm, zeros_df])\n",
    "starve_log_tpm = starve_log_tpm.loc[log_tpm_df.index]\n",
    "log_tpm_df = pd.concat([starve_log_tpm, log_tpm_df], axis = 1)\n",
    "\n",
    "\n",
    "tpm_df = 2**log_tpm_df\n",
    "overlap = list(set(tpm_df.index).intersection(set(metab_df.index)))\n",
    "metab_df = metab_df.loc[overlap]\n",
    "tpm_df = tpm_df.loc[overlap]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# scale tpm_df by the minimum and maximum\n",
    "new_df = pd.DataFrame(index = tpm_df.index, columns = tpm_df.columns)\n",
    "\n",
    "vals = metab_df[metab_cols].values.flatten()\n",
    "vals = vals[vals != 0]\n",
    "global_min = min(vals)\n",
    "global_max = max(vals)\n",
    "\n",
    "to_rem = []\n",
    "for gene in overlap:\n",
    "    min_val = min(metab_df[metab_cols].loc[gene])\n",
    "    max_val = max(metab_df[metab_cols].loc[gene])\n",
    "    if min_val == max_val: # no good data, just pull from overall max and min\n",
    "        min_val = global_min\n",
    "        max_val = global_max\n",
    "        to_rem.append(gene) # used to use this to remove samples\n",
    "\n",
    "    tpm_min_val = min(tpm_df.loc[gene])\n",
    "    tpm_max_val = max(tpm_df.loc[gene])\n",
    "    \n",
    "    for sample in tpm_df.columns:\n",
    "        scaled_tpm = (tpm_df.loc[gene][sample] - tpm_min_val) / (tpm_max_val - tpm_min_val)\n",
    "        new_val = scaled_tpm*(max_val - min_val) + min_val\n",
    "        new_df.at[gene, sample] = new_val\n",
    "new_df.to_csv('../data/external/validation_data_sets/converted_log_tpm_in_M.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eac3ca0",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
