{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "787b4bb4-f92c-405b-945d-342205c2312d",
   "metadata": {},
   "source": [
    "# Import Packages and Construct Bitome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "141c4647-47bb-40fa-aa04-4a7df7f5e0d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import re\n",
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "sys.path.append('../../bitome2')\n",
    "\n",
    "from bitome.core import Bitome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9c340b08-8450-492b-82bd-7dcab64dd10b",
   "metadata": {},
   "outputs": [],
   "source": [
    "K12_DATA_PATH = Path('../data/bitome2/mg1655')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "0bdf55e5-9781-40d4-a25a-d2dcb36bff66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from GenBank annotation\n",
    "origin = (3925743, 3925975)\n",
    "# from 1. Duggin, I. G. & Bell, S. D. J. Mol. Biol. (2009). with the following terA/terC sequences:\n",
    "# ter_a = 'AATTAGTATGTTGTAACTAAAGT'\n",
    "# ter_c = 'ATATAGGATGTTGTAACTAATAT'\n",
    "terminus = (1341745, 1609180)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "69f0acee-dfd9-41e3-af97-ed9bbfc6e844",
   "metadata": {},
   "outputs": [],
   "source": [
    "mg1655 = Bitome(\n",
    "    Path(K12_DATA_PATH, 'NC_000913.3.gb'),\n",
    "    name='MG1655',\n",
    "    origin=origin, terminus=terminus,\n",
    "    gene_table=Path(K12_DATA_PATH, 'gene_info_supp.csv'),\n",
    "    tu_table=Path(K12_DATA_PATH, 'tu.csv'),\n",
    "    operon_table=Path(K12_DATA_PATH, 'operon.csv'),\n",
    "    tss_table=Path(K12_DATA_PATH, 'tss.csv'),\n",
    "    # TODO PARSE TTS data\n",
    "    tts_table=None,\n",
    "    tfbs_table=Path(K12_DATA_PATH, 'tfbs.csv'),\n",
    "    terminator_table=Path(K12_DATA_PATH, 'terminator.csv'),\n",
    "    attenuator_table=Path(K12_DATA_PATH, 'attenuator.csv'),\n",
    "    rbs_table=Path(K12_DATA_PATH, 'rbs.csv'),\n",
    "    riboswitch_table=Path(K12_DATA_PATH, 'riboswitch.csv')\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2775e470-44b2-466a-afcc-09643bdd74de",
   "metadata": {},
   "source": [
    "# Promoter Table Example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "c0d737ee-748b-4830-bfd8-08b2924cc156",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genome_loc_sin': 0.8090169943749473,\n",
       " 'genome_loc_cos': 0.5877852522924732,\n",
       " 'rep_region': 'lagging',\n",
       " 'origin_dist': 1506716,\n",
       " 'tu_len': 4319.0,\n",
       " 'tu_gc': 0.5448020375086826,\n",
       " 'tu_order': 1.0,\n",
       " 'tss_dist': 26.0,\n",
       " 'tss_base': 'A',\n",
       " 'utr_len': 26.0,\n",
       " 'utr_gc': 0.3846153846153846,\n",
       " 'box_10_seq': 'TATGCT',\n",
       " 'box_10_tss_dist': 12.0,\n",
       " 'box_10_ext_gc': 0.33333333333333337,\n",
       " 'box_35_seq': 'ATGTCA',\n",
       " 'box_35_tss_dist': 36.0,\n",
       " 'spacer_len': 18.0,\n",
       " 'spacer_gc': 0.38888888888888884,\n",
       " 'tu_tm': 47.122741534424165,\n",
       " 'tu_gatc': False,\n",
       " 'sigma_factor': 'Sigma70, Sigma38'}"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tu_id = 'ECK120009218'\n",
    "\n",
    "tu_row = mg1655.tu_table.loc[tu_id]\n",
    "\n",
    "# get the first gene of this TU\n",
    "def get_tu_first_gene(tu_id, strand):\n",
    "    \n",
    "    tu_genes = mg1655._tu_to_genes[tu_id]\n",
    "    tu_gene_df = mg1655.gene_table.loc[tu_genes]\n",
    "    if strand == 1:\n",
    "        sorted_gene_df = tu_gene_df.sort_values(by='left', ascending=True)\n",
    "    else:\n",
    "        sorted_gene_df = tu_gene_df.sort_values(by='right', ascending=False)\n",
    "    return sorted_gene_df.index[0]\n",
    "\n",
    "first_gene_id = get_tu_first_gene(tu_row.name, tu_strand)\n",
    "first_gene_row = mg1655.gene_table.loc[first_gene_id]\n",
    "\n",
    "gene_info_dict = mg1655.genome_organization_for_gene(first_gene_row, primary_tu=True)\n",
    "gene_info_dict['sigma_factor'] = tu_row.sigma_factor\n",
    "gene_info_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d17eb232-c8d7-46e6-999d-3ecaf78103be",
   "metadata": {},
   "source": [
    "# Regulatory States"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "24311a9c-4ea5-46ee-8b74-cf443e827c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "REGULONDB_PATH = Path('../data/bitome2/mg1655/regulondb10.6/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "9aceec31-fb4b-4ad3-9b28-b7be4d5e70b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_regulon_db_file(filename) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Given a filename for a raw TXT file from RegulonDB, parses into a DataFrame\n",
    "    :param Union[str, Path] filename: the filename of the RegulonDB TXT file to read into a pandas DataFrame\n",
    "    :return pd.DataFrame regulon_df: a pandas DataFrame parsed from a RegulonDB raw TXT file\n",
    "    \"\"\"\n",
    "\n",
    "    full_filename = Path(REGULONDB_PATH, filename)\n",
    "\n",
    "    with open(full_filename, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    comment = True\n",
    "    i = 0\n",
    "    while comment:\n",
    "        if lines[i].startswith('#'):\n",
    "            i += 1\n",
    "        else:\n",
    "            comment = False\n",
    "\n",
    "    names = [line[5:-1] for line in lines if re.match(r'# \\d', line)]\n",
    "\n",
    "    df = pd.read_csv(full_filename, index_col=None, skiprows=i, sep='\\t', header=None, names=names)\n",
    "\n",
    "    return df.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "ab529742-5870-45a6-aead-e46808035b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_strand_int(strand):\n",
    "    if strand in ['forward', '+', 1, 'FWD']:\n",
    "        return 1\n",
    "    elif strand in ['reverse', '-', -1, 'REV']:\n",
    "        return -1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "ec3356b3-b87a-446c-9bdd-6448c1a76d0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>mode</th>\n",
       "      <th>REGULATORY_INTERACTION_ID</th>\n",
       "      <th>PROMOTER_ID</th>\n",
       "      <th>final_state</th>\n",
       "      <th>tf</th>\n",
       "      <th>strand</th>\n",
       "      <th>REG_PHRASE_ID</th>\n",
       "      <th>PHRASE</th>\n",
       "      <th>REG_PHRASE_FUNCTION</th>\n",
       "      <th>TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>612650</td>\n",
       "      <td>612668</td>\n",
       "      <td>repressor</td>\n",
       "      <td>ECK120033853</td>\n",
       "      <td>ECK120010221</td>\n",
       "      <td>Fur-Fe&lt;SUP&gt;2+&lt;/SUP&gt;</td>\n",
       "      <td>Fur</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12T292271</td>\n",
       "      <td>[Fur,-]</td>\n",
       "      <td>repressor</td>\n",
       "      <td>proximal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>612650</td>\n",
       "      <td>612668</td>\n",
       "      <td>repressor</td>\n",
       "      <td>ECK120033883</td>\n",
       "      <td>ECK120010299</td>\n",
       "      <td>Fur-Fe&lt;SUP&gt;2+&lt;/SUP&gt;</td>\n",
       "      <td>Fur</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12T292271</td>\n",
       "      <td>[Fur,-]</td>\n",
       "      <td>repressor</td>\n",
       "      <td>proximal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4516744</td>\n",
       "      <td>4516762</td>\n",
       "      <td>repressor</td>\n",
       "      <td>ECK120033831</td>\n",
       "      <td>ECK120010548</td>\n",
       "      <td>Fur-Fe&lt;SUP&gt;2+&lt;/SUP&gt;</td>\n",
       "      <td>Fur</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12T292271</td>\n",
       "      <td>[Fur,-]</td>\n",
       "      <td>repressor</td>\n",
       "      <td>proximal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1619005</td>\n",
       "      <td>1619019</td>\n",
       "      <td>activator</td>\n",
       "      <td>ECK120034760</td>\n",
       "      <td>ECK120010245</td>\n",
       "      <td>Fis</td>\n",
       "      <td>Fis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12T292285</td>\n",
       "      <td>[Fis,+]</td>\n",
       "      <td>activator</td>\n",
       "      <td>proximal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3305958</td>\n",
       "      <td>3305975</td>\n",
       "      <td>activator</td>\n",
       "      <td>ECK120031815</td>\n",
       "      <td>ECK120011032</td>\n",
       "      <td>TyrR-tyrosine</td>\n",
       "      <td>TyrR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12T292433</td>\n",
       "      <td>[TyrR,+]</td>\n",
       "      <td>activator</td>\n",
       "      <td>remote</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      left    right       mode REGULATORY_INTERACTION_ID   PROMOTER_ID  \\\n",
       "0   612650   612668  repressor              ECK120033853  ECK120010221   \n",
       "1   612650   612668  repressor              ECK120033883  ECK120010299   \n",
       "2  4516744  4516762  repressor              ECK120033831  ECK120010548   \n",
       "3  1619005  1619019  activator              ECK120034760  ECK120010245   \n",
       "4  3305958  3305975  activator              ECK120031815  ECK120011032   \n",
       "\n",
       "           final_state    tf  strand REG_PHRASE_ID    PHRASE  \\\n",
       "0  Fur-Fe<SUP>2+</SUP>   Fur     NaN  ECK12T292271   [Fur,-]   \n",
       "1  Fur-Fe<SUP>2+</SUP>   Fur     NaN  ECK12T292271   [Fur,-]   \n",
       "2  Fur-Fe<SUP>2+</SUP>   Fur     NaN  ECK12T292271   [Fur,-]   \n",
       "3                  Fis   Fis     NaN  ECK12T292285   [Fis,+]   \n",
       "4        TyrR-tyrosine  TyrR     NaN  ECK12T292433  [TyrR,+]   \n",
       "\n",
       "  REG_PHRASE_FUNCTION      TYPE  \n",
       "0           repressor  proximal  \n",
       "1           repressor  proximal  \n",
       "2           repressor  proximal  \n",
       "3           activator  proximal  \n",
       "4           activator    remote  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_phrase_df = read_regulon_db_file('reg_phrase.txt')\n",
    "reg_phrase_df = reg_phrase_df[['REG_PHRASE_ID', 'PHRASE', 'REG_PHRASE_FUNCTION']]\n",
    "reg_phrase_link_df = read_regulon_db_file('reg_phrase_ri_link.txt')\n",
    "\n",
    "reg_phrase_merged_df = reg_phrase_df.merge(reg_phrase_link_df, on='REG_PHRASE_ID', how='left')\n",
    "tfbs_reg_phrase = mg1655.tfbs_table.merge(reg_phrase_merged_df, on='REGULATORY_INTERACTION_ID', how='left')\n",
    "tfbs_reg_phrase.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "a210a55e-4940-4633-9334-369a1faf0bc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[GalR,-]\n",
      "[GalS,-]\n",
      "[CRP,+]\n",
      "[H-NS,-]\n",
      "[HU,-]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>mode</th>\n",
       "      <th>REGULATORY_INTERACTION_ID</th>\n",
       "      <th>PROMOTER_ID</th>\n",
       "      <th>final_state</th>\n",
       "      <th>tf</th>\n",
       "      <th>strand</th>\n",
       "      <th>REG_PHRASE_ID</th>\n",
       "      <th>PHRASE</th>\n",
       "      <th>REG_PHRASE_FUNCTION</th>\n",
       "      <th>TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1630</th>\n",
       "      <td>792059</td>\n",
       "      <td>792092</td>\n",
       "      <td>repressor</td>\n",
       "      <td>ECK120030995</td>\n",
       "      <td>ECK120010122</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12T292364</td>\n",
       "      <td>[HU,-]</td>\n",
       "      <td>repressor</td>\n",
       "      <td>proximal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        left   right       mode REGULATORY_INTERACTION_ID   PROMOTER_ID  \\\n",
       "1630  792059  792092  repressor              ECK120030995  ECK120010122   \n",
       "\n",
       "     final_state  tf  strand REG_PHRASE_ID  PHRASE REG_PHRASE_FUNCTION  \\\n",
       "1630          HU  HU     NaN  ECK12T292364  [HU,-]           repressor   \n",
       "\n",
       "          TYPE  \n",
       "1630  proximal  "
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# hard-code the promoter ID we care about\n",
    "prom_id = 'ECK120010122'\n",
    "\n",
    "phrase_for_gal = tfbs_reg_phrase[tfbs_reg_phrase['PROMOTER_ID'] == prom_id]\n",
    "\n",
    "for phrase, phrase_df in phrase_for_gal.groupby('REG_PHRASE_ID'):\n",
    "    print(phrase_df.iloc[0]['PHRASE'])\n",
    "    \n",
    "phrase_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3722c172-3090-4419-b76c-2ef76081c60f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Promoter Sweep Variables"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1f9f765-07fc-4dfb-b07b-e71f084c3769",
   "metadata": {},
   "source": [
    "## Motifs\n",
    "\n",
    "Just a single sigma shown, but same for any motif with PSSM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "fdfdc169-b2a8-41d4-9571-1a221c49e32a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sigma_motif_dict = mg1655.create_sigma_motifs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "d4bc8f21-6eba-42a9-b882-7ec13273590c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>match_sequence</th>\n",
       "      <th>log_odds</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>ATTTCA</td>\n",
       "      <td>-7.88548</td>\n",
       "      <td>792081.0</td>\n",
       "      <td>792086.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TATTTC</td>\n",
       "      <td>1.290253</td>\n",
       "      <td>792082.0</td>\n",
       "      <td>792087.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>TTATTT</td>\n",
       "      <td>-2.124794</td>\n",
       "      <td>792083.0</td>\n",
       "      <td>792088.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>GTTATT</td>\n",
       "      <td>-4.060738</td>\n",
       "      <td>792084.0</td>\n",
       "      <td>792089.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>GGTTAT</td>\n",
       "      <td>-2.798393</td>\n",
       "      <td>792085.0</td>\n",
       "      <td>792090.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>CCTCTC</td>\n",
       "      <td>-8.610154</td>\n",
       "      <td>792272.0</td>\n",
       "      <td>792277.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>CCCTCT</td>\n",
       "      <td>-4.975958</td>\n",
       "      <td>792273.0</td>\n",
       "      <td>792278.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>ACCCTC</td>\n",
       "      <td>-10.79639</td>\n",
       "      <td>792274.0</td>\n",
       "      <td>792279.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CACCCT</td>\n",
       "      <td>0.199219</td>\n",
       "      <td>792275.0</td>\n",
       "      <td>792280.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>GCACCC</td>\n",
       "      <td>-9.652913</td>\n",
       "      <td>792276.0</td>\n",
       "      <td>792281.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    match_sequence  log_odds      left     right\n",
       "147         ATTTCA  -7.88548  792081.0  792086.0\n",
       "16          TATTTC  1.290253  792082.0  792087.0\n",
       "56          TTATTT -2.124794  792083.0  792088.0\n",
       "79          GTTATT -4.060738  792084.0  792089.0\n",
       "61          GGTTAT -2.798393  792085.0  792090.0\n",
       "..             ...       ...       ...       ...\n",
       "155         CCTCTC -8.610154  792272.0  792277.0\n",
       "95          CCCTCT -4.975958  792273.0  792278.0\n",
       "191         ACCCTC -10.79639  792274.0  792279.0\n",
       "24          CACCCT  0.199219  792275.0  792280.0\n",
       "175         GCACCC -9.652913  792276.0  792281.0\n",
       "\n",
       "[196 rows x 4 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mg1655.promoter_motif_search(tu_row.tss, tu_strand, sigma_motif_dict['Sigma70']['-10'].pssm, n_up=200, n_down=0, n_best_matches=200).sort_values(by='right')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3acb4455-d401-473f-9d0a-585632e5976e",
   "metadata": {},
   "source": [
    "## Shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "2124c9be-cc6c-482a-a059-f80c8cfe346b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fivemer_shape_df = pd.read_csv('../data/5mer_shape_lookup.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "53b9414a-a954-4d5d-a220-f9a06ff27121",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dna_shape(seq):\n",
    "    \"\"\"\n",
    "    Given a sequence, compute all 14 shape params\n",
    "    \"\"\"\n",
    "    \n",
    "    seq_fivemers = [seq[i:i+5] for i in range(len(seq) - 5 + 1)]\n",
    "    seq_fivemer_shape_df = fivemer_shape_df.loc[seq_fivemers]\n",
    "    \n",
    "    return seq_fivemer_shape_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "60c6ecde-f2f3-4076-af8e-428c8af2bb88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Opening</th>\n",
       "      <th>Rise</th>\n",
       "      <th>Stretch</th>\n",
       "      <th>EP</th>\n",
       "      <th>Tilt</th>\n",
       "      <th>Shear</th>\n",
       "      <th>ProT</th>\n",
       "      <th>Buckle</th>\n",
       "      <th>Shift</th>\n",
       "      <th>HelT</th>\n",
       "      <th>Stagger</th>\n",
       "      <th>Slide</th>\n",
       "      <th>MGW</th>\n",
       "      <th>Roll</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATACC</th>\n",
       "      <td>-0.57</td>\n",
       "      <td>3.340</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-7.03</td>\n",
       "      <td>-0.320</td>\n",
       "      <td>0.11</td>\n",
       "      <td>-8.36</td>\n",
       "      <td>3.27</td>\n",
       "      <td>0.130</td>\n",
       "      <td>34.625</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-1.495</td>\n",
       "      <td>5.46</td>\n",
       "      <td>1.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCA</th>\n",
       "      <td>0.27</td>\n",
       "      <td>3.345</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>-5.40</td>\n",
       "      <td>-0.610</td>\n",
       "      <td>0.27</td>\n",
       "      <td>-5.96</td>\n",
       "      <td>2.66</td>\n",
       "      <td>0.045</td>\n",
       "      <td>34.340</td>\n",
       "      <td>0.07</td>\n",
       "      <td>-1.600</td>\n",
       "      <td>5.13</td>\n",
       "      <td>-1.565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCAT</th>\n",
       "      <td>0.26</td>\n",
       "      <td>3.465</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>-5.41</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.27</td>\n",
       "      <td>-4.33</td>\n",
       "      <td>0.80</td>\n",
       "      <td>-0.210</td>\n",
       "      <td>34.715</td>\n",
       "      <td>0.06</td>\n",
       "      <td>-1.660</td>\n",
       "      <td>4.97</td>\n",
       "      <td>1.410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCATA</th>\n",
       "      <td>-0.60</td>\n",
       "      <td>3.245</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-6.63</td>\n",
       "      <td>0.270</td>\n",
       "      <td>0.09</td>\n",
       "      <td>-6.69</td>\n",
       "      <td>-2.86</td>\n",
       "      <td>-0.185</td>\n",
       "      <td>33.115</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>-1.420</td>\n",
       "      <td>5.88</td>\n",
       "      <td>0.200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATAA</th>\n",
       "      <td>-1.00</td>\n",
       "      <td>3.310</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-6.88</td>\n",
       "      <td>0.390</td>\n",
       "      <td>-0.10</td>\n",
       "      <td>-8.06</td>\n",
       "      <td>4.86</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>33.010</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-1.390</td>\n",
       "      <td>6.02</td>\n",
       "      <td>2.320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGGC</th>\n",
       "      <td>0.28</td>\n",
       "      <td>3.450</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-5.64</td>\n",
       "      <td>-0.505</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>-1.50</td>\n",
       "      <td>3.50</td>\n",
       "      <td>0.200</td>\n",
       "      <td>33.455</td>\n",
       "      <td>0.14</td>\n",
       "      <td>-1.845</td>\n",
       "      <td>4.51</td>\n",
       "      <td>-2.685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGCG</th>\n",
       "      <td>0.40</td>\n",
       "      <td>3.420</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>-5.21</td>\n",
       "      <td>0.580</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>-2.82</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.015</td>\n",
       "      <td>35.325</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-1.725</td>\n",
       "      <td>4.77</td>\n",
       "      <td>-2.130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGCGA</th>\n",
       "      <td>0.42</td>\n",
       "      <td>3.370</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-5.06</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.99</td>\n",
       "      <td>0.17</td>\n",
       "      <td>-0.070</td>\n",
       "      <td>34.400</td>\n",
       "      <td>0.05</td>\n",
       "      <td>-1.580</td>\n",
       "      <td>5.40</td>\n",
       "      <td>0.810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGAT</th>\n",
       "      <td>0.44</td>\n",
       "      <td>3.370</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-5.24</td>\n",
       "      <td>0.485</td>\n",
       "      <td>-0.26</td>\n",
       "      <td>-6.96</td>\n",
       "      <td>-2.84</td>\n",
       "      <td>-0.130</td>\n",
       "      <td>34.155</td>\n",
       "      <td>-0.08</td>\n",
       "      <td>-1.555</td>\n",
       "      <td>5.28</td>\n",
       "      <td>1.745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGATA</th>\n",
       "      <td>-0.73</td>\n",
       "      <td>3.245</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-7.12</td>\n",
       "      <td>-0.085</td>\n",
       "      <td>0.10</td>\n",
       "      <td>-7.63</td>\n",
       "      <td>-2.58</td>\n",
       "      <td>-0.090</td>\n",
       "      <td>33.785</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>-1.480</td>\n",
       "      <td>5.32</td>\n",
       "      <td>-1.910</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Opening   Rise  Stretch    EP   Tilt  Shear  ProT  Buckle  Shift  \\\n",
       "ATACC    -0.57  3.340    -0.03 -7.03 -0.320   0.11 -8.36    3.27  0.130   \n",
       "TACCA     0.27  3.345    -0.04 -5.40 -0.610   0.27 -5.96    2.66  0.045   \n",
       "ACCAT     0.26  3.465    -0.05 -5.41  0.500   0.27 -4.33    0.80 -0.210   \n",
       "CCATA    -0.60  3.245    -0.02 -6.63  0.270   0.09 -6.69   -2.86 -0.185   \n",
       "CATAA    -1.00  3.310    -0.02 -6.88  0.390  -0.10 -8.06    4.86 -0.040   \n",
       "...        ...    ...      ...   ...    ...    ...   ...     ...    ...   \n",
       "AAGGC     0.28  3.450    -0.03 -5.64 -0.505  -0.28 -1.50    3.50  0.200   \n",
       "AGGCG     0.40  3.420    -0.04 -5.21  0.580  -0.30 -2.82    0.23  0.015   \n",
       "GGCGA     0.42  3.370    -0.03 -5.06  0.165   0.25 -2.99    0.17 -0.070   \n",
       "GCGAT     0.44  3.370    -0.03 -5.24  0.485  -0.26 -6.96   -2.84 -0.130   \n",
       "CGATA    -0.73  3.245    -0.02 -7.12 -0.085   0.10 -7.63   -2.58 -0.090   \n",
       "\n",
       "         HelT  Stagger  Slide   MGW   Roll  \n",
       "ATACC  34.625     0.00 -1.495  5.46  1.400  \n",
       "TACCA  34.340     0.07 -1.600  5.13 -1.565  \n",
       "ACCAT  34.715     0.06 -1.660  4.97  1.410  \n",
       "CCATA  33.115    -0.07 -1.420  5.88  0.200  \n",
       "CATAA  33.010    -0.01 -1.390  6.02  2.320  \n",
       "...       ...      ...    ...   ...    ...  \n",
       "AAGGC  33.455     0.14 -1.845  4.51 -2.685  \n",
       "AGGCG  35.325     0.02 -1.725  4.77 -2.130  \n",
       "GGCGA  34.400     0.05 -1.580  5.40  0.810  \n",
       "GCGAT  34.155    -0.08 -1.555  5.28  1.745  \n",
       "CGATA  33.785    -0.06 -1.480  5.32 -1.910  \n",
       "\n",
       "[197 rows x 14 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_dna_shape(str(mg1655.get_sequence(tu_row.tss - 200, tu_row.tss, -1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4a0c882-3540-4a01-92a1-6afd98013e89",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
